// n8n-nodes-base/dist/nodes/MongoDb/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.connectMongoClient = exports.stringifyObjectIDs = exports.prepareFields = exports.prepareItems = exports.validateAndResolveMongoCredentials = exports.buildMongoConnectionParams = exports.buildParameterizedConnString = void 0;
const tls_1 = require("tls");
const n8n_workflow_1 = require("n8n-workflow");
const get_1 = __importDefault(require("lodash/get"));
const set_1 = __importDefault(require("lodash/set"));
const mongodb_1 = require("mongodb");
const utilities_1 = require("../../utils/utilities");
function buildParameterizedConnString(credentials) {
    if (credentials.port) {
        return `mongodb://${credentials.user}:${credentials.password}@${credentials.host}:${credentials.port}`;
    }
    else {
        return `mongodb+srv://${credentials.user}:${credentials.password}@${credentials.host}`;
    }
}
exports.buildParameterizedConnString = buildParameterizedConnString;
function buildMongoConnectionParams(self, credentials) {
    const sanitizedDbName = credentials.database && credentials.database.trim().length > 0
        ? credentials.database.trim()
        : '';
    if (credentials.configurationType === 'connectionString') {
        if (credentials.connectionString && credentials.connectionString.trim().length > 0) {
            return {
                connectionString: credentials.connectionString.trim(),
                database: sanitizedDbName,
            };
        }
        else {
            throw new n8n_workflow_1.NodeOperationError(self.getNode(), 'Cannot override credentials: valid MongoDB connection string not provided ');
        }
    }
    else {
        return {
            connectionString: buildParameterizedConnString(credentials),
            database: sanitizedDbName,
        };
    }
}
exports.buildMongoConnectionParams = buildMongoConnectionParams;
function validateAndResolveMongoCredentials(self, credentials) {
    if (credentials === undefined) {
        throw new n8n_workflow_1.NodeOperationError(self.getNode(), 'No credentials got returned!');
    }
    else {
        return buildMongoConnectionParams(self, credentials);
    }
}
exports.validateAndResolveMongoCredentials = validateAndResolveMongoCredentials;
function prepareItems(items, fields, updateKey = '', useDotNotation = false, dateFields = []) {
    let data = items;
    if (updateKey) {
        if (!fields.includes(updateKey)) {
            fields.push(updateKey);
        }
        data = items.filter((item) => item.json[updateKey] !== undefined);
    }
    const preperedItems = data.map(({ json }) => {
        const updateItem = {};
        for (const field of fields) {
            let fieldData;
            if (useDotNotation) {
                fieldData = (0, get_1.default)(json, field, null);
            }
            else {
                fieldData = json[field] !== undefined ? json[field] : null;
            }
            if (fieldData && dateFields.includes(field)) {
                fieldData = new Date(fieldData);
            }
            if (useDotNotation) {
                (0, set_1.default)(updateItem, field, fieldData);
            }
            else {
                updateItem[field] = fieldData;
            }
        }
        return updateItem;
    });
    return preperedItems;
}
exports.prepareItems = prepareItems;
function prepareFields(fields) {
    return fields
        .split(',')
        .map((field) => field.trim())
        .filter((field) => !!field);
}
exports.prepareFields = prepareFields;
function stringifyObjectIDs(items) {
    items.forEach((item) => {
        if (item._id instanceof mongodb_1.ObjectId) {
            item.json._id = item._id.toString();
        }
        if (item.id instanceof mongodb_1.ObjectId) {
            item.json.id = item.id.toString();
        }
    });
    return items;
}
exports.stringifyObjectIDs = stringifyObjectIDs;
async function connectMongoClient(connectionString, credentials = {}) {
    let client;
    if (credentials.tls) {
        const ca = credentials.ca ? (0, utilities_1.formatPrivateKey)(credentials.ca) : undefined;
        const cert = credentials.cert ? (0, utilities_1.formatPrivateKey)(credentials.cert) : undefined;
        const key = credentials.key ? (0, utilities_1.formatPrivateKey)(credentials.key) : undefined;
        const passphrase = credentials.passphrase || undefined;
        const secureContext = (0, tls_1.createSecureContext)({
            ca,
            cert,
            key,
            passphrase,
        });
        client = await mongodb_1.MongoClient.connect(connectionString, {
            tls: true,
            secureContext,
        });
    }
    else {
        client = await mongodb_1.MongoClient.connect(connectionString);
    }
    return client;
}
exports.connectMongoClient = connectMongoClient;
