// n8n-nodes-base/dist/nodes/GoToWebinar/GoToWebinar.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoToWebinar = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const descriptions_1 = require("./descriptions");
const GenericFunctions_1 = require("./GenericFunctions");
const lodash_1 = require("lodash");
const moment = require("moment-timezone");
class GoToWebinar {
    constructor() {
        this.description = {
            displayName: 'GoToWebinar',
            name: 'goToWebinar',
            icon: 'file:gotowebinar.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume the GoToWebinar API',
            defaults: {
                name: 'GoToWebinar',
                color: '#0097e1',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'goToWebinarOAuth2Api',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    options: [
                        {
                            name: 'Attendee',
                            value: 'attendee',
                        },
                        {
                            name: 'Co-Organizer',
                            value: 'coorganizer',
                        },
                        {
                            name: 'Panelist',
                            value: 'panelist',
                        },
                        {
                            name: 'Registrant',
                            value: 'registrant',
                        },
                        {
                            name: 'Session',
                            value: 'session',
                        },
                        {
                            name: 'Webinar',
                            value: 'webinar',
                        },
                    ],
                    default: 'attendee',
                    description: 'Resource to consume',
                },
                ...descriptions_1.attendeeOperations,
                ...descriptions_1.attendeeFields,
                ...descriptions_1.coorganizerOperations,
                ...descriptions_1.coorganizerFields,
                ...descriptions_1.panelistOperations,
                ...descriptions_1.panelistFields,
                ...descriptions_1.registrantOperations,
                ...descriptions_1.registrantFields,
                ...descriptions_1.sessionOperations,
                ...descriptions_1.sessionFields,
                ...descriptions_1.webinarOperations,
                ...descriptions_1.webinarFields,
            ],
        };
        this.methods = {
            loadOptions: {
                async getWebinars() {
                    return await GenericFunctions_1.loadWebinars.call(this);
                },
                async getAnswers() {
                    return await GenericFunctions_1.loadAnswers.call(this);
                },
                async getWebinarSessions() {
                    return await GenericFunctions_1.loadWebinarSessions.call(this);
                },
                async getTimezones() {
                    const returnData = [];
                    for (const timezone of moment.tz.names()) {
                        const timezoneName = timezone;
                        const timezoneId = timezone;
                        returnData.push({
                            name: timezoneName,
                            value: timezoneId,
                        });
                    }
                    return returnData;
                },
                async getRegistranSimpleQuestions() {
                    return await GenericFunctions_1.loadRegistranSimpleQuestions.call(this);
                },
                async getRegistranMultiChoiceQuestions() {
                    return await GenericFunctions_1.loadRegistranMultiChoiceQuestions.call(this);
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        let responseData;
        const returnData = [];
        const { oauthTokenData } = await this.getCredentials('goToWebinarOAuth2Api');
        const accountKey = oauthTokenData.account_key;
        const organizerKey = oauthTokenData.organizer_key;
        for (let i = 0; i < items.length; i++) {
            try {
                if (resource === 'attendee') {
                    if (operation === 'get') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const sessionKey = this.getNodeParameter('sessionKey', i);
                        const registrantKey = this.getNodeParameter('registrantKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/sessions/${sessionKey}/attendees/${registrantKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'GET', endpoint, {}, {});
                    }
                    else if (operation === 'getAll') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const sessionKey = this.getNodeParameter('sessionKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/sessions/${sessionKey}/attendees`;
                        responseData = await GenericFunctions_1.handleGetAll.call(this, endpoint, {}, {}, resource);
                    }
                    else if (operation === 'getDetails') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const sessionKey = this.getNodeParameter('sessionKey', i);
                        const registrantKey = this.getNodeParameter('registrantKey', i);
                        const details = this.getNodeParameter('details', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/sessions/${sessionKey}/attendees/${registrantKey}/${details}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'GET', endpoint, {}, {});
                    }
                }
                else if (resource === 'coorganizer') {
                    if (operation === 'create') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const body = {
                            external: this.getNodeParameter('isExternal', i),
                        };
                        if (body.external === false) {
                            body.organizerKey = this.getNodeParameter('organizerKey', i);
                        }
                        if (body.external === true) {
                            body.givenName = this.getNodeParameter('givenName', i);
                            body.email = this.getNodeParameter('email', i);
                        }
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/coorganizers`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'POST', endpoint, {}, [body]);
                    }
                    else if (operation === 'delete') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const coorganizerKey = this.getNodeParameter('coorganizerKey', i);
                        const qs = {
                            external: this.getNodeParameter('isExternal', i),
                        };
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/coorganizers/${coorganizerKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'DELETE', endpoint, qs, {});
                        responseData = { success: true };
                    }
                    else if (operation === 'getAll') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/coorganizers`;
                        responseData = await GenericFunctions_1.handleGetAll.call(this, endpoint, {}, {}, resource);
                    }
                    else if (operation === 'reinvite') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const coorganizerKey = this.getNodeParameter('coorganizerKey', i);
                        const qs = {
                            external: this.getNodeParameter('isExternal', i),
                        };
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/coorganizers/${coorganizerKey}/resendInvitation`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'POST', endpoint, qs, {});
                        responseData = { success: true };
                    }
                }
                else if (resource === 'panelist') {
                    if (operation === 'create') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const body = [
                            {
                                name: this.getNodeParameter('name', i),
                                email: this.getNodeParameter('email', i),
                            },
                        ];
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/panelists`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'POST', endpoint, {}, body);
                    }
                    else if (operation === 'delete') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const panelistKey = this.getNodeParameter('panelistKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/panelists/${panelistKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'DELETE', endpoint, {}, {});
                        responseData = { success: true };
                    }
                    else if (operation === 'getAll') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/panelists`;
                        responseData = await GenericFunctions_1.handleGetAll.call(this, endpoint, {}, {}, resource);
                    }
                    else if (operation === 'reinvite') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const panelistKey = this.getNodeParameter('panelistKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/panelists/${panelistKey}/resendInvitation`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'POST', endpoint, {}, {});
                        responseData = { success: true };
                    }
                }
                else if (resource === 'registrant') {
                    if (operation === 'create') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const qs = {};
                        const body = {
                            firstName: this.getNodeParameter('firstName', i),
                            lastName: this.getNodeParameter('lastName', i),
                            email: this.getNodeParameter('email', i),
                            responses: [],
                        };
                        let additionalFields = this.getNodeParameter('additionalFields', i);
                        if (additionalFields.resendConfirmation) {
                            qs.resendConfirmation = additionalFields.resendConfirmation;
                            additionalFields = lodash_1.omit(additionalFields, ['resendConfirmation']);
                        }
                        if (additionalFields.fullAddress) {
                            Object.assign(body, additionalFields.fullAddress.details);
                            additionalFields = lodash_1.omit(additionalFields, ['fullAddress']);
                        }
                        if (additionalFields.simpleResponses) {
                            body.responses.push(...additionalFields.simpleResponses.details);
                            additionalFields = lodash_1.omit(additionalFields, ['simpleResponses']);
                        }
                        if (additionalFields.multiChoiceResponses) {
                            body.responses.push(...additionalFields.multiChoiceResponses.details);
                            additionalFields = lodash_1.omit(additionalFields, ['multiChoiceResponses']);
                        }
                        Object.assign(body, additionalFields);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/registrants`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'POST', endpoint, qs, body);
                    }
                    else if (operation === 'delete') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const registrantKey = this.getNodeParameter('registrantKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/registrants/${registrantKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'DELETE', endpoint, {}, {});
                        responseData = { success: true };
                    }
                    else if (operation === 'get') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const registrantKey = this.getNodeParameter('registrantKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/registrants/${registrantKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'GET', endpoint, {}, {});
                    }
                    else if (operation === 'getAll') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/registrants`;
                        responseData = await GenericFunctions_1.handleGetAll.call(this, endpoint, {}, {}, resource);
                    }
                }
                else if (resource === 'session') {
                    if (operation === 'get') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const sessionKey = this.getNodeParameter('sessionKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/sessions/${sessionKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'GET', endpoint, {}, {});
                    }
                    else if (operation === 'getAll') {
                        const qs = {};
                        const returnAll = this.getNodeParameter('returnAll', 0);
                        if (!returnAll) {
                            qs.limit = this.getNodeParameter('limit', 0);
                        }
                        const { webinarKey, times, } = this.getNodeParameter('additionalFields', i);
                        if (times) {
                            qs.fromTime = moment(times.timesProperties.fromTime).format();
                            qs.toTime = moment(times.timesProperties.toTime).format();
                        }
                        else {
                            qs.fromTime = moment().subtract(1, 'years').format();
                            qs.toTime = moment().format();
                        }
                        if (webinarKey !== undefined) {
                            const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/sessions`;
                            responseData = await GenericFunctions_1.goToWebinarApiRequestAllItems.call(this, 'GET', endpoint, qs, {}, resource);
                        }
                        else {
                            const endpoint = `organizers/${organizerKey}/sessions`;
                            responseData = await GenericFunctions_1.goToWebinarApiRequestAllItems.call(this, 'GET', endpoint, qs, {}, resource);
                        }
                    }
                    else if (operation === 'getDetails') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const sessionKey = this.getNodeParameter('sessionKey', i);
                        const details = this.getNodeParameter('details', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}/sessions/${sessionKey}/${details}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'GET', endpoint, {}, {});
                    }
                }
                else if (resource === 'webinar') {
                    if (operation === 'create') {
                        const timesProperties = this.getNodeParameter('times.timesProperties', i, []);
                        const body = {
                            subject: this.getNodeParameter('subject', i),
                            times: timesProperties,
                        };
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        Object.assign(body, additionalFields);
                        const endpoint = `organizers/${organizerKey}/webinars`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'POST', endpoint, {}, body);
                    }
                    else if (operation === 'delete') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const { sendCancellationEmails } = this.getNodeParameter('additionalFields', i);
                        const qs = {};
                        if (sendCancellationEmails) {
                            qs.sendCancellationEmails = sendCancellationEmails;
                        }
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}`;
                        await GenericFunctions_1.goToWebinarApiRequest.call(this, 'DELETE', endpoint, qs, {});
                        responseData = { success: true };
                    }
                    else if (operation === 'get') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequest.call(this, 'GET', endpoint, {}, {});
                    }
                    else if (operation === 'getAll') {
                        const qs = {};
                        const returnAll = this.getNodeParameter('returnAll', 0);
                        if (!returnAll) {
                            qs.limit = this.getNodeParameter('limit', 0);
                        }
                        const { times } = this.getNodeParameter('additionalFields', i);
                        if (times) {
                            qs.fromTime = moment(times.timesProperties.fromTime).format();
                            qs.toTime = moment(times.timesProperties.toTime).format();
                        }
                        else {
                            qs.fromTime = moment().subtract(1, 'years').format();
                            qs.toTime = moment().format();
                        }
                        const endpoint = `accounts/${accountKey}/webinars`;
                        responseData = await GenericFunctions_1.goToWebinarApiRequestAllItems.call(this, 'GET', endpoint, qs, {}, resource);
                    }
                    else if (operation === 'update') {
                        const webinarKey = this.getNodeParameter('webinarKey', i);
                        const qs = {
                            notifyParticipants: this.getNodeParameter('notifyParticipants', i),
                        };
                        let body = {};
                        let updateFields = this.getNodeParameter('updateFields', i);
                        if (updateFields.times) {
                            const { times } = updateFields;
                            body = {
                                times: times.timesProperties,
                            };
                            updateFields = lodash_1.omit(updateFields, ['times']);
                        }
                        Object.assign(body, updateFields);
                        if (lodash_1.isEmpty(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`);
                        }
                        const endpoint = `organizers/${organizerKey}/webinars/${webinarKey}`;
                        await GenericFunctions_1.goToWebinarApiRequest.call(this, 'PUT', endpoint, qs, body);
                        responseData = { success: true };
                    }
                }
            }
            catch (error) {
                if (this.continueOnFail()) {
                    returnData.push({ error: error.message });
                    continue;
                }
                throw error;
            }
            Array.isArray(responseData)
                ? returnData.push(...responseData)
                : returnData.push(responseData);
        }
        return [this.helpers.returnJsonArray(returnData)];
    }
}
exports.GoToWebinar = GoToWebinar;
