// n8n-workflow/dist/errors/node-api.error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeApiError = void 0;
const xml2js_1 = require("xml2js");
const node_error_1 = require("./abstract/node.error");
const utils_1 = require("../utils");
const axios_1 = require("axios");
const Constants_1 = require("../Constants");
const ERROR_MESSAGE_PROPERTIES = [
    'cause',
    'error',
    'message',
    'Message',
    'msg',
    'messages',
    'description',
    'reason',
    'detail',
    'details',
    'errors',
    'errorMessage',
    'errorMessages',
    'ErrorMessage',
    'error_message',
    '_error_message',
    'errorDescription',
    'error_description',
    'error_summary',
    'title',
    'text',
    'field',
    'err',
    'type',
];
const ERROR_STATUS_PROPERTIES = [
    'statusCode',
    'status',
    'code',
    'status_code',
    'errorCode',
    'error_code',
];
const ERROR_NESTING_PROPERTIES = ['error', 'err', 'response', 'body', 'data'];
const STATUS_CODE_MESSAGES = {
    '4XX': 'Your request is invalid or could not be processed by the service',
    '400': 'Bad request - please check your parameters',
    '401': 'Authorization failed - please check your credentials',
    '402': 'Payment required - perhaps check your payment details?',
    '403': 'Forbidden - perhaps check your credentials?',
    '404': 'The resource you are requesting could not be found',
    '405': 'Method not allowed - please check you are using the right HTTP method',
    '429': 'The service is receiving too many requests from you',
    '5XX': 'The service failed to process your request',
    '500': 'The service was not able to process your request',
    '502': 'Bad gateway - the service failed to handle your request',
    '503': 'Service unavailable - try again later or consider setting this node to retry automatically (in the node settings)',
    '504': 'Gateway timed out - perhaps try again later?',
};
class NodeApiError extends node_error_1.NodeError {
    constructor(node, errorResponse, { message, description, httpCode, parseXml, runIndex, itemIndex, level, functionality, messageMapping, } = {}) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        if (errorResponse instanceof NodeApiError) {
            return errorResponse;
        }
        super(node, errorResponse);
        this.httpCode = null;
        this.addToMessages(errorResponse.message);
        if (!httpCode && errorResponse instanceof axios_1.AxiosError) {
            httpCode = (_b = (_a = errorResponse.response) === null || _a === void 0 ? void 0 : _a.status) === null || _b === void 0 ? void 0 : _b.toString();
        }
        if (errorResponse.error) {
            (0, utils_1.removeCircularRefs)(errorResponse.error);
        }
        if (!description &&
            (errorResponse.description || ((_c = errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.reason) === null || _c === void 0 ? void 0 : _c.description))) {
            this.description = (errorResponse.description ||
                ((_d = errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.reason) === null || _d === void 0 ? void 0 : _d.description));
        }
        if (!message &&
            (errorResponse.message || ((_e = errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.reason) === null || _e === void 0 ? void 0 : _e.message) || description)) {
            this.message = (errorResponse.message ||
                ((_f = errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.reason) === null || _f === void 0 ? void 0 : _f.message) ||
                description);
        }
        if (errorResponse.reason) {
            const reason = errorResponse.reason;
            if (reason.isAxiosError && reason.response) {
                errorResponse = reason.response;
            }
        }
        if (httpCode) {
            this.httpCode = httpCode;
        }
        else if (errorResponse.httpCode) {
            this.httpCode = errorResponse.httpCode;
        }
        else {
            this.httpCode =
                (_g = this.findProperty(errorResponse, ERROR_STATUS_PROPERTIES, ERROR_NESTING_PROPERTIES)) !== null && _g !== void 0 ? _g : null;
        }
        this.level = level !== null && level !== void 0 ? level : 'warning';
        if ((errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.response) &&
            typeof (errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.response) === 'object' &&
            !Array.isArray(errorResponse.response) &&
            errorResponse.response.data &&
            typeof errorResponse.response.data === 'object' &&
            !Array.isArray(errorResponse.response.data)) {
            const data = errorResponse.response.data;
            if (data.message) {
                description = data.message;
            }
            else if (data.error && (data.error || {}).message) {
                description = data.error.message;
            }
            this.context.data = data;
        }
        if (description) {
            this.description = description;
        }
        if (!this.description) {
            if (parseXml) {
                this.setDescriptionFromXml(errorResponse.error);
            }
            else {
                this.description = this.findProperty(errorResponse, ERROR_MESSAGE_PROPERTIES, ERROR_NESTING_PROPERTIES);
            }
        }
        if (message) {
            this.message = message;
        }
        else {
            this.setDefaultStatusCodeMessage();
        }
        if (this.message === this.description) {
            this.description = undefined;
        }
        [this.message, this.messages] = this.setDescriptiveErrorMessage(this.message, this.messages, this.httpCode ||
            (errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.code) ||
            ((_h = errorResponse === null || errorResponse === void 0 ? void 0 : errorResponse.reason) === null || _h === void 0 ? void 0 : _h.code) ||
            undefined, messageMapping);
        if (functionality !== undefined)
            this.context.functionality = functionality;
        if (runIndex !== undefined)
            this.context.runIndex = runIndex;
        if (itemIndex !== undefined)
            this.context.itemIndex = itemIndex;
    }
    setDescriptionFromXml(xml) {
        (0, xml2js_1.parseString)(xml, { explicitArray: false }, (_, result) => {
            if (!result)
                return;
            const topLevelKey = Object.keys(result)[0];
            this.description = this.findProperty(result[topLevelKey], ERROR_MESSAGE_PROPERTIES, ['Error'].concat(ERROR_NESTING_PROPERTIES));
        });
    }
    setDefaultStatusCodeMessage() {
        if (!this.httpCode && this.message && this.message.toLowerCase().includes('bad gateway')) {
            this.httpCode = '502';
        }
        if (!this.httpCode) {
            this.httpCode = null;
            if (!this.message) {
                if (this.description) {
                    this.message = this.description;
                    this.description = undefined;
                }
                else {
                    this.message = Constants_1.UNKNOWN_ERROR_MESSAGE;
                    this.description = Constants_1.UNKNOWN_ERROR_DESCRIPTION;
                }
            }
            return;
        }
        if (STATUS_CODE_MESSAGES[this.httpCode]) {
            this.addToMessages(this.message);
            this.message = STATUS_CODE_MESSAGES[this.httpCode];
            return;
        }
        switch (this.httpCode.charAt(0)) {
            case '4':
                this.addToMessages(this.message);
                this.message = STATUS_CODE_MESSAGES['4XX'];
                break;
            case '5':
                this.addToMessages(this.message);
                this.message = STATUS_CODE_MESSAGES['5XX'];
                break;
            default:
                if (!this.message) {
                    if (this.description) {
                        this.message = this.description;
                        this.description = undefined;
                    }
                    else {
                        this.message = Constants_1.UNKNOWN_ERROR_MESSAGE;
                        this.description = Constants_1.UNKNOWN_ERROR_DESCRIPTION;
                    }
                }
        }
        if (this.node.type === Constants_1.NO_OP_NODE_TYPE && this.message === Constants_1.UNKNOWN_ERROR_MESSAGE) {
            this.message = `${Constants_1.UNKNOWN_ERROR_MESSAGE_CRED} - ${this.httpCode}`;
        }
    }
}
exports.NodeApiError = NodeApiError;
