// n8n-nodes-base/dist/nodes/SpreadsheetFile/SpreadsheetFile.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpreadsheetFile = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const SpreadsheetFileV1_node_1 = require("./v1/SpreadsheetFileV1.node");
const SpreadsheetFileV2_node_1 = require("./v2/SpreadsheetFileV2.node");
class SpreadsheetFile extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            hidden: true,
            displayName: 'Spreadsheet File',
            name: 'spreadsheetFile',
            icon: 'fa:table',
            group: ['transform'],
            description: 'Reads and writes data from a spreadsheet file like CSV, XLS, ODS, etc',
            defaultVersion: 2,
        };
        const nodeVersions = {
            1: new SpreadsheetFileV1_node_1.SpreadsheetFileV1(baseDescription),
            2: new SpreadsheetFileV2_node_1.SpreadsheetFileV2(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.SpreadsheetFile = SpreadsheetFile;
