// n8n-nodes-base/dist/nodes/Mattermost/v1/actions/message/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.descriptions = exports.postEphemeral = exports.post = exports.delete = void 0;
const del = require("./del");
exports.delete = del;
const post = require("./post");
exports.post = post;
const postEphemeral = require("./postEphemeral");
exports.postEphemeral = postEphemeral;
exports.descriptions = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'message',
                ],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'delete',
                description: 'Soft delete a post, by marking the post as deleted in the database',
            },
            {
                name: 'Post',
                value: 'post',
                description: 'Post a message into a channel',
            },
            {
                name: 'Post Ephemeral',
                value: 'postEphemeral',
                description: 'Post an ephemeral message into a channel',
            },
        ],
        default: 'post',
        description: 'The operation to perform',
    },
    ...del.description,
    ...post.description,
    ...postEphemeral.description,
];
