// n8n-nodes-base/dist/nodes/Discord/v2/actions/message/react.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const utilities_1 = require("../../../../../utils/utilities");
const utils_1 = require("../../helpers/utils");
const transport_1 = require("../../transport");
const common_description_1 = require("../common.description");
const properties = [
    common_description_1.channelRLC,
    common_description_1.messageIdString,
    {
        displayName: 'Emoji',
        name: 'emoji',
        type: 'string',
        default: '',
        required: true,
        description: 'The emoji you want to react with',
    },
];
const displayOptions = {
    show: {
        resource: ['message'],
        operation: ['react'],
    },
    hide: {
        authentication: ['webhook'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(_guildId, userGuilds) {
    const returnData = [];
    const items = this.getInputData();
    const getChannelId = await utils_1.setupChannelGetter.call(this, userGuilds);
    for (let i = 0; i < items.length; i++) {
        try {
            const channelId = await getChannelId(i);
            const messageId = this.getNodeParameter('messageId', i);
            const emoji = this.getNodeParameter('emoji', i);
            await transport_1.discordApiRequest.call(this, 'PUT', `/channels/${channelId}/messages/${messageId}/reactions/${encodeURIComponent(emoji)}/@me`);
            const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
            returnData.push(...executionData);
        }
        catch (error) {
            const err = utils_1.parseDiscordError.call(this, error, i);
            if (this.continueOnFail(error)) {
                returnData.push(...utils_1.prepareErrorData.call(this, err, i));
                continue;
            }
            throw err;
        }
    }
    return returnData;
}
exports.execute = execute;
