// n8n-nodes-base/dist/nodes/Bitwarden/descriptions/CollectionDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectionFields = exports.collectionOperations = void 0;
exports.collectionOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        default: 'get',
        description: 'Operation to perform',
        options: [
            {
                name: 'Delete',
                value: 'delete',
            },
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
            {
                name: 'Update',
                value: 'update',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
            },
        },
    },
];
exports.collectionFields = [
    {
        displayName: 'Collection ID',
        name: 'collectionId',
        type: 'string',
        required: true,
        description: 'The identifier of the collection.',
        default: '',
        placeholder: '5e59c8c7-e05a-4d17-8e85-acc301343926',
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'delete',
                    'get',
                    'update',
                ],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all available results for the query.',
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 10,
        description: 'Number of results to return for the query.',
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    {
        displayName: 'Update Fields',
        name: 'updateFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        required: true,
        options: [
            {
                displayName: 'Group',
                name: 'groups',
                type: 'multiOptions',
                description: 'The group to assign this collection to.',
                default: [],
                typeOptions: {
                    loadOptionsMethod: 'getGroups',
                },
            },
            {
                displayName: 'External ID',
                name: 'externalId',
                type: 'string',
                description: 'The external identifier to set to this collection.',
                default: '',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'update',
                ],
            },
        },
    },
];
