// n8n-nodes-base/dist/nodes/TheHiveProject/actions/log/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.search = exports.get = exports.executeResponder = exports.deleteLog = exports.deleteAttachment = exports.create = exports.addAttachment = void 0;
const addAttachment = __importStar(require("./addAttachment.operation"));
exports.addAttachment = addAttachment;
const create = __importStar(require("./create.operation"));
exports.create = create;
const deleteAttachment = __importStar(require("./deleteAttachment.operation"));
exports.deleteAttachment = deleteAttachment;
const deleteLog = __importStar(require("./deleteLog.operation"));
exports.deleteLog = deleteLog;
const executeResponder = __importStar(require("./executeResponder.operation"));
exports.executeResponder = executeResponder;
const get = __importStar(require("./get.operation"));
exports.get = get;
const search = __importStar(require("./search.operation"));
exports.search = search;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        noDataExpression: true,
        type: 'options',
        required: true,
        default: 'create',
        options: [
            {
                name: 'Add Attachment',
                value: 'addAttachment',
                action: 'Add attachment to a task log',
            },
            {
                name: 'Create',
                value: 'create',
                action: 'Create a task log',
            },
            {
                name: 'Delete',
                value: 'deleteLog',
                action: 'Delete task log',
            },
            {
                name: 'Delete Attachment',
                value: 'deleteAttachment',
                action: 'Delete attachment from a task log',
            },
            {
                name: 'Execute Responder',
                value: 'executeResponder',
                action: 'Execute responder on a task log',
            },
            {
                name: 'Get',
                value: 'get',
                action: 'Get a task log',
            },
            {
                name: 'Search',
                value: 'search',
                action: 'Search task logs',
            },
        ],
        displayOptions: {
            show: {
                resource: ['log'],
            },
        },
    },
    ...addAttachment.description,
    ...create.description,
    ...deleteAttachment.description,
    ...deleteLog.description,
    ...executeResponder.description,
    ...get.description,
    ...search.description,
];
