// n8n-nodes-base/dist/nodes/TheHiveProject/actions/log/search.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const descriptions_1 = require("../../descriptions");
const transport_1 = require("../../transport");
const utilities_1 = require("../../../../utils/utilities");
const properties = [
    {
        displayName: 'Search in All Tasks',
        name: 'allTasks',
        type: 'boolean',
        default: true,
        description: 'Whether to search in all tasks or only in selected task',
    },
    {
        ...descriptions_1.taskRLC,
        displayOptions: {
            show: {
                allTasks: [false],
            },
        },
    },
    ...descriptions_1.returnAllAndLimit,
    descriptions_1.genericFiltersCollection,
    descriptions_1.sortCollection,
    descriptions_1.searchOptions,
];
const displayOptions = {
    show: {
        resource: ['log'],
        operation: ['search'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const allTasks = this.getNodeParameter('allTasks', i);
    const filtersValues = this.getNodeParameter('filters.values', i, []);
    const sortFields = this.getNodeParameter('sort.fields', i, []);
    const returnAll = this.getNodeParameter('returnAll', i);
    const { returnCount, extraData } = this.getNodeParameter('options', i);
    let limit;
    let scope;
    if (allTasks) {
        scope = { query: 'listLog' };
    }
    else {
        const taskId = this.getNodeParameter('taskId', i, '', { extractValue: true });
        scope = { query: 'getTask', id: taskId, restrictTo: 'logs' };
    }
    if (!returnAll) {
        limit = this.getNodeParameter('limit', i);
    }
    responseData = await transport_1.theHiveApiQuery.call(this, scope, filtersValues, sortFields, limit, returnCount, extraData);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
