// n8n-nodes-base/dist/nodes/SentryIo/OrganizationDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.organizationFields = exports.organizationOperations = void 0;
exports.organizationOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create an organization',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Get organization by slug',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Get all organizations',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update an organization',
            },
        ],
        default: 'get',
        description: 'The operation to perform',
    },
];
exports.organizationFields = [
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                operation: [
                    'getAll',
                ],
                resource: [
                    'organization',
                ],
            },
        },
        default: false,
        description: 'If all results should be returned or only up to a given limit.',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                operation: [
                    'getAll',
                ],
                resource: [
                    'organization',
                ],
                returnAll: [
                    false,
                ],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 500,
        },
        default: 100,
        description: 'How many results to return.',
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        options: [
            {
                displayName: 'Member',
                name: 'member',
                type: 'boolean',
                default: true,
                description: 'Restrict results to organizations which you have membership.',
            },
            {
                displayName: 'Owner',
                name: 'owner',
                type: 'boolean',
                default: true,
                description: 'Restrict results to organizations which you are the owner.',
            },
        ],
    },
    {
        displayName: 'Organization Slug',
        name: 'organizationSlug',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getOrganizations',
        },
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'get',
                ],
            },
        },
        required: true,
        description: 'The slug of the organization the team should be created for.',
    },
    {
        displayName: 'Name',
        name: 'name',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'create',
                ],
            },
        },
        required: true,
        description: 'The slug of the organization the team should be created for.',
    },
    {
        displayName: 'Agree to Terms',
        name: 'agreeTerms',
        type: 'boolean',
        default: false,
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'create',
                ],
            },
        },
        description: 'Signaling you agree to the applicable terms of service and privacy policy of Sentry.io.',
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'create',
                ],
            },
        },
        options: [
            {
                displayName: 'Slug',
                name: 'slug',
                type: 'string',
                default: '',
                description: 'The unique URL slug for this organization. If this is not provided a slug is automatically generated based on the name.',
            },
        ],
    },
    {
        displayName: 'Slug',
        name: 'organization_slug',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getOrganizations',
        },
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'update',
                ],
            },
        },
        required: true,
        description: 'The slug of the organization to update.',
    },
    {
        displayName: 'Update Fields',
        name: 'updateFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'organization',
                ],
                operation: [
                    'update',
                ],
            },
        },
        options: [
            {
                displayName: 'Name',
                name: 'name',
                type: 'string',
                default: '',
                description: 'The new name of the organization.',
            },
            {
                displayName: 'Slug',
                name: 'slug',
                type: 'string',
                default: '',
                description: 'The new URL slug for this organization.',
            },
        ],
    },
];
