// n8n-nodes-base/dist/nodes/Discord/v2/actions/member/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.roleRemove = exports.roleAdd = exports.getAll = void 0;
const common_description_1 = require("../common.description");
const getAll = __importStar(require("./getAll.operation"));
exports.getAll = getAll;
const roleAdd = __importStar(require("./roleAdd.operation"));
exports.roleAdd = roleAdd;
const roleRemove = __importStar(require("./roleRemove.operation"));
exports.roleRemove = roleRemove;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['member'],
                authentication: ['botToken', 'oAuth2'],
            },
        },
        options: [
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Retrieve the members of a server',
                action: 'Get many members',
            },
            {
                name: 'Role Add',
                value: 'roleAdd',
                description: 'Add a role to a member',
                action: 'Add a role to a member',
            },
            {
                name: 'Role Remove',
                value: 'roleRemove',
                description: 'Remove a role from a member',
                action: 'Remove a role from a member',
            },
        ],
        default: 'getAll',
    },
    {
        ...common_description_1.guildRLC,
        displayOptions: {
            show: {
                resource: ['member'],
                authentication: ['botToken', 'oAuth2'],
            },
        },
    },
    ...getAll.description,
    ...roleAdd.description,
    ...roleRemove.description,
];
