// n8n/dist/services/ai/utils/summarizeNodeTypeProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.summarizeNodeTypeProperties = exports.summarizeProperty = exports.summarizeOption = void 0;
function summarizeOption(option) {
    if ('value' in option) {
        return {
            name: option.name,
            value: option.value,
        };
    }
    else if ('values' in option) {
        return {
            name: option.name,
            values: option.values.map(summarizeProperty),
        };
    }
    else {
        return summarizeProperty(option);
    }
}
exports.summarizeOption = summarizeOption;
function summarizeProperty(property) {
    return {
        name: property.displayName,
        type: property.type,
        ...(property.displayOptions ? { displayOptions: property.displayOptions } : {}),
        ...(property.options
            ? { options: property.options.map(summarizeOption) }
            : {}),
    };
}
exports.summarizeProperty = summarizeProperty;
function summarizeNodeTypeProperties(nodeTypeProperties) {
    return nodeTypeProperties.map(summarizeProperty);
}
exports.summarizeNodeTypeProperties = summarizeNodeTypeProperties;
