// n8n-nodes-base/dist/nodes/WooCommerce/descriptions/CustomerDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customerFields = exports.customerOperations = void 0;
const shared_1 = require("./shared");
exports.customerOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create a customer',
            },
            {
                name: 'Delete',
                value: 'delete',
                description: 'Delete a customer',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Retrieve a customer',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Retrieve all customers',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update a customer',
            },
        ],
        default: 'create',
    },
];
exports.customerFields = [
    {
        displayName: 'Email',
        name: 'email',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    shared_1.customerCreateFields,
    {
        displayName: 'Customer ID',
        name: 'customerId',
        description: 'ID of the customer to delete',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'delete',
                ],
            },
        },
    },
    {
        displayName: 'Customer ID',
        name: 'customerId',
        description: 'ID of the customer to retrieve',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Whether to return all results or only up to a given limit',
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 50,
        description: 'How many results to return',
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        options: [
            {
                displayName: 'Email',
                name: 'email',
                type: 'string',
                default: '',
                description: 'Email address to filter customers by',
            },
            {
                displayName: 'Sort Order',
                name: 'order',
                description: 'Order to sort customers in',
                type: 'options',
                options: [
                    {
                        name: 'Ascending',
                        value: 'asc',
                    },
                    {
                        name: 'Descending',
                        value: 'desc',
                    },
                ],
                default: 'asc',
            },
            {
                displayName: 'Order By',
                name: 'orderby',
                description: 'Field to sort customers by',
                type: 'options',
                options: [
                    {
                        name: 'ID',
                        value: 'id',
                    },
                    {
                        name: 'Include',
                        value: 'include',
                    },
                    {
                        name: 'Name',
                        value: 'name',
                    },
                    {
                        name: 'Registered Date',
                        value: 'registered_date',
                    },
                ],
                default: 'id',
            },
        ],
    },
    {
        displayName: 'Customer ID',
        name: 'customerId',
        description: 'ID of the customer to update',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'customer',
                ],
                operation: [
                    'update',
                ],
            },
        },
    },
    shared_1.customerUpdateFields,
];
