// n8n-nodes-base/dist/nodes/Microsoft/Outlook/v2/actions/folder/get.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = exports.properties = void 0;
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utils_1 = require("../../helpers/utils");
const utilities_1 = require("../../../../../../utils/utilities");
exports.properties = [
    descriptions_1.folderRLC,
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        options: [
            {
                displayName: 'Fields',
                name: 'fields',
                type: 'multiOptions',
                description: 'The fields to add to the output',
                options: descriptions_1.folderFields,
                default: [],
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['folder'],
        operation: ['get'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
async function execute(index) {
    const qs = {};
    const folderId = (0, utils_1.decodeOutlookId)(this.getNodeParameter('folderId', index, undefined, {
        extractValue: true,
    }));
    const options = this.getNodeParameter('options', index);
    if (options.fields) {
        qs.$select = options.fields.join(',');
    }
    if (options.filter) {
        qs.$filter = options.filter;
    }
    const responseData = await transport_1.microsoftApiRequest.call(this, 'GET', `/mailFolders/${folderId}`, {}, qs);
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: index } });
    return executionData;
}
exports.execute = execute;
