// n8n-nodes-base/dist/nodes/Files/ReadWriteFile/actions/read.operation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = exports.properties = void 0;
const fast_glob_1 = __importDefault(require("fast-glob"));
const utils_1 = require("../helpers/utils");
const utilities_1 = require("../../../../utils/utilities");
exports.properties = [
    {
        displayName: 'File(s) Selector',
        name: 'fileSelector',
        type: 'string',
        default: '',
        required: true,
        placeholder: 'e.g. /home/user/Pictures/**/*.png',
        hint: 'Supports patterns, learn more <a href="https://github.com/micromatch/picomatch#basic-globbing" target="_blank">here</a>',
        description: "Specify a file's path or path pattern to read multiple files",
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        options: [
            {
                displayName: 'File Extension',
                name: 'fileExtension',
                type: 'string',
                default: '',
                placeholder: 'e.g. zip',
                description: 'Extension of the file in the output binary',
            },
            {
                displayName: 'File Name',
                name: 'fileName',
                type: 'string',
                default: '',
                placeholder: 'e.g. data.zip',
                description: 'Name of the file in the output binary',
            },
            {
                displayName: 'Mime Type',
                name: 'mimeType',
                type: 'string',
                default: '',
                placeholder: 'e.g. application/zip',
                description: 'Mime type of the file in the output binary',
            },
            {
                displayName: 'Put Output File in Field',
                name: 'dataPropertyName',
                type: 'string',
                default: 'data',
                placeholder: 'e.g. data',
                description: "By default 'data' is used",
                hint: 'The name of the output binary field to put the file in',
            },
        ],
    },
];
const displayOptions = {
    show: {
        operation: ['read'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
async function execute(items) {
    const returnData = [];
    let fileSelector;
    for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
        try {
            fileSelector = this.getNodeParameter('fileSelector', itemIndex);
            const options = this.getNodeParameter('options', itemIndex, {});
            let dataPropertyName = 'data';
            if (options.dataPropertyName) {
                dataPropertyName = options.dataPropertyName;
            }
            const files = await (0, fast_glob_1.default)(fileSelector);
            const newItems = [];
            for (const filePath of files) {
                const stream = await this.helpers.createReadStream(filePath);
                const binaryData = await this.helpers.prepareBinaryData(stream, filePath);
                if (options.fileName !== undefined) {
                    binaryData.fileName = options.fileName;
                }
                if (options.fileExtension !== undefined) {
                    binaryData.fileExtension = options.fileExtension;
                }
                if (options.mimeType !== undefined) {
                    binaryData.mimeType = options.mimeType;
                }
                newItems.push({
                    binary: {
                        [dataPropertyName]: binaryData,
                    },
                    json: {
                        mimeType: binaryData.mimeType,
                        fileType: binaryData.fileType,
                        fileName: binaryData.fileName,
                        directory: binaryData.directory,
                        fileExtension: binaryData.fileExtension,
                        fileSize: binaryData.fileSize,
                    },
                    pairedItem: {
                        item: itemIndex,
                    },
                });
            }
            returnData.push(...newItems);
        }
        catch (error) {
            const nodeOperatioinError = utils_1.errorMapper.call(this, error, itemIndex, {
                filePath: fileSelector,
                operation: 'read',
            });
            if (this.continueOnFail(error)) {
                returnData.push({
                    json: {
                        error: nodeOperatioinError.message,
                    },
                    pairedItem: {
                        item: itemIndex,
                    },
                });
                continue;
            }
            throw nodeOperatioinError;
        }
    }
    return returnData;
}
exports.execute = execute;
