// @n8n/n8n-nodes-langchain/dist/nodes/vector_store/VectorStoreInMemoryInsert/VectorStoreInMemoryInsert.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorStoreInMemoryInsert = void 0;
const processDocuments_1 = require("../shared/processDocuments");
const MemoryVectorStoreManager_1 = require("../shared/MemoryVectorStoreManager");
class VectorStoreInMemoryInsert {
    constructor() {
        this.description = {
            displayName: 'In Memory Vector Store Insert',
            name: 'vectorStoreInMemoryInsert',
            icon: 'fa:database',
            group: ['transform'],
            version: 1,
            hidden: true,
            description: 'Insert data into an in-memory vector store',
            defaults: {
                name: 'In Memory Vector Store Insert',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Vector Stores'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/root-nodes/n8n-nodes-langchain.vectorstoreinmemory/',
                        },
                    ],
                },
            },
            inputs: [
                "main",
                {
                    displayName: 'Document',
                    maxConnections: 1,
                    type: "ai_document",
                    required: true,
                },
                {
                    displayName: 'Embedding',
                    maxConnections: 1,
                    type: "ai_embedding",
                    required: true,
                },
            ],
            outputs: ["main"],
            properties: [
                {
                    displayName: 'The embbded data are stored in the server memory, so they will be lost when the server is restarted. Additionally, if the amount of data is too large, it may cause the server to crash due to insufficient memory.',
                    name: 'notice',
                    type: 'notice',
                    default: '',
                },
                {
                    displayName: 'Clear Store',
                    name: 'clearStore',
                    type: 'boolean',
                    default: false,
                    description: 'Whether to clear the store before inserting new data',
                },
                {
                    displayName: 'Memory Key',
                    name: 'memoryKey',
                    type: 'string',
                    default: 'vector_store_key',
                    description: 'The key to use to store the vector memory in the workflow data. The key will be prefixed with the workflow ID to avoid collisions.',
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData(0);
        const embeddings = (await this.getInputConnectionData("ai_embedding", 0));
        const memoryKey = this.getNodeParameter('memoryKey', 0);
        const clearStore = this.getNodeParameter('clearStore', 0);
        const documentInput = (await this.getInputConnectionData("ai_document", 0));
        const { processedDocuments, serializedDocuments } = await (0, processDocuments_1.processDocuments)(documentInput, items);
        const workflowId = this.getWorkflow().id;
        const vectorStoreInstance = MemoryVectorStoreManager_1.MemoryVectorStoreManager.getInstance(embeddings);
        await vectorStoreInstance.addDocuments(`${workflowId}__${memoryKey}`, processedDocuments, clearStore);
        return [serializedDocuments];
    }
}
exports.VectorStoreInMemoryInsert = VectorStoreInMemoryInsert;
