// n8n-nodes-base/dist/nodes/Wise/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testPublicKey = exports.livePublicKey = exports.getTriggerName = exports.handleBinaryData = exports.wiseApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function wiseApiRequest(method, endpoint, body = {}, qs = {}, option = {}) {
    const { apiToken, environment } = await this.getCredentials('wiseApi');
    const rootUrl = environment === 'live'
        ? 'https://api.transferwise.com/'
        : 'https://api.sandbox.transferwise.tech/';
    const options = {
        headers: {
            'user-agent': 'n8n',
            'Authorization': `Bearer ${apiToken}`,
        },
        method,
        uri: `${rootUrl}${endpoint}`,
        qs,
        body,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    if (Object.keys(option)) {
        Object.assign(options, option);
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.wiseApiRequest = wiseApiRequest;
async function handleBinaryData(items, i, endpoint) {
    var _a;
    const data = await wiseApiRequest.call(this, 'GET', endpoint, {}, {}, { encoding: null });
    const binaryProperty = this.getNodeParameter('binaryProperty', i);
    items[i].binary = (_a = items[i].binary) !== null && _a !== void 0 ? _a : {};
    items[i].binary[binaryProperty] = await this.helpers.prepareBinaryData(data);
    items[i].binary[binaryProperty].fileName = this.getNodeParameter('fileName', i);
    items[i].binary[binaryProperty].fileExtension = 'pdf';
    return items;
}
exports.handleBinaryData = handleBinaryData;
function getTriggerName(eventName) {
    const events = {
        'tranferStateChange': 'transfers#state-change',
        'transferActiveCases': 'transfers#active-cases',
        'balanceCredit': 'balances#credit',
    };
    return events[eventName];
}
exports.getTriggerName = getTriggerName;
exports.livePublicKey = `
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvO8vXV+JksBzZAY6GhSO
XdoTCfhXaaiZ+qAbtaDBiu2AGkGVpmEygFmWP4Li9m5+Ni85BhVvZOodM9epgW3F
bA5Q1SexvAF1PPjX4JpMstak/QhAgl1qMSqEevL8cmUeTgcMuVWCJmlge9h7B1CS
D4rtlimGZozG39rUBDg6Qt2K+P4wBfLblL0k4C4YUdLnpGYEDIth+i8XsRpFlogx
CAFyH9+knYsDbR43UJ9shtc42Ybd40Afihj8KnYKXzchyQ42aC8aZ/h5hyZ28yVy
Oj3Vos0VdBIs/gAyJ/4yyQFCXYte64I7ssrlbGRaco4nKF3HmaNhxwyKyJafz19e
HwIDAQAB
-----END PUBLIC KEY-----`;
exports.testPublicKey = `
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAwpb91cEYuyJNQepZAVfP
ZIlPZfNUefH+n6w9SW3fykqKu938cR7WadQv87oF2VuT+fDt7kqeRziTmPSUhqPU
ys/V2Q1rlfJuXbE+Gga37t7zwd0egQ+KyOEHQOpcTwKmtZ81ieGHynAQzsn1We3j
wt760MsCPJ7GMT141ByQM+yW1Bx+4SG3IGjXWyqOWrcXsxAvIXkpUD/jK/L958Cg
nZEgz0BSEh0QxYLITnW1lLokSx/dTianWPFEhMC9BgijempgNXHNfcVirg1lPSyg
z7KqoKUN0oHqWLr2U1A+7kqrl6O2nx3CKs1bj1hToT1+p4kcMoHXA7kA+VBLUpEs
VwIDAQAB
-----END PUBLIC KEY-----`;
