// n8n-nodes-base/dist/utils/binary.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractDataFromPDF = exports.createBinaryFromJson = exports.convertJsonToSpreadsheetBinary = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const xlsx_1 = require("xlsx");
const get_1 = __importDefault(require("lodash/get"));
const iconv_lite_1 = __importDefault(require("iconv-lite"));
const pdfjs_dist_1 = require("pdfjs-dist");
const utilities_1 = require("./utilities");
async function convertJsonToSpreadsheetBinary(items, fileFormat, options, defaultFileName = 'spreadsheet') {
    const itemData = [];
    for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
        itemData.push((0, utilities_1.flattenObject)(items[itemIndex].json));
    }
    let sheetToJsonOptions;
    if (options.headerRow === false) {
        sheetToJsonOptions = { skipHeader: true };
    }
    const sheet = xlsx_1.utils.json_to_sheet(itemData, sheetToJsonOptions);
    const writingOptions = {
        bookType: fileFormat,
        bookSST: false,
        type: 'buffer',
    };
    if (['xlsx', 'ods'].includes(fileFormat) && options.compression) {
        writingOptions.compression = true;
    }
    const sheetName = options.sheetName || 'Sheet';
    const workbook = {
        SheetNames: [sheetName],
        Sheets: {
            [sheetName]: sheet,
        },
    };
    const buffer = (0, xlsx_1.write)(workbook, writingOptions);
    const fileName = options.fileName !== undefined ? options.fileName : `${defaultFileName}.${fileFormat}`;
    const binaryData = await this.helpers.prepareBinaryData(buffer, fileName);
    return binaryData;
}
exports.convertJsonToSpreadsheetBinary = convertJsonToSpreadsheetBinary;
async function createBinaryFromJson(data, options) {
    let value;
    if (options.sourceKey) {
        value = (0, get_1.default)(data, options.sourceKey);
    }
    else {
        value = data;
    }
    if (value === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `The value in "${options.sourceKey}" is not set`, {
            itemIndex: options.itemIndex || 0,
        });
    }
    let buffer;
    if (!options.dataIsBase64) {
        let valueAsString = value;
        if (typeof value === 'object') {
            options.mimeType = 'application/json';
            if (options.format) {
                valueAsString = JSON.stringify(value, null, 2);
            }
            else {
                valueAsString = JSON.stringify(value);
            }
        }
        buffer = iconv_lite_1.default.encode(valueAsString, options.encoding || 'utf8', {
            addBOM: options.addBOM,
        });
    }
    else {
        buffer = Buffer.from(value, n8n_workflow_1.BINARY_ENCODING);
    }
    const binaryData = await this.helpers.prepareBinaryData(buffer, options.fileName, options.mimeType);
    if (!binaryData.fileName) {
        const fileExtension = binaryData.fileExtension ? `.${binaryData.fileExtension}` : '';
        binaryData.fileName = `file${fileExtension}`;
    }
    return binaryData;
}
exports.createBinaryFromJson = createBinaryFromJson;
const parseText = (textContent) => {
    let lastY = undefined;
    const text = [];
    for (const item of textContent.items) {
        if ('str' in item) {
            if (lastY == item.transform[5] || !lastY) {
                text.push(item.str);
            }
            else {
                text.push(`\n${item.str}`);
            }
            lastY = item.transform[5];
        }
    }
    return text.join('');
};
async function extractDataFromPDF(binaryPropertyName, password, maxPages, joinPages = true, itemIndex = 0) {
    const binaryData = this.helpers.assertBinaryData(itemIndex, binaryPropertyName);
    const params = { password, isEvalSupported: false };
    if (binaryData.id) {
        params.data = await this.helpers.binaryToBuffer(await this.helpers.getBinaryStream(binaryData.id));
    }
    else {
        params.data = Buffer.from(binaryData.data, n8n_workflow_1.BINARY_ENCODING).buffer;
    }
    const document = await (0, pdfjs_dist_1.getDocument)(params).promise;
    const { info, metadata } = await document
        .getMetadata()
        .catch(() => ({ info: null, metadata: null }));
    const pages = [];
    if (maxPages !== 0) {
        let pagesToRead = document.numPages;
        if (maxPages && maxPages < document.numPages) {
            pagesToRead = maxPages;
        }
        for (let i = 1; i <= pagesToRead; i++) {
            const page = await document.getPage(i);
            const text = await page.getTextContent().then(parseText);
            pages.push(text);
        }
    }
    const text = joinPages ? pages.join('\n\n') : pages;
    const returnData = {
        numpages: document.numPages,
        numrender: document.numPages,
        info,
        metadata: metadata === null || metadata === void 0 ? void 0 : metadata.getAll(),
        text,
        version: pdfjs_dist_1.version,
    };
    return returnData;
}
exports.extractDataFromPDF = extractDataFromPDF;
