// n8n-nodes-base/dist/nodes/Google/Drive/v2/actions/folder/deleteFolder.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
const common_descriptions_1 = require("../common.descriptions");
const utilities_1 = require("../../../../../../utils/utilities");
const properties = [
    {
        ...common_descriptions_1.folderNoRootRLC,
        description: 'The folder to delete',
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        options: [
            {
                displayName: 'Delete Permanently',
                name: 'deletePermanently',
                type: 'boolean',
                default: false,
                description: 'Whether to delete the folder immediately. If false, the folder will be moved to the trash.',
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['folder'],
        operation: ['deleteFolder'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    const returnData = [];
    const folderId = this.getNodeParameter('folderNoRootId', i, undefined, {
        extractValue: true,
    });
    const deletePermanently = this.getNodeParameter('options.deletePermanently', i, false);
    const qs = {
        supportsAllDrives: true,
    };
    if (deletePermanently) {
        await transport_1.googleApiRequest.call(this, 'DELETE', `/drive/v3/files/${folderId}`, undefined, qs);
    }
    else {
        await transport_1.googleApiRequest.call(this, 'PATCH', `/drive/v3/files/${folderId}`, { trashed: true }, qs);
    }
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({
        fileId: folderId,
        success: true,
    }), { itemData: { item: i } });
    returnData.push(...executionData);
    return returnData;
}
exports.execute = execute;
