// n8n-nodes-base/dist/nodes/Google/Drive/v2/actions/folder/Folder.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.share = exports.deleteFolder = exports.create = void 0;
const create = __importStar(require("./create.operation"));
exports.create = create;
const deleteFolder = __importStar(require("./deleteFolder.operation"));
exports.deleteFolder = deleteFolder;
const share = __importStar(require("./share.operation"));
exports.share = share;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['folder'],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create a folder',
                action: 'Create folder',
            },
            {
                name: 'Delete',
                value: 'deleteFolder',
                description: 'Permanently delete a folder',
                action: 'Delete folder',
            },
            {
                name: 'Share',
                value: 'share',
                description: 'Add sharing permissions to a folder',
                action: 'Share folder',
            },
        ],
        default: 'create',
    },
    ...create.description,
    ...deleteFolder.description,
    ...share.description,
];
