// n8n-nodes-base/dist/nodes/Mattermost/v1/actions/user/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.descriptions = exports.invite = exports.getById = exports.getByEmail = exports.getAll = exports.deactive = exports.create = void 0;
const create = require("./create");
exports.create = create;
const deactive = require("./deactive");
exports.deactive = deactive;
const getAll = require("./getAll");
exports.getAll = getAll;
const getByEmail = require("./getByEmail");
exports.getByEmail = getByEmail;
const getById = require("./getById");
exports.getById = getById;
const invite = require("./invite");
exports.invite = invite;
exports.descriptions = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'user',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create a new user',
            },
            {
                name: 'Deactive',
                value: 'deactive',
                description: 'Deactivates the user and revokes all its sessions by archiving its user object.',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Retrieve all users',
            },
            {
                name: 'Get By Email',
                value: 'getByEmail',
                description: 'Get a user by email',
            },
            {
                name: 'Get By ID',
                value: 'getById',
                description: 'Get a user by id',
            },
            {
                name: 'Invite',
                value: 'invite',
                description: 'Invite user to team',
            },
        ],
        default: '',
        description: 'The operation to perform.',
    },
    ...create.description,
    ...deactive.description,
    ...getAll.description,
    ...getByEmail.description,
    ...getById.description,
    ...invite.description,
];
