// n8n-nodes-base/dist/nodes/Google/Drive/v2/actions/file/File.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.upload = exports.update = exports.share = exports.move = exports.download = exports.deleteFile = exports.createFromText = exports.copy = void 0;
const copy = __importStar(require("./copy.operation"));
exports.copy = copy;
const createFromText = __importStar(require("./createFromText.operation"));
exports.createFromText = createFromText;
const deleteFile = __importStar(require("./deleteFile.operation"));
exports.deleteFile = deleteFile;
const download = __importStar(require("./download.operation"));
exports.download = download;
const move = __importStar(require("./move.operation"));
exports.move = move;
const share = __importStar(require("./share.operation"));
exports.share = share;
const update = __importStar(require("./update.operation"));
exports.update = update;
const upload = __importStar(require("./upload.operation"));
exports.upload = upload;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['file'],
            },
        },
        options: [
            {
                name: 'Copy',
                value: 'copy',
                description: 'Create a copy of an existing file',
                action: 'Copy file',
            },
            {
                name: 'Create From Text',
                value: 'createFromText',
                description: 'Create a file from a provided text',
                action: 'Create file from text',
            },
            {
                name: 'Delete',
                value: 'deleteFile',
                description: 'Permanently delete a file',
                action: 'Delete a file',
            },
            {
                name: 'Download',
                value: 'download',
                description: 'Download a file',
                action: 'Download file',
            },
            {
                name: 'Move',
                value: 'move',
                description: 'Move a file to another folder',
                action: 'Move file',
            },
            {
                name: 'Share',
                value: 'share',
                description: 'Add sharing permissions to a file',
                action: 'Share file',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update a file',
                action: 'Update file',
            },
            {
                name: 'Upload',
                value: 'upload',
                description: 'Upload an existing file to Google Drive',
                action: 'Upload file',
            },
        ],
        default: 'upload',
    },
    ...copy.description,
    ...deleteFile.description,
    ...createFromText.description,
    ...download.description,
    ...move.description,
    ...share.description,
    ...update.description,
    ...upload.description,
];
