// n8n-nodes-base/dist/nodes/Google/Drive/v2/actions/file/share.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
const common_descriptions_1 = require("../common.descriptions");
const utilities_1 = require("../../../../../../utils/utilities");
const properties = [
    {
        ...common_descriptions_1.fileRLC,
        description: 'The file to share',
    },
    common_descriptions_1.permissionsOptions,
    common_descriptions_1.shareOptions,
];
const displayOptions = {
    show: {
        resource: ['file'],
        operation: ['share'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    const returnData = [];
    const fileId = this.getNodeParameter('fileId', i, undefined, {
        extractValue: true,
    });
    const permissions = this.getNodeParameter('permissionsUi', i);
    const shareOption = this.getNodeParameter('options', i);
    const body = {};
    const qs = {
        supportsAllDrives: true,
    };
    if (permissions.permissionsValues) {
        Object.assign(body, permissions.permissionsValues);
    }
    Object.assign(qs, shareOption);
    const response = await transport_1.googleApiRequest.call(this, 'POST', `/drive/v3/files/${fileId}/permissions`, body, qs);
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(response), { itemData: { item: i } });
    returnData.push(...executionData);
    return returnData;
}
exports.execute = execute;
