// n8n-nodes-base/dist/nodes/Google/Drive/v2/actions/file/upload.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
const common_descriptions_1 = require("../common.descriptions");
const utils_1 = require("../../helpers/utils");
const utilities_1 = require("../../../../../../utils/utilities");
const properties = [
    {
        displayName: 'Input Data Field Name',
        name: 'inputDataFieldName',
        type: 'string',
        placeholder: '“e.g. data',
        default: 'data',
        required: true,
        hint: 'The name of the input field containing the binary file data to update the file',
        description: 'Find the name of input field containing the binary data to update the file in the Input panel on the left, in the Binary tab',
    },
    {
        displayName: 'File Name',
        name: 'name',
        type: 'string',
        default: '',
        placeholder: 'e.g. My New File',
        description: 'If not specified, the original file name will be used',
    },
    {
        ...common_descriptions_1.driveRLC,
        displayName: 'Parent Drive',
        description: 'The drive where to upload the file',
    },
    {
        ...common_descriptions_1.folderRLC,
        displayName: 'Parent Folder',
        description: 'The folder where to upload the file',
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        options: [
            ...common_descriptions_1.updateCommonOptions,
            {
                displayName: 'Simplify Output',
                name: 'simplifyOutput',
                type: 'boolean',
                default: true,
                description: 'Whether to return a simplified version of the response instead of all fields',
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['file'],
        operation: ['upload'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    var _a;
    const returnData = [];
    const inputDataFieldName = this.getNodeParameter('inputDataFieldName', i);
    const { contentLength, fileContent, originalFilename, mimeType } = await utils_1.getItemBinaryData.call(this, inputDataFieldName, i);
    const name = this.getNodeParameter('name', i) || originalFilename;
    const driveId = this.getNodeParameter('driveId', i, undefined, {
        extractValue: true,
    });
    const folderId = this.getNodeParameter('folderId', i, undefined, {
        extractValue: true,
    });
    let uploadId;
    if (Buffer.isBuffer(fileContent)) {
        const response = await transport_1.googleApiRequest.call(this, 'POST', '/upload/drive/v3/files', fileContent, {
            uploadType: 'media',
        }, undefined, {
            headers: {
                'Content-Type': mimeType,
                'Content-Length': contentLength,
            },
        });
        uploadId = response.id;
    }
    else {
        const resumableUpload = await transport_1.googleApiRequest.call(this, 'POST', '/upload/drive/v3/files', undefined, { uploadType: 'resumable' }, undefined, {
            returnFullResponse: true,
        });
        const uploadUrl = resumableUpload.headers.location;
        let offset = 0;
        for await (const chunk of fileContent) {
            const nextOffset = offset + Number(chunk.length);
            try {
                const response = await this.helpers.httpRequest({
                    method: 'PUT',
                    url: uploadUrl,
                    headers: {
                        'Content-Length': chunk.length,
                        'Content-Range': `bytes ${offset}-${nextOffset - 1}/${contentLength}`,
                    },
                    body: chunk,
                });
                uploadId = response === null || response === void 0 ? void 0 : response.id;
            }
            catch (error) {
                if (((_a = error.response) === null || _a === void 0 ? void 0 : _a.status) !== 308)
                    throw error;
            }
            offset = nextOffset;
        }
    }
    const options = this.getNodeParameter('options', i, {});
    const qs = (0, utils_1.setUpdateCommonParams)({
        addParents: (0, utils_1.setParentFolder)(folderId, driveId),
        includeItemsFromAllDrives: true,
        supportsAllDrives: true,
        spaces: 'appDataFolder, drive',
        corpora: 'allDrives',
    }, options);
    if (!options.simplifyOutput) {
        qs.fields = '*';
    }
    const body = (0, utils_1.setFileProperties)({
        mimeType,
        name,
        originalFilename,
    }, options);
    const response = await transport_1.googleApiRequest.call(this, 'PATCH', `/drive/v3/files/${uploadId}`, body, qs);
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(response), { itemData: { item: i } });
    returnData.push(...executionData);
    return returnData;
}
exports.execute = execute;
