// n8n-nodes-base/dist/nodes/Postgres/v2/actions/database/executeQuery.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const utils_1 = require("../../helpers/utils");
const common_descriptions_1 = require("../common.descriptions");
const utilities_1 = require("../../../../../utils/utilities");
const properties = [
    {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: '',
        placeholder: 'e.g. SELECT id, name FROM product WHERE quantity > $1 AND price <= $2',
        noDataExpression: true,
        required: true,
        description: "The SQL query to execute. You can use n8n expressions and $1, $2, $3, etc to refer to the 'Query Parameters' set in options below.",
        typeOptions: {
            editor: 'sqlEditor',
            sqlDialect: 'PostgreSQL',
        },
        hint: 'Consider using query parameters to prevent SQL injection attacks. Add them in the options below',
    },
    common_descriptions_1.optionsCollection,
];
const displayOptions = {
    show: {
        resource: ['database'],
        operation: ['executeQuery'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(runQueries, items, nodeOptions, _db) {
    var _a;
    items = (0, utils_1.replaceEmptyStringsByNulls)(items, nodeOptions.replaceEmptyStrings);
    const queries = [];
    for (let i = 0; i < items.length; i++) {
        let query = this.getNodeParameter('query', i);
        for (const resolvable of (0, utilities_1.getResolvables)(query)) {
            query = query.replace(resolvable, this.evaluateExpression(resolvable, i));
        }
        let values = [];
        let queryReplacement = this.getNodeParameter('options.queryReplacement', i, '');
        if (typeof queryReplacement === 'number') {
            queryReplacement = String(queryReplacement);
        }
        if (typeof queryReplacement === 'string') {
            const node = this.getNode();
            const rawReplacements = (_a = node.parameters.options) === null || _a === void 0 ? void 0 : _a.queryReplacement;
            if (rawReplacements) {
                const rawValues = rawReplacements
                    .replace(/^=+/, '')
                    .split(',')
                    .filter((entry) => entry)
                    .map((entry) => entry.trim());
                for (const rawValue of rawValues) {
                    const resolvables = (0, utilities_1.getResolvables)(rawValue);
                    if (resolvables.length) {
                        for (const resolvable of resolvables) {
                            values.push(this.evaluateExpression(`${resolvable}`, i));
                        }
                    }
                    else {
                        values.push(rawValue);
                    }
                }
            }
        }
        else {
            if (Array.isArray(queryReplacement)) {
                values = queryReplacement;
            }
            else {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Query Parameters must be a string of comma-separated values or an array of values', { itemIndex: i });
            }
        }
        queries.push({ query, values });
    }
    return await runQueries(queries, items, nodeOptions);
}
exports.execute = execute;
