// n8n-nodes-base/dist/credentials/AirtableApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirtableApi = void 0;
class AirtableApi {
    constructor() {
        this.name = 'airtableApi';
        this.displayName = 'Airtable API';
        this.documentationUrl = 'airtable';
        this.properties = [
            {
                displayName: "This type of connection (API Key) was deprecated and can't be used anymore. Please create a new credential of type 'Access Token' instead.",
                name: 'deprecated',
                type: 'notice',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                qs: {
                    api_key: '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.AirtableApi = AirtableApi;
