// n8n-nodes-base/dist/credentials/AirtableOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirtableOAuth2Api = void 0;
const scopes = ['schema.bases:read', 'data.records:read', 'data.records:write'];
class AirtableOAuth2Api {
    constructor() {
        this.name = 'airtableOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'Airtable OAuth2 API';
        this.documentationUrl = 'airtable';
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'pkce',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://airtable.com/oauth2/v1/authorize',
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://airtable.com/oauth2/v1/token',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: `${scopes.join(' ')}`,
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'header',
            },
        ];
    }
}
exports.AirtableOAuth2Api = AirtableOAuth2Api;
