// n8n-nodes-base/dist/credentials/AirtableTokenApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirtableTokenApi = void 0;
class AirtableTokenApi {
    constructor() {
        this.name = 'airtableTokenApi';
        this.displayName = 'Airtable Personal Access Token API';
        this.documentationUrl = 'airtable';
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: `Make sure you enabled the following scopes for your token:<br>
				<code>data.records:read</code><br>
				<code>data.records:write</code><br>
				<code>schema.bases:read</code><br>
				`,
                name: 'notice',
                type: 'notice',
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.airtable.com/v0/meta/whoami',
            },
        };
    }
}
exports.AirtableTokenApi = AirtableTokenApi;
