// n8n-nodes-base/dist/credentials/AlienVaultApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlienVaultApi = void 0;
class AlienVaultApi {
    constructor() {
        this.name = 'alienVaultApi';
        this.displayName = 'AlienVault API';
        this.documentationUrl = 'alienvault';
        this.icon = 'file:icons/AlienVault.png';
        this.httpRequestNode = {
            name: 'AlienVault',
            docsUrl: 'https://otx.alienvault.com/api',
            apiBaseUrl: 'https://otx.alienvault.com/api/v1/',
        };
        this.properties = [
            {
                displayName: 'OTX Key',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-OTX-API-KEY': '={{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://otx.alienvault.com',
                url: '/api/v1/user/me',
            },
        };
    }
}
exports.AlienVaultApi = AlienVaultApi;
