// n8n-nodes-base/dist/credentials/BitwardenApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitwardenApi = void 0;
class BitwardenApi {
    constructor() {
        this.name = 'bitwardenApi';
        this.displayName = 'Bitwarden API';
        this.documentationUrl = 'bitwarden';
        this.properties = [
            {
                displayName: 'Client ID',
                name: 'clientId',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Client Secret',
                name: 'clientSecret',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Environment',
                name: 'environment',
                type: 'options',
                default: 'cloudHosted',
                options: [
                    {
                        name: 'Cloud-hosted',
                        value: 'cloudHosted',
                    },
                    {
                        name: 'Self-hosted',
                        value: 'selfHosted',
                    },
                ],
            },
            {
                displayName: 'Self-hosted domain',
                name: 'domain',
                type: 'string',
                default: '',
                placeholder: 'https://www.mydomain.com',
                displayOptions: {
                    show: {
                        environment: [
                            'selfHosted',
                        ],
                    },
                },
            },
        ];
    }
}
exports.BitwardenApi = BitwardenApi;
