// n8n-nodes-base/dist/credentials/CiscoMerakiApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CiscoMerakiApi = void 0;
class CiscoMerakiApi {
    constructor() {
        this.name = 'ciscoMerakiApi';
        this.displayName = 'Cisco Meraki API';
        this.documentationUrl = 'ciscomeraki';
        this.icon = { light: 'file:icons/Cisco.svg', dark: 'file:icons/Cisco.dark.svg' };
        this.httpRequestNode = {
            name: 'Cisco Meraki',
            docsUrl: 'https://developer.cisco.com/meraki/api/',
            apiBaseUrl: 'https://api.meraki.com/api/v1/',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-Cisco-Meraki-API-Key': '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.CiscoMerakiApi = CiscoMerakiApi;
