// n8n-nodes-base/dist/credentials/CrowdStrikeOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrowdStrikeOAuth2Api = void 0;
class CrowdStrikeOAuth2Api {
    constructor() {
        this.name = 'crowdStrikeOAuth2Api';
        this.displayName = 'CrowdStrike OAuth2 API';
        this.documentationUrl = 'crowdstrike';
        this.icon = { light: 'file:icons/CrowdStrike.svg', dark: 'file:icons/CrowdStrike.dark.svg' };
        this.httpRequestNode = {
            name: 'CrowdStrike',
            docsUrl: 'https://developer.crowdstrike.com/',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'Session Token',
                name: 'sessionToken',
                type: 'hidden',
                typeOptions: {
                    expirable: true,
                },
                default: '',
            },
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                required: true,
                default: '',
            },
            {
                displayName: 'Client ID',
                name: 'clientId',
                type: 'string',
                required: true,
                default: '',
            },
            {
                displayName: 'Client Secret',
                name: 'clientSecret',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.sessionToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.url}}',
                url: 'user-management/queries/users/v1',
            },
        };
    }
    async preAuthentication(credentials) {
        const url = credentials.url;
        const { access_token } = (await this.helpers.httpRequest({
            method: 'POST',
            url: `${url.endsWith('/') ? url.slice(0, -1) : url}/oauth2/token`,
            body: {
                client_id: credentials.clientId,
                client_secret: credentials.clientSecret,
            },
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
            },
        }));
        return { sessionToken: access_token };
    }
}
exports.CrowdStrikeOAuth2Api = CrowdStrikeOAuth2Api;
