// n8n-nodes-base/dist/credentials/CrowdDevApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrowdDevApi = void 0;
class CrowdDevApi {
    constructor() {
        this.name = 'crowdDevApi';
        this.displayName = 'crowd.dev API';
        this.documentationUrl = 'crowddev';
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                default: 'https://app.crowd.dev',
            },
            {
                displayName: 'Tenant ID',
                name: 'tenantId',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Token',
                name: 'token',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'Ignore SSL Issues',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                description: 'Whether to connect even if SSL certificate validation is not possible',
                default: false,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{"Bearer " + $credentials.token}}',
                },
            },
        };
        this.test = {
            request: {
                method: 'POST',
                baseURL: '={{$credentials.url.replace(/\\/$/, "") + "/api/tenant/" + $credentials.tenantId}}',
                url: '/member/query',
                skipSslCertificateValidation: '={{ $credentials.allowUnauthorizedCerts }}',
                body: {
                    limit: 1,
                    offset: 0,
                },
            },
        };
    }
}
exports.CrowdDevApi = CrowdDevApi;
