// n8n-nodes-base/dist/credentials/DiscordOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscordOAuth2Api = void 0;
class DiscordOAuth2Api {
    constructor() {
        this.name = 'discordOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'Discord OAuth2 API';
        this.documentationUrl = 'discord';
        this.properties = [
            {
                displayName: 'Bot Token',
                name: 'botToken',
                type: 'string',
                default: '',
                typeOptions: {
                    password: true,
                },
            },
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'authorizationCode',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://discord.com/api/oauth2/authorize',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://discord.com/api/oauth2/token',
                required: true,
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: 'identify guilds guilds.join bot',
                required: true,
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'permissions=1642758929655',
            },
        ];
    }
}
exports.DiscordOAuth2Api = DiscordOAuth2Api;
