// n8n-nodes-base/dist/credentials/DiscordBotApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscordBotApi = void 0;
class DiscordBotApi {
    constructor() {
        this.name = 'discordBotApi';
        this.displayName = 'Discord Bot API';
        this.documentationUrl = 'discord';
        this.properties = [
            {
                displayName: 'Bot Token',
                name: 'botToken',
                type: 'string',
                default: '',
                required: true,
                typeOptions: {
                    password: true,
                },
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bot {{$credentials.botToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://discord.com/api/v10/',
                url: '/users/@me/guilds',
            },
        };
    }
}
exports.DiscordBotApi = DiscordBotApi;
