// n8n-nodes-base/dist/credentials/FortiGateApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FortiGateApi = void 0;
class FortiGateApi {
    constructor() {
        this.name = 'fortiGateApi';
        this.displayName = 'Fortinet FortiGate API';
        this.documentationUrl = 'fortigate';
        this.icon = 'file:icons/Fortinet.svg';
        this.httpRequestNode = {
            name: 'Fortinet FortiGate',
            docsUrl: 'https://docs.fortinet.com/document/fortigate/7.4.1/administration-guide/940602/using-apis',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                required: true,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                qs: {
                    access_token: '={{$credentials.accessToken}}',
                },
            },
        };
    }
}
exports.FortiGateApi = FortiGateApi;
