// n8n-nodes-base/dist/credentials/GristApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GristApi = void 0;
class GristApi {
    constructor() {
        this.name = 'gristApi';
        this.displayName = 'Grist API';
        this.documentationUrl = 'grist';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'Plan Type',
                name: 'planType',
                type: 'options',
                default: 'free',
                options: [
                    {
                        name: 'Free',
                        value: 'free',
                    },
                    {
                        name: 'Paid',
                        value: 'paid',
                    },
                ],
            },
            {
                displayName: 'Custom Subdomain',
                name: 'customSubdomain',
                type: 'string',
                default: '',
                required: true,
                description: 'Custom subdomain of your team',
                displayOptions: {
                    show: {
                        planType: [
                            'paid',
                        ],
                    },
                },
            },
        ];
    }
}
exports.GristApi = GristApi;
