// n8n-nodes-base/dist/credentials/HttpSslAuth.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpSslAuth = void 0;
class HttpSslAuth {
    constructor() {
        this.name = 'httpSslAuth';
        this.displayName = 'SSL Certificates';
        this.documentationUrl = 'httpRequest';
        this.icon = 'node:n8n-nodes-base.httpRequest';
        this.properties = [
            {
                displayName: 'CA',
                name: 'ca',
                type: 'string',
                description: 'Certificate Authority certificate',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'Certificate',
                name: 'cert',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'Private Key',
                name: 'key',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'Passphrase',
                name: 'passphrase',
                type: 'string',
                description: 'Optional passphrase for the private key, if the private key is encrypted',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
        ];
    }
}
exports.HttpSslAuth = HttpSslAuth;
