// n8n-nodes-base/dist/credentials/HubspotDeveloperApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HubspotDeveloperApi = void 0;
const scopes = [
    'contacts',
];
class HubspotDeveloperApi {
    constructor() {
        this.name = 'hubspotDeveloperApi';
        this.displayName = 'Hubspot Developer API';
        this.documentationUrl = 'hubspot';
        this.extends = [
            'oAuth2Api',
        ];
        this.properties = [
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://app.hubspot.com/oauth/authorize',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://api.hubapi.com/oauth/v1/token',
                required: true,
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'grant_type=authorization_code',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'body',
                description: 'Resource to consume.',
            },
            {
                displayName: 'Developer API Key',
                name: 'apiKey',
                type: 'string',
                default: '',
            },
            {
                displayName: 'APP ID',
                name: 'appId',
                type: 'string',
                required: true,
                default: '',
                description: 'The APP ID',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
            },
        ];
    }
}
exports.HubspotDeveloperApi = HubspotDeveloperApi;
