// n8n-nodes-base/dist/credentials/HubspotOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HubspotOAuth2Api = void 0;
const scopes = [
    'contacts',
    'forms',
    'tickets',
];
class HubspotOAuth2Api {
    constructor() {
        this.name = 'hubspotOAuth2Api';
        this.extends = [
            'oAuth2Api',
        ];
        this.displayName = 'Hubspot OAuth2 API';
        this.documentationUrl = 'hubspot';
        this.properties = [
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://app.hubspot.com/oauth/authorize',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://api.hubapi.com/oauth/v1/token',
                required: true,
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'grant_type=authorization_code',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'body',
                description: 'Resource to consume.',
            },
        ];
    }
}
exports.HubspotOAuth2Api = HubspotOAuth2Api;
