// n8n-nodes-base/dist/credentials/HybridAnalysisApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HybridAnalysisApi = void 0;
class HybridAnalysisApi {
    constructor() {
        this.name = 'hybridAnalysisApi';
        this.displayName = 'Hybrid Analysis API';
        this.documentationUrl = 'hybridanalysis';
        this.icon = 'file:icons/Hybrid.png';
        this.httpRequestNode = {
            name: 'Hybrid Analysis',
            docsUrl: 'https://www.hybrid-analysis.com/docs/api/v2',
            apiBaseUrl: 'https://www.hybrid-analysis.com/api/',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'api-key': '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.HybridAnalysisApi = HybridAnalysisApi;
