// n8n-nodes-base/dist/credentials/KibanaApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KibanaApi = void 0;
class KibanaApi {
    constructor() {
        this.name = 'kibanaApi';
        this.displayName = 'Kibana API';
        this.documentationUrl = 'kibana';
        this.icon = 'file:icons/Kibana.svg';
        this.httpRequestNode = {
            name: 'Kibana',
            docsUrl: 'https://www.elastic.co/guide/en/kibana/current/api.html',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                required: true,
                default: '',
                placeholder: 'http://localhost:5601',
            },
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                required: true,
                default: '',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'kbn-xsrf': true,
                },
                auth: {
                    username: '={{$credentials.username}}',
                    password: '={{$credentials.password}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/api/features',
            },
        };
    }
}
exports.KibanaApi = KibanaApi;
