// n8n-nodes-base/dist/credentials/LinearOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinearOAuth2Api = void 0;
const scopes = ['read', 'write', 'issues:create', 'comments:create'];
class LinearOAuth2Api {
    constructor() {
        this.name = 'linearOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'Linear OAuth2 API';
        this.documentationUrl = 'linear';
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'authorizationCode',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://linear.app/oauth/authorize',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://api.linear.app/oauth/token',
                required: true,
            },
            {
                displayName: 'Actor',
                name: 'actor',
                type: 'options',
                options: [
                    {
                        name: 'User',
                        value: 'user',
                        description: 'Resources are created as the user who authorized the application',
                    },
                    {
                        name: 'Application',
                        value: 'application',
                        description: 'Resources are created as the application',
                    },
                ],
                default: 'user',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
                required: true,
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '={{"actor="+$self["actor"]}}',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'body',
            },
        ];
    }
}
exports.LinearOAuth2Api = LinearOAuth2Api;
