// n8n-nodes-base/dist/credentials/LinkedInCommunityManagementOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkedInCommunityManagementOAuth2Api = void 0;
const scopes = ['w_member_social', 'w_organization_social', 'r_basicprofile'];
class LinkedInCommunityManagementOAuth2Api {
    constructor() {
        this.name = 'linkedInCommunityManagementOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'LinkedIn Community Management OAuth2 API';
        this.documentationUrl = 'linkedIn';
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'authorizationCode',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://www.linkedin.com/oauth/v2/authorization',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://www.linkedin.com/oauth/v2/accessToken',
                required: true,
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'body',
            },
        ];
    }
}
exports.LinkedInCommunityManagementOAuth2Api = LinkedInCommunityManagementOAuth2Api;
