// n8n-nodes-base/dist/credentials/MispApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MispApi = void 0;
class MispApi {
    constructor() {
        this.name = 'mispApi';
        this.displayName = 'MISP API';
        this.documentationUrl = 'misp';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Allow Unauthorized Certificates',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                description: 'Whether to connect even if SSL certificate validation is not possible',
                default: false,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.baseUrl.replace(new RegExp("/$"), "")}}',
                url: '/tags',
                skipSslCertificateValidation: '={{$credentials.allowUnauthorizedCerts}}',
            },
        };
    }
}
exports.MispApi = MispApi;
