// n8n-nodes-base/dist/credentials/MistApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MistApi = void 0;
class MistApi {
    constructor() {
        this.name = 'mistApi';
        this.displayName = 'Mist API';
        this.icon = 'file:icons/Mist.svg';
        this.documentationUrl = 'mist';
        this.httpRequestNode = {
            name: 'Mist',
            docsUrl: 'https://www.mist.com/documentation/mist-api-introduction/',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'API Token',
                name: 'token',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
            {
                displayName: 'Region',
                name: 'region',
                type: 'options',
                options: [
                    {
                        name: 'Europe',
                        value: 'eu',
                    },
                    {
                        name: 'Global',
                        value: 'global',
                    },
                ],
                default: 'eu',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Token {{$credentials.token}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '=https://api{{$credentials.region === "eu" ? ".eu" : ""}}.mist.com',
                url: '/api/v1/self',
                method: 'GET',
            },
        };
    }
}
exports.MistApi = MistApi;
