// n8n-nodes-base/dist/credentials/MondayComApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MondayComApi = void 0;
class MondayComApi {
    constructor() {
        this.name = 'mondayComApi';
        this.displayName = 'Monday.com API';
        this.documentationUrl = 'mondayCom';
        this.properties = [
            {
                displayName: 'Token V2',
                name: 'apiToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiToken}}',
                },
            },
        };
        this.test = {
            request: {
                headers: {
                    'API-Version': '2023-10',
                    'Content-Type': 'application/json',
                },
                baseURL: 'https://api.monday.com/v2',
                method: 'POST',
                body: JSON.stringify({
                    query: 'query { me { name }}',
                }),
            },
        };
    }
}
exports.MondayComApi = MondayComApi;
