// n8n-nodes-base/dist/credentials/OktaApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OktaApi = void 0;
class OktaApi {
    constructor() {
        this.name = 'oktaApi';
        this.displayName = 'Okta API';
        this.documentationUrl = 'okta';
        this.icon = { light: 'file:icons/Okta.svg', dark: 'file:icons/Okta.dark.svg' };
        this.httpRequestNode = {
            name: 'Okta',
            docsUrl: 'https://developer.okta.com/docs/reference/',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                required: true,
                default: '',
                placeholder: 'https://dev-123456.okta.com',
            },
            {
                displayName: 'SSWS Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=SSWS {{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/api/v1/api-tokens',
            },
        };
    }
}
exports.OktaApi = OktaApi;
