// n8n-nodes-base/dist/credentials/OpenCTIApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenCTIApi = void 0;
class OpenCTIApi {
    constructor() {
        this.name = 'openCtiApi';
        this.displayName = 'OpenCTI API';
        this.documentationUrl = 'opencti';
        this.icon = 'file:icons/OpenCTI.png';
        this.httpRequestNode = {
            name: 'OpenCTI',
            docsUrl: 'https://docs.opencti.io/latest/deployment/integrations/?h=api#graphql-api',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                required: true,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.OpenCTIApi = OpenCTIApi;
