// n8n-nodes-base/dist/credentials/QRadarApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QRadarApi = void 0;
class QRadarApi {
    constructor() {
        this.name = 'qRadarApi';
        this.displayName = 'QRadar API';
        this.icon = { light: 'file:icons/IBM.svg', dark: 'file:icons/IBM.dark.svg' };
        this.documentationUrl = 'qradar';
        this.httpRequestNode = {
            name: 'QRadar',
            docsUrl: 'https://www.ibm.com/docs/en/qradar-common',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    SEC: '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.QRadarApi = QRadarApi;
