// n8n-nodes-base/dist/credentials/QualysApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QualysApi = void 0;
class QualysApi {
    constructor() {
        this.name = 'qualysApi';
        this.displayName = 'Qualys API';
        this.icon = 'file:icons/Qualys.svg';
        this.documentationUrl = 'qualys';
        this.httpRequestNode = {
            name: 'Qualys',
            docsUrl: 'https://qualysguard.qg2.apps.qualys.com/qwebhelp/fo_portal/api_doc/index.htm',
            apiBaseUrl: 'https://qualysapi.qualys.com/api/',
        };
        this.properties = [
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                required: true,
            },
            {
                displayName: 'Requested With',
                name: 'requestedWith',
                type: 'string',
                default: 'n8n application',
                description: 'User description, like a user agent',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-Requested-With': '={{$credentials.requestedWith}}',
                },
                auth: {
                    username: '={{$credentials.username}}',
                    password: '={{$credentials.password}}',
                },
            },
        };
    }
}
exports.QualysApi = QualysApi;
