// n8n-nodes-base/dist/credentials/RecordedFutureApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecordedFutureApi = void 0;
class RecordedFutureApi {
    constructor() {
        this.name = 'recordedFutureApi';
        this.displayName = 'Recorded Future API';
        this.documentationUrl = 'recordedfuture';
        this.icon = {
            light: 'file:icons/RecordedFuture.svg',
            dark: 'file:icons/RecordedFuture.dark.svg',
        };
        this.httpRequestNode = {
            name: 'Recorded Future',
            docsUrl: 'https://api.recordedfuture.com',
            apiBaseUrl: 'https://api.recordedfuture.com/v2/',
        };
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-RFToken': '={{$credentials.accessToken}}',
                },
            },
        };
    }
}
exports.RecordedFutureApi = RecordedFutureApi;
