// n8n-nodes-base/dist/credentials/SeaTableApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeaTableApi = void 0;
class SeaTableApi {
    constructor() {
        this.name = 'seaTableApi';
        this.displayName = 'SeaTable API';
        this.documentationUrl = 'seaTable';
        this.properties = [
            {
                displayName: 'Environment',
                name: 'environment',
                type: 'options',
                default: 'cloudHosted',
                options: [
                    {
                        name: 'Cloud-hosted',
                        value: 'cloudHosted',
                    },
                    {
                        name: 'Self-hosted',
                        value: 'selfHosted',
                    },
                ],
            },
            {
                displayName: 'Self-hosted domain',
                name: 'domain',
                type: 'string',
                default: '',
                placeholder: 'https://www.mydomain.com',
                displayOptions: {
                    show: {
                        environment: [
                            'selfHosted',
                        ],
                    },
                },
            },
            {
                displayName: 'API Token (of a Base)',
                name: 'token',
                type: 'string',
                default: '',
            },
        ];
    }
}
exports.SeaTableApi = SeaTableApi;
