// n8n-nodes-base/dist/credentials/SekoiaApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SekoiaApi = void 0;
class SekoiaApi {
    constructor() {
        this.name = 'sekoiaApi';
        this.displayName = 'Sekoia API';
        this.icon = 'file:icons/Sekoia.svg';
        this.documentationUrl = 'sekoia';
        this.httpRequestNode = {
            name: 'Sekoia',
            docsUrl: 'https://docs.sekoia.io/cti/features/integrations/api/',
            apiBaseUrl: 'https://api.sekoia.io/',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.SekoiaApi = SekoiaApi;
