// n8n-nodes-base/dist/credentials/ServiceNowOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceNowOAuth2Api = void 0;
class ServiceNowOAuth2Api {
    constructor() {
        this.name = 'serviceNowOAuth2Api';
        this.extends = [
            'oAuth2Api',
        ];
        this.displayName = 'ServiceNow OAuth2 API';
        this.documentationUrl = 'serviceNow';
        this.properties = [
            {
                displayName: 'Subdomain',
                name: 'subdomain',
                type: 'string',
                default: '',
                placeholder: 'n8n',
                description: 'The subdomain of your ServiceNow environment',
                required: true,
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: '=https://{{$self["subdomain"]}}.service-now.com/oauth_auth.do',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: '=https://{{$self["subdomain"]}}.service-now.com/oauth_token.do',
                required: true,
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: 'useraccount',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'response_type=code',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'grant_type=authorization_code',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'header',
            },
        ];
    }
}
exports.ServiceNowOAuth2Api = ServiceNowOAuth2Api;
