// n8n-nodes-base/dist/credentials/ShufflerApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShufflerApi = void 0;
class ShufflerApi {
    constructor() {
        this.name = 'shufflerApi';
        this.displayName = 'Shuffler API';
        this.icon = 'file:icons/Shuffler.svg';
        this.documentationUrl = 'shuffler';
        this.httpRequestNode = {
            name: 'Shuffler',
            docsUrl: 'https://shuffler.io/docs/API',
            apiBaseUrl: 'https://shuffler.io/api/v1/',
        };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://shuffler.io/api',
                url: '/v1/users/getusers',
                method: 'GET',
            },
        };
    }
}
exports.ShufflerApi = ShufflerApi;
