// n8n-nodes-base/dist/credentials/SurveyMonkeyApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SurveyMonkeyApi = void 0;
class SurveyMonkeyApi {
    constructor() {
        this.name = 'surveyMonkeyApi';
        this.displayName = 'SurveyMonkey API';
        this.documentationUrl = 'surveyMonkey';
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                default: '',
                description: `The access token must have the following scopes:</br />
			- Create/modify webhooks</br />
			- View webhooks</br />
			- View surveys</br />
			- View collectors</br />
			- View responses<br />
			- View response details`,
            },
            {
                displayName: 'Client ID',
                name: 'clientId',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Client Secret',
                name: 'clientSecret',
                type: 'string',
                default: '',
            },
        ];
    }
}
exports.SurveyMonkeyApi = SurveyMonkeyApi;
