// n8n-nodes-base/dist/credentials/TheHiveProjectApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TheHiveProjectApi = void 0;
class TheHiveProjectApi {
    constructor() {
        this.name = 'theHiveProjectApi';
        this.displayName = 'The Hive 5 API';
        this.documentationUrl = 'theHive';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'ApiKey',
                type: 'string',
                default: '',
                typeOptions: {
                    password: true,
                },
            },
            {
                displayName: 'URL',
                name: 'url',
                default: '',
                type: 'string',
                description: 'The URL of TheHive instance',
                placeholder: 'https://localhost:9000',
            },
            {
                displayName: 'Ignore SSL Issues',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                description: 'Whether to connect even if SSL certificate validation is not possible',
                default: false,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials?.ApiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.url}}',
                url: '/api/case',
            },
        };
    }
}
exports.TheHiveProjectApi = TheHiveProjectApi;
