// n8n-nodes-base/dist/credentials/TheHiveApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TheHiveApi = void 0;
class TheHiveApi {
    constructor() {
        this.name = 'theHiveApi';
        this.displayName = 'The Hive API';
        this.documentationUrl = 'theHive';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'ApiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'URL',
                name: 'url',
                default: '',
                type: 'string',
                description: 'The URL of TheHive instance',
                placeholder: 'https://localhost:9000',
            },
            {
                displayName: 'API Version',
                name: 'apiVersion',
                default: '',
                type: 'options',
                description: 'The version of api to be used',
                options: [
                    {
                        name: 'TheHive 4+ (api v1)',
                        value: 'v1',
                        description: 'API version with TheHive 4 support, also works with TheHive 5 but not all features are supported',
                    },
                    {
                        name: 'TheHive 3 (api v0)',
                        value: '',
                        description: 'API version with TheHive 3 support',
                    },
                ],
            },
            {
                displayName: 'Ignore SSL Issues',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                description: 'Whether to connect even if SSL certificate validation is not possible',
                default: false,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials?.ApiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.url}}',
                url: '/api/case',
            },
        };
    }
}
exports.TheHiveApi = TheHiveApi;
