// n8n-nodes-base/dist/credentials/TrellixEpoApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrellixEpoApi = void 0;
class TrellixEpoApi {
    constructor() {
        this.name = 'trellixEpoApi';
        this.displayName = 'Trellix (McAfee) ePolicy Orchestrator API';
        this.documentationUrl = 'trellixepo';
        this.icon = 'file:icons/Trellix.svg';
        this.httpRequestNode = {
            name: 'Trellix (McAfee) ePolicy Orchestrator',
            docsUrl: 'https://docs.trellix.com/en/bundle/epolicy-orchestrator-web-api-reference-guide',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                required: true,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                auth: {
                    username: '={{$credentials.username}}',
                    password: '={{$credentials.password}}',
                },
            },
        };
    }
}
exports.TrellixEpoApi = TrellixEpoApi;
