// n8n-nodes-base/dist/credentials/TwakeServerApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwakeServerApi = void 0;
class TwakeServerApi {
    constructor() {
        this.name = 'twakeServerApi';
        this.displayName = 'Twake Server API';
        this.icon = 'file:icons/Twake.png';
        this.documentationUrl = 'twake';
        this.httpRequestNode = {
            name: 'Twake Server',
            docsUrl: 'https://doc.twake.app/developers-api/home',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'Host URL',
                name: 'hostUrl',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Public ID',
                name: 'publicId',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Private API Key',
                name: 'privateApiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
    }
}
exports.TwakeServerApi = TwakeServerApi;
