// n8n-nodes-base/dist/credentials/VenafiTlsProtectCloudApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VenafiTlsProtectCloudApi = void 0;
class VenafiTlsProtectCloudApi {
    constructor() {
        this.name = 'venafiTlsProtectCloudApi';
        this.displayName = 'Venafi TLS Protect Cloud';
        this.documentationUrl = 'venafitlsprotectcloud';
        this.properties = [
            {
                displayName: 'Region',
                name: 'region',
                type: 'options',
                options: [
                    {
                        name: 'US',
                        value: 'cloud',
                    },
                    {
                        name: 'EU',
                        value: 'eu',
                    },
                ],
                default: 'cloud',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'tppl-api-key': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '=https://api.venafi.{{$credentials.region ?? "cloud"}}',
                url: '/v1/preferences',
            },
        };
    }
}
exports.VenafiTlsProtectCloudApi = VenafiTlsProtectCloudApi;
