// n8n-nodes-base/dist/credentials/VirusTotalApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirusTotalApi = void 0;
class VirusTotalApi {
    constructor() {
        this.name = 'virusTotalApi';
        this.displayName = 'VirusTotal API';
        this.documentationUrl = 'virustotal';
        this.icon = 'file:icons/VirusTotal.svg';
        this.httpRequestNode = {
            name: 'VirusTotal',
            docsUrl: 'https://developers.virustotal.com/reference/overview',
            apiBaseUrl: 'https://www.virustotal.com/api/v3/',
        };
        this.properties = [
            {
                displayName: 'API Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'x-apikey': '={{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://www.virustotal.com/api/v3',
                url: '/popular_threat_categories',
            },
        };
    }
}
exports.VirusTotalApi = VirusTotalApi;
