// n8n-nodes-base/dist/credentials/WhatsAppTriggerApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatsAppTriggerApi = void 0;
class WhatsAppTriggerApi {
    constructor() {
        this.name = 'whatsAppTriggerApi';
        this.displayName = 'WhatsApp API';
        this.documentationUrl = 'whatsApp';
        this.properties = [
            {
                displayName: 'Client ID',
                name: 'clientId',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'Client Secret',
                name: 'clientSecret',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                required: true,
            },
        ];
        this.test = {
            request: {
                method: 'POST',
                baseURL: 'https://graph.facebook.com/v19.0/oauth/access_token',
                body: {
                    client_id: '={{$credentials.clientId}}',
                    client_secret: '={{$credentials.clientSecret}}',
                    grant_type: 'client_credentials',
                },
            },
        };
    }
}
exports.WhatsAppTriggerApi = WhatsAppTriggerApi;
