// n8n-nodes-base/dist/nodes/Airtable/v2/actions/base/Base.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.getSchema = exports.getMany = void 0;
const getMany = __importStar(require("./getMany.operation"));
exports.getMany = getMany;
const getSchema = __importStar(require("./getSchema.operation"));
exports.getSchema = getSchema;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Get Many',
                value: 'getMany',
                description: 'List all the bases',
                action: 'Get many bases',
            },
            {
                name: 'Get Schema',
                value: 'getSchema',
                description: 'Get the schema of the tables in a base',
                action: 'Get base schema',
            },
        ],
        default: 'getMany',
        displayOptions: {
            show: {
                resource: ['base'],
            },
        },
    },
    ...getMany.description,
    ...getSchema.description,
];
