// n8n-nodes-base/dist/nodes/Discourse/SearchDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchFields = exports.searchOperations = void 0;
exports.searchOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        description: 'Choose an operation',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'search',
                ],
            },
        },
        options: [
            {
                name: 'Query',
                value: 'query',
                description: 'Search for something',
            },
        ],
        default: 'query',
    },
];
exports.searchFields = [
    {
        displayName: 'Term',
        name: 'term',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'search',
                ],
                operation: [
                    'query',
                ],
            },
        },
        default: '',
        description: 'Term to search for.',
    },
    {
        displayName: 'Simplify Response',
        name: 'simple',
        type: 'boolean',
        displayOptions: {
            show: {
                resource: [
                    'search',
                ],
                operation: [
                    'query',
                ],
            },
        },
        default: true,
        description: 'Return a simplified version of the response instead of the raw data.',
    },
];
