// n8n-nodes-base/dist/nodes/Mattermost/v1/actions/user/invite/description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userInviteDescription = void 0;
exports.userInviteDescription = [
    {
        displayName: 'Team ID',
        name: 'teamId',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getTeams',
        },
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'user',
                ],
                operation: [
                    'invite',
                ],
            },
        },
        default: '',
    },
    {
        displayName: 'Emails',
        name: 'emails',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'user',
                ],
                operation: [
                    'invite',
                ],
            },
        },
        default: '',
        description: `User's email. Multiple can be set separated by comma.`,
    },
];
