// n8n-nodes-base/dist/nodes/Code/Code.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Code = void 0;
const set_1 = __importDefault(require("lodash/set"));
const JavascriptCodeDescription_1 = require("./descriptions/JavascriptCodeDescription");
const PythonCodeDescription_1 = require("./descriptions/PythonCodeDescription");
const JavaScriptSandbox_1 = require("./JavaScriptSandbox");
const PythonSandbox_1 = require("./PythonSandbox");
const Sandbox_1 = require("./Sandbox");
const utils_1 = require("./utils");
const { CODE_ENABLE_STDOUT } = process.env;
class Code {
    constructor() {
        this.description = {
            displayName: 'Code',
            name: 'code',
            icon: 'file:code.svg',
            group: ['transform'],
            version: [1, 2],
            defaultVersion: 2,
            description: 'Run custom JavaScript or Python code',
            defaults: {
                name: 'Code',
            },
            inputs: ['main'],
            outputs: ['main'],
            parameterPane: 'wide',
            properties: [
                {
                    displayName: 'Mode',
                    name: 'mode',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Run Once for All Items',
                            value: 'runOnceForAllItems',
                            description: 'Run this code only once, no matter how many input items there are',
                        },
                        {
                            name: 'Run Once for Each Item',
                            value: 'runOnceForEachItem',
                            description: 'Run this code as many times as there are input items',
                        },
                    ],
                    default: 'runOnceForAllItems',
                },
                {
                    displayName: 'Language',
                    name: 'language',
                    type: 'options',
                    noDataExpression: true,
                    displayOptions: {
                        show: {
                            '@version': [2],
                        },
                    },
                    options: [
                        {
                            name: 'JavaScript',
                            value: 'javaScript',
                        },
                        {
                            name: 'Python (Beta)',
                            value: 'python',
                        },
                    ],
                    default: 'javaScript',
                },
                {
                    displayName: 'Language',
                    name: 'language',
                    type: 'hidden',
                    displayOptions: {
                        show: {
                            '@version': [1],
                        },
                    },
                    default: 'javaScript',
                },
                ...JavascriptCodeDescription_1.javascriptCodeDescription,
                ...PythonCodeDescription_1.pythonCodeDescription,
            ],
        };
    }
    async execute() {
        const nodeMode = this.getNodeParameter('mode', 0);
        const workflowMode = this.getMode();
        const node = this.getNode();
        const language = node.typeVersion === 2
            ? this.getNodeParameter('language', 0)
            : 'javaScript';
        const codeParameterName = language === 'python' ? 'pythonCode' : 'jsCode';
        const getSandbox = (index = 0) => {
            const code = this.getNodeParameter(codeParameterName, index);
            const context = Sandbox_1.getSandboxContext.call(this, index);
            if (nodeMode === 'runOnceForAllItems') {
                context.items = context.$input.all();
            }
            else {
                context.item = context.$input.item;
            }
            const Sandbox = language === 'python' ? PythonSandbox_1.PythonSandbox : JavaScriptSandbox_1.JavaScriptSandbox;
            const sandbox = new Sandbox(context, code, index, this.helpers);
            sandbox.on('output', workflowMode === 'manual'
                ? this.sendMessageToUI
                : CODE_ENABLE_STDOUT === 'true'
                    ? (...args) => console.log(`[Workflow "${this.getWorkflow().id}"][Node "${node.name}"]`, ...args)
                    : () => { });
            return sandbox;
        };
        if (nodeMode === 'runOnceForAllItems') {
            const sandbox = getSandbox();
            let items;
            try {
                items = (await sandbox.runCodeAllItems());
            }
            catch (error) {
                if (!this.continueOnFail(error)) {
                    (0, set_1.default)(error, 'node', node);
                    throw error;
                }
                items = [{ json: { error: error.message } }];
            }
            for (const item of items) {
                (0, utils_1.standardizeOutput)(item.json);
            }
            return [items];
        }
        const returnData = [];
        const items = this.getInputData();
        for (let index = 0; index < items.length; index++) {
            const sandbox = getSandbox(index);
            let result;
            try {
                result = await sandbox.runCodeEachItem();
            }
            catch (error) {
                if (!this.continueOnFail(error)) {
                    (0, set_1.default)(error, 'node', node);
                    throw error;
                }
                returnData.push({
                    json: { error: error.message },
                    pairedItem: {
                        item: index,
                    },
                });
            }
            if (result) {
                returnData.push({
                    json: (0, utils_1.standardizeOutput)(result.json),
                    pairedItem: { item: index },
                    ...(result.binary && { binary: result.binary }),
                });
            }
        }
        return [returnData];
    }
}
exports.Code = Code;
