// n8n-nodes-base/dist/nodes/Mattermost/v1/actions/reaction/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.descriptions = exports.getAll = exports.delete = exports.create = void 0;
const create = require("./create");
exports.create = create;
const del = require("./del");
exports.delete = del;
const getAll = require("./getAll");
exports.getAll = getAll;
exports.descriptions = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'reaction',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Add a reaction to a post.',
            },
            {
                name: 'Delete',
                value: 'delete',
                description: 'Remove a reaction from a post',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Get all the reactions to one or more posts',
            },
        ],
        default: 'create',
        description: 'The operation to perform',
    },
    ...create.description,
    ...del.description,
    ...getAll.description,
];
