// n8n-nodes-base/dist/nodes/Microsoft/Sql/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteOperation = exports.updateOperation = exports.insertOperation = exports.mssqlChunk = exports.configurePool = exports.formatColumns = exports.executeQueryQueue = exports.createTableStruct = exports.copyInputItem = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const mssql_1 = __importDefault(require("mssql"));
const utilities_1 = require("../../../utils/utilities");
function copyInputItem(item, properties) {
    const newItem = {};
    for (const property of properties) {
        if (item.json[property] === undefined) {
            newItem[property] = null;
        }
        else {
            newItem[property] = (0, n8n_workflow_1.deepCopy)(item.json[property]);
        }
    }
    return newItem;
}
exports.copyInputItem = copyInputItem;
function createTableStruct(getNodeParam, items, additionalProperties = [], keyName) {
    return items.reduce((tables, item, index) => {
        const table = getNodeParam('table', index);
        const columnString = getNodeParam('columns', index);
        const columns = columnString.split(',').map((column) => column.trim());
        const itemCopy = copyInputItem(item, columns.concat(additionalProperties));
        const keyParam = keyName ? getNodeParam(keyName, index) : undefined;
        if (tables[table] === undefined) {
            tables[table] = {};
        }
        if (tables[table][columnString] === undefined) {
            tables[table][columnString] = [];
        }
        if (keyName) {
            itemCopy[keyName] = keyParam;
        }
        tables[table][columnString].push(itemCopy);
        return tables;
    }, {});
}
exports.createTableStruct = createTableStruct;
async function executeQueryQueue(tables, buildQueryQueue) {
    return await Promise.all(Object.keys(tables).map(async (table) => {
        const columnsResults = Object.keys(tables[table]).map(async (columnString) => {
            return await Promise.all(buildQueryQueue({
                table,
                columnString,
                items: tables[table][columnString],
            }));
        });
        return await Promise.all(columnsResults);
    }));
}
exports.executeQueryQueue = executeQueryQueue;
function formatColumns(columns) {
    return columns
        .split(',')
        .map((column) => `[${column.trim()}]`)
        .join(', ');
}
exports.formatColumns = formatColumns;
function configurePool(credentials) {
    const config = {
        server: credentials.server,
        port: credentials.port,
        database: credentials.database,
        user: credentials.user,
        password: credentials.password,
        domain: credentials.domain ? credentials.domain : undefined,
        connectionTimeout: credentials.connectTimeout,
        requestTimeout: credentials.requestTimeout,
        options: {
            encrypt: credentials.tls,
            enableArithAbort: false,
            tdsVersion: credentials.tdsVersion,
            trustServerCertificate: credentials.allowUnauthorizedCerts,
        },
    };
    return new mssql_1.default.ConnectionPool(config);
}
exports.configurePool = configurePool;
const escapeTableName = (table) => {
    table = table.trim();
    if (table.startsWith('[') && table.endsWith(']')) {
        return table;
    }
    else {
        return `[${table}]`;
    }
};
const MSSQL_PARAMETER_LIMIT = 2100;
function mssqlChunk(rows) {
    const chunked = [[]];
    let currentParamCount = 0;
    for (const row of rows) {
        const rowValues = Object.values(row);
        const valueCount = rowValues.length;
        if (currentParamCount + valueCount >= MSSQL_PARAMETER_LIMIT) {
            chunked.push([]);
            currentParamCount = 0;
        }
        chunked[chunked.length - 1].push(row);
        currentParamCount += valueCount;
    }
    return chunked;
}
exports.mssqlChunk = mssqlChunk;
async function insertOperation(tables, pool) {
    return await executeQueryQueue(tables, ({ table, columnString, items }) => {
        return mssqlChunk(items).map(async (insertValues) => {
            const request = pool.request();
            const valuesPlaceholder = [];
            for (const [rIndex, entry] of insertValues.entries()) {
                const row = Object.values(entry);
                valuesPlaceholder.push(`(${row.map((_, vIndex) => `@r${rIndex}v${vIndex}`).join(', ')})`);
                for (const [vIndex, value] of row.entries()) {
                    request.input(`r${rIndex}v${vIndex}`, value);
                }
            }
            const query = `INSERT INTO ${escapeTableName(table)} (${formatColumns(columnString)}) VALUES ${valuesPlaceholder.join(', ')};`;
            return await request.query(query);
        });
    });
}
exports.insertOperation = insertOperation;
async function updateOperation(tables, pool) {
    return await executeQueryQueue(tables, ({ table, columnString, items }) => {
        return items.map(async (item) => {
            const request = pool.request();
            const columns = columnString.split(',').map((column) => column.trim());
            const setValues = [];
            const condition = `${item.updateKey} = @condition`;
            request.input('condition', item[item.updateKey]);
            for (const [index, col] of columns.entries()) {
                setValues.push(`[${col}] = @v${index}`);
                request.input(`v${index}`, item[col]);
            }
            const query = `UPDATE ${escapeTableName(table)} SET ${setValues.join(', ')} WHERE ${condition};`;
            return await request.query(query);
        });
    });
}
exports.updateOperation = updateOperation;
async function deleteOperation(tables, pool) {
    const queriesResults = await Promise.all(Object.keys(tables).map(async (table) => {
        const deleteKeyResults = Object.keys(tables[table]).map(async (deleteKey) => {
            const deleteItemsList = (0, utilities_1.chunk)(tables[table][deleteKey].map((item) => copyInputItem(item, [deleteKey])), 1000);
            const queryQueue = deleteItemsList.map(async (deleteValues) => {
                const request = pool.request();
                const valuesPlaceholder = [];
                for (const [index, entry] of deleteValues.entries()) {
                    valuesPlaceholder.push(`@v${index}`);
                    request.input(`v${index}`, entry[deleteKey]);
                }
                const query = `DELETE FROM ${escapeTableName(table)} WHERE [${deleteKey}] IN (${valuesPlaceholder.join(', ')});`;
                return await request.query(query);
            });
            return await Promise.all(queryQueue);
        });
        return await Promise.all(deleteKeyResults);
    }));
    return (0, utilities_1.flatten)(queriesResults).reduce((acc, resp) => (acc += resp.rowsAffected.reduce((sum, val) => (sum += val))), 0);
}
exports.deleteOperation = deleteOperation;
