// n8n/dist/databases/dsl/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSchemaBuilder = void 0;
const Column_1 = require("./Column");
const Table_1 = require("./Table");
const Indices_1 = require("./Indices");
const createSchemaBuilder = (tablePrefix, queryRunner) => ({
    column: (name) => new Column_1.Column(name),
    createTable: (tableName) => new Table_1.CreateTable(tableName, tablePrefix, queryRunner),
    dropTable: (tableName) => new Table_1.DropTable(tableName, tablePrefix, queryRunner),
    addColumns: (tableName, columns) => new Table_1.AddColumns(tableName, columns, tablePrefix, queryRunner),
    dropColumns: (tableName, columnNames) => new Table_1.DropColumns(tableName, columnNames, tablePrefix, queryRunner),
    createIndex: (tableName, columnNames, isUnique = false, customIndexName) => new Indices_1.CreateIndex(tableName, columnNames, isUnique, tablePrefix, queryRunner, customIndexName),
    dropIndex: (tableName, columnNames, customIndexName) => new Indices_1.DropIndex(tableName, columnNames, tablePrefix, queryRunner, customIndexName),
    addForeignKey: (tableName, columnName, reference, customConstraintName) => new Table_1.AddForeignKey(tableName, columnName, reference, tablePrefix, queryRunner, customConstraintName),
    dropForeignKey: (tableName, columnName, reference, customConstraintName) => new Table_1.DropForeignKey(tableName, columnName, reference, tablePrefix, queryRunner, customConstraintName),
    addNotNull: (tableName, columnName) => new Table_1.AddNotNull(tableName, columnName, tablePrefix, queryRunner),
    dropNotNull: (tableName, columnName) => new Table_1.DropNotNull(tableName, columnName, tablePrefix, queryRunner),
});
exports.createSchemaBuilder = createSchemaBuilder;
