// n8n-nodes-base/dist/nodes/MessageBird/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageBirdApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function messageBirdApiRequest(method, resource, body, query = {}) {
    const credentials = await this.getCredentials('messageBirdApi');
    if (credentials === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials returned!');
    }
    const options = {
        headers: {
            Accept: 'application/json',
            Authorization: `AccessKey ${credentials.accessKey}`,
        },
        method,
        qs: query,
        body,
        uri: `https://rest.messagebird.com${resource}`,
        json: true,
    };
    try {
        if (Object.keys(body).length === 0) {
            delete options.body;
        }
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.messageBirdApiRequest = messageBirdApiRequest;
