// @n8n/n8n-nodes-langchain/dist/nodes/memory/MemoryZep/MemoryZep.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryZep = void 0;
const zep_1 = require("@langchain/community/memory/zep");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
const descriptions_1 = require("../descriptions");
const helpers_1 = require("../../../utils/helpers");
class MemoryZep {
    constructor() {
        this.description = {
            displayName: 'Zep',
            name: 'memoryZep',
            icon: 'file:zep.png',
            group: ['transform'],
            version: [1, 1.1, 1.2],
            description: 'Use Zep Memory',
            defaults: {
                name: 'Zep',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Memory'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.memoryzep/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_memory"],
            outputNames: ['Memory'],
            credentials: [
                {
                    name: 'zepApi',
                    required: true,
                },
            ],
            properties: [
                (0, sharedFields_1.getConnectionHintNoticeField)(["ai_agent"]),
                {
                    displayName: 'Session ID',
                    name: 'sessionId',
                    type: 'string',
                    required: true,
                    default: '',
                    displayOptions: {
                        show: {
                            '@version': [1],
                        },
                    },
                },
                {
                    displayName: 'Session ID',
                    name: 'sessionId',
                    type: 'string',
                    default: '={{ $json.sessionId }}',
                    description: 'The key to use to store the memory',
                    displayOptions: {
                        show: {
                            '@version': [1.1],
                        },
                    },
                },
                {
                    ...descriptions_1.sessionIdOption,
                    displayOptions: {
                        show: {
                            '@version': [{ _cnd: { gte: 1.2 } }],
                        },
                    },
                },
                descriptions_1.sessionKeyProperty,
            ],
        };
    }
    async supplyData(itemIndex) {
        const credentials = (await this.getCredentials('zepApi'));
        const nodeVersion = this.getNode().typeVersion;
        let sessionId;
        if (nodeVersion >= 1.2) {
            sessionId = (0, helpers_1.getSessionId)(this, itemIndex);
        }
        else {
            sessionId = this.getNodeParameter('sessionId', itemIndex);
        }
        const memory = new zep_1.ZepMemory({
            sessionId,
            baseURL: credentials.apiUrl,
            apiKey: credentials.apiKey,
            memoryKey: 'chat_history',
            returnMessages: true,
            inputKey: 'input',
            outputKey: 'output',
        });
        return {
            response: (0, logWrapper_1.logWrapper)(memory, this),
        };
    }
}
exports.MemoryZep = MemoryZep;
