// n8n-nodes-base/dist/nodes/Dropcontact/GenericFunction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateCrendetials = exports.dropcontactApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function dropcontactApiRequest(method, endpoint, body, qs) {
    const { apiKey } = await this.getCredentials('dropcontactApi');
    const options = {
        headers: {
            'user-agent': 'n8n',
            'X-Access-Token': apiKey,
        },
        method,
        uri: `https://api.dropcontact.io${endpoint}`,
        qs,
        body,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.dropcontactApiRequest = dropcontactApiRequest;
async function validateCrendetials(decryptedCredentials) {
    const credentials = decryptedCredentials;
    const { apiKey } = credentials;
    const options = {
        headers: {
            'user-agent': 'n8n',
            'X-Access-Token': apiKey,
        },
        method: 'POST',
        body: {
            data: [{ email: '' }],
        },
        uri: `https://api.dropcontact.io/batch`,
        json: true,
    };
    return this.helpers.request(options);
}
exports.validateCrendetials = validateCrendetials;
