// n8n-nodes-base/dist/nodes/Mattermost/v1/actions/channel/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.descriptions = exports.statistics = exports.addUser = exports.restore = exports.members = exports.delete = exports.create = void 0;
const create = require("./create");
exports.create = create;
const del = require("./del");
exports.delete = del;
const members = require("./members");
exports.members = members;
const restore = require("./restore");
exports.restore = restore;
const addUser = require("./addUser");
exports.addUser = addUser;
const statistics = require("./statistics");
exports.statistics = statistics;
exports.descriptions = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'channel',
                ],
            },
        },
        options: [
            {
                name: 'Add User',
                value: 'addUser',
                description: 'Add a user to a channel',
            },
            {
                name: 'Create',
                value: 'create',
                description: 'Create a new channel',
            },
            {
                name: 'Delete',
                value: 'delete',
                description: 'Soft delete a channel',
            },
            {
                name: 'Member',
                value: 'members',
                description: 'Get a page of members for a channel',
            },
            {
                name: 'Restore',
                value: 'restore',
                description: 'Restores a soft deleted channel',
            },
            {
                name: 'Statistics',
                value: 'statistics',
                description: 'Get statistics for a channel',
            },
        ],
        default: 'create',
        description: 'The operation to perform.',
    },
    ...create.description,
    ...del.description,
    ...members.description,
    ...restore.description,
    ...addUser.description,
    ...statistics.description,
];
