// n8n-nodes-base/dist/nodes/TheHiveProject/actions/node.description.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
const alert = __importStar(require("./alert"));
const case_ = __importStar(require("./case"));
const comment = __importStar(require("./comment"));
const log = __importStar(require("./log"));
const observable = __importStar(require("./observable"));
const query = __importStar(require("./query"));
const task = __importStar(require("./task"));
const page = __importStar(require("./page"));
exports.description = {
    displayName: 'TheHive 5',
    name: 'theHiveProject',
    icon: 'file:thehiveproject.svg',
    group: ['transform'],
    subtitle: '={{$parameter["operation"]}} : {{$parameter["resource"]}}',
    version: 1,
    description: 'Consume TheHive 5 API',
    defaults: {
        name: 'TheHive 5',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'theHiveProjectApi',
            required: true,
        },
    ],
    properties: [
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            noDataExpression: true,
            required: true,
            options: [
                {
                    name: 'Alert',
                    value: 'alert',
                },
                {
                    name: 'Case',
                    value: 'case',
                },
                {
                    name: 'Comment',
                    value: 'comment',
                },
                {
                    name: 'Observable',
                    value: 'observable',
                },
                {
                    name: 'Page',
                    value: 'page',
                },
                {
                    name: 'Query',
                    value: 'query',
                },
                {
                    name: 'Task',
                    value: 'task',
                },
                {
                    name: 'Task Log',
                    value: 'log',
                },
            ],
            default: 'alert',
        },
        ...alert.description,
        ...case_.description,
        ...comment.description,
        ...log.description,
        ...observable.description,
        ...page.description,
        ...query.description,
        ...task.description,
    ],
};
