// n8n-nodes-base/dist/nodes/TheHiveProject/actions/page/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.update = exports.search = exports.deletePage = exports.create = void 0;
const create = __importStar(require("./create.operation"));
exports.create = create;
const deletePage = __importStar(require("./deletePage.operation"));
exports.deletePage = deletePage;
const search = __importStar(require("./search.operation"));
exports.search = search;
const update = __importStar(require("./update.operation"));
exports.update = update;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        noDataExpression: true,
        type: 'options',
        required: true,
        default: 'create',
        options: [
            {
                name: 'Create',
                value: 'create',
                action: 'Create a page',
            },
            {
                name: 'Delete',
                value: 'deletePage',
                action: 'Delete a page',
            },
            {
                name: 'Search',
                value: 'search',
                action: 'Search pages',
            },
            {
                name: 'Update',
                value: 'update',
                action: 'Update a page',
            },
        ],
        displayOptions: {
            show: {
                resource: ['page'],
            },
        },
    },
    ...create.description,
    ...deletePage.description,
    ...search.description,
    ...update.description,
];
