// n8n-nodes-base/dist/nodes/TheHiveProject/actions/page/search.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const descriptions_1 = require("../../descriptions");
const transport_1 = require("../../transport");
const utilities_1 = require("../../../../utils/utilities");
const properties = [
    {
        displayName: 'Search in Knowledge Base',
        name: 'searchInKnowledgeBase',
        type: 'boolean',
        default: true,
        description: 'Whether to search in knowledge base or only in the selected case',
    },
    {
        ...descriptions_1.caseRLC,
        displayOptions: {
            show: {
                searchInKnowledgeBase: [false],
            },
        },
    },
    ...descriptions_1.returnAllAndLimit,
    descriptions_1.genericFiltersCollection,
    descriptions_1.sortCollection,
    {
        ...descriptions_1.searchOptions,
        displayOptions: {
            show: {
                returnAll: [true],
            },
        },
    },
];
const displayOptions = {
    show: {
        resource: ['page'],
        operation: ['search'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const searchInKnowledgeBase = this.getNodeParameter('searchInKnowledgeBase', i);
    const filtersValues = this.getNodeParameter('filters.values', i, []);
    const sortFields = this.getNodeParameter('sort.fields', i, []);
    const returnAll = this.getNodeParameter('returnAll', i);
    let returnCount = false;
    if (!returnAll) {
        returnCount = this.getNodeParameter('options.returnCount', i, false);
    }
    let limit;
    let scope;
    if (searchInKnowledgeBase) {
        scope = { query: 'listOrganisationPage' };
    }
    else {
        const caseId = this.getNodeParameter('caseId', i, '', { extractValue: true });
        scope = { query: 'getCase', id: caseId, restrictTo: 'pages' };
    }
    if (!returnAll) {
        limit = this.getNodeParameter('limit', i);
    }
    responseData = await transport_1.theHiveApiQuery.call(this, scope, filtersValues, sortFields, limit, returnCount);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
