// n8n-nodes-base/dist/nodes/HomeAssistant/CameraProxyDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cameraProxyFields = exports.cameraProxyOperations = void 0;
exports.cameraProxyOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'cameraProxy',
                ],
            },
        },
        options: [
            {
                name: 'Get Screenshot',
                value: 'getScreenshot',
                description: 'Get the camera screenshot',
            },
        ],
        default: 'getScreenshot',
        description: 'The operation to perform.',
    },
];
exports.cameraProxyFields = [
    {
        displayName: 'Camera Entity ID',
        name: 'cameraEntityId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                operation: [
                    'getScreenshot',
                ],
                resource: [
                    'cameraProxy',
                ],
            },
        },
        description: 'The camera entity ID.',
    },
    {
        displayName: 'Binary Property',
        name: 'binaryPropertyName',
        type: 'string',
        required: true,
        default: 'data',
        displayOptions: {
            show: {
                operation: [
                    'getScreenshot',
                ],
                resource: [
                    'cameraProxy',
                ],
            },
        },
        description: 'Name of the binary property to which to write the data of the read file.',
    },
];
