// n8n-nodes-base/dist/nodes/Lemlist/descriptions/ActivityDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activityFields = exports.activityOperations = void 0;
exports.activityOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        default: 'getAll',
        description: 'Operation to perform',
        options: [
            {
                name: 'Get All',
                value: 'getAll',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'activity',
                ],
            },
        },
    },
];
exports.activityFields = [
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'activity',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 5,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
            maxValue: 1000,
        },
        displayOptions: {
            show: {
                resource: [
                    'activity',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'activity',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        options: [
            {
                displayName: 'Campaign ID',
                name: 'campaignId',
                type: 'options',
                required: true,
                default: '',
                typeOptions: {
                    loadOptionsMethod: 'getCampaigns',
                },
                description: 'ID of the campaign to retrieve activity for.',
            },
            {
                displayName: 'Type',
                name: 'type',
                type: 'options',
                default: 'emailsOpened',
                description: 'Type of activity to retrieve.',
                options: [
                    {
                        name: 'Emails Bounced',
                        value: 'emailsBounced',
                    },
                    {
                        name: 'Emails Clicked',
                        value: 'emailsClicked',
                    },
                    {
                        name: 'Emails Opened',
                        value: 'emailsOpened',
                    },
                    {
                        name: 'Emails Replied',
                        value: 'emailsReplied',
                    },
                    {
                        name: 'Emails Send Failed',
                        value: 'emailsSendFailed',
                    },
                    {
                        name: 'Emails Sent',
                        value: 'emailsSent',
                    },
                    {
                        name: 'Emails Unsubscribed',
                        value: 'emailsUnsubscribed',
                    },
                ],
            },
        ],
    },
];
