// n8n-nodes-base/dist/nodes/Lemlist/descriptions/CampaignDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.campaignFields = exports.campaignOperations = void 0;
exports.campaignOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        default: 'getAll',
        description: 'Operation to perform',
        options: [
            {
                name: 'Get All',
                value: 'getAll',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'campaign',
                ],
            },
        },
    },
];
exports.campaignFields = [
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'campaign',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 5,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
            maxValue: 1000,
        },
        displayOptions: {
            show: {
                resource: [
                    'campaign',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
];
