// n8n-nodes-base/dist/nodes/FacebookLeadAds/methods/listSearch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formList = exports.pageList = void 0;
const GenericFunctions_1 = require("../GenericFunctions");
const filterMatches = (name, filter) => !filter || (name === null || name === void 0 ? void 0 : name.toLowerCase().includes(filter.toLowerCase()));
async function pageList(filter, paginationToken) {
    var _a;
    const { data: pages, paging } = await GenericFunctions_1.facebookPageList.call(this, paginationToken);
    return {
        results: pages
            .filter((page) => filterMatches(page.name, filter))
            .map((page) => ({
            name: page.name,
            value: page.id,
            url: `https://facebook.com/${page.id}`,
        })),
        paginationToken: (_a = paging === null || paging === void 0 ? void 0 : paging.cursors) === null || _a === void 0 ? void 0 : _a.after,
    };
}
exports.pageList = pageList;
async function formList(filter, paginationToken) {
    var _a;
    const pageId = this.getNodeParameter('page', '', { extractValue: true });
    const { data: forms, paging } = await GenericFunctions_1.facebookFormList.call(this, pageId, paginationToken);
    return {
        results: forms
            .filter((form) => filterMatches(form.name, filter))
            .map((form) => ({
            name: form.name,
            value: form.id,
        })),
        paginationToken: (_a = paging === null || paging === void 0 ? void 0 : paging.cursors) === null || _a === void 0 ? void 0 : _a.after,
    };
}
exports.formList = formList;
