// @n8n/n8n-nodes-langchain/dist/nodes/vendors/OpenAi/actions/audio/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.translate = exports.transcribe = exports.generate = void 0;
const generate = __importStar(require("./generate.operation"));
exports.generate = generate;
const transcribe = __importStar(require("./transcribe.operation"));
exports.transcribe = transcribe;
const translate = __importStar(require("./translate.operation"));
exports.translate = translate;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Generate Audio',
                value: 'generate',
                action: 'Generate audio',
                description: 'Creates audio from a text prompt',
            },
            {
                name: 'Transcribe a Recording',
                value: 'transcribe',
                action: 'Transcribe a recording',
                description: 'Transcribes audio into the text',
            },
            {
                name: 'Translate a Recording',
                value: 'translate',
                action: 'Translate a recording',
                description: 'Translate audio into the text in the english language',
            },
        ],
        default: 'generate',
        displayOptions: {
            show: {
                resource: ['audio'],
            },
        },
    },
    {
        displayName: 'OpenAI API limits the size of the audio file to 25 MB',
        name: 'fileSizeLimitNotice',
        type: 'notice',
        default: ' ',
        displayOptions: {
            show: {
                resource: ['audio'],
                operation: ['translate', 'transcribe'],
            },
        },
    },
    ...generate.description,
    ...transcribe.description,
    ...translate.description,
];
