// n8n/dist/databases/entities/AbstractEntity.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithTimestampsAndStringId = exports.WithTimestamps = exports.WithStringId = exports.datetimeColumnType = exports.jsonColumnType = void 0;
const typeorm_1 = require("@n8n/typeorm");
const config_1 = __importDefault(require("../../config"));
const generators_1 = require("../utils/generators");
const dbType = config_1.default.getEnv('database.type');
const timestampSyntax = {
    sqlite: "STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')",
    postgresdb: 'CURRENT_TIMESTAMP(3)',
    mysqldb: 'CURRENT_TIMESTAMP(3)',
    mariadb: 'CURRENT_TIMESTAMP(3)',
}[dbType];
exports.jsonColumnType = dbType === 'sqlite' ? 'simple-json' : 'json';
exports.datetimeColumnType = dbType === 'postgresdb' ? 'timestamptz' : 'datetime';
const tsColumnOptions = {
    precision: 3,
    default: () => timestampSyntax,
    type: exports.datetimeColumnType,
};
function mixinStringId(base) {
    class Derived extends base {
        generateId() {
            if (!this.id) {
                this.id = (0, generators_1.generateNanoId)();
            }
        }
    }
    __decorate([
        (0, typeorm_1.PrimaryColumn)('varchar'),
        __metadata("design:type", String)
    ], Derived.prototype, "id", void 0);
    __decorate([
        (0, typeorm_1.BeforeInsert)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Derived.prototype, "generateId", null);
    return Derived;
}
function mixinTimestamps(base) {
    class Derived extends base {
        setUpdateDate() {
            this.updatedAt = new Date();
        }
    }
    __decorate([
        (0, typeorm_1.CreateDateColumn)(tsColumnOptions),
        __metadata("design:type", Date)
    ], Derived.prototype, "createdAt", void 0);
    __decorate([
        (0, typeorm_1.UpdateDateColumn)(tsColumnOptions),
        __metadata("design:type", Date)
    ], Derived.prototype, "updatedAt", void 0);
    __decorate([
        (0, typeorm_1.BeforeUpdate)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Derived.prototype, "setUpdateDate", null);
    return Derived;
}
class BaseEntity {
}
exports.WithStringId = mixinStringId(BaseEntity);
exports.WithTimestamps = mixinTimestamps(BaseEntity);
exports.WithTimestampsAndStringId = mixinStringId(exports.WithTimestamps);
