// @n8n/n8n-nodes-langchain/dist/nodes/chains/ChainSummarization/ChainSummarization.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainSummarization = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const ChainSummarizationV1_node_1 = require("./V1/ChainSummarizationV1.node");
const ChainSummarizationV2_node_1 = require("./V2/ChainSummarizationV2.node");
class ChainSummarization extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            displayName: 'Summarization Chain',
            name: 'chainSummarization',
            icon: 'fa:link',
            group: ['transform'],
            description: 'Transforms text into a concise summary',
            codex: {
                alias: ['LangChain'],
                categories: ['AI'],
                subcategories: {
                    AI: ['Chains', 'Root Nodes'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/root-nodes/n8n-nodes-langchain.chainsummarization/',
                        },
                    ],
                },
            },
            defaultVersion: 2,
        };
        const nodeVersions = {
            1: new ChainSummarizationV1_node_1.ChainSummarizationV1(baseDescription),
            2: new ChainSummarizationV2_node_1.ChainSummarizationV2(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.ChainSummarization = ChainSummarization;
