// n8n-nodes-base/dist/nodes/Mocean/Mocean.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mocean = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
class Mocean {
    constructor() {
        this.description = {
            displayName: 'Mocean',
            name: 'mocean',
            icon: 'file:mocean.png',
            group: ['transform'],
            version: 1,
            description: 'Send SMS and voice messages via Mocean',
            defaults: {
                name: 'Mocean',
                color: '#772244',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'moceanApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    options: [
                        {
                            name: 'SMS',
                            value: 'sms',
                        },
                        {
                            name: 'Voice',
                            value: 'voice',
                        },
                    ],
                    default: 'sms',
                },
                {
                    displayName: 'Operation',
                    name: 'operation',
                    type: 'options',
                    displayOptions: {
                        show: {
                            resource: [
                                'sms',
                                'voice',
                            ],
                        },
                    },
                    options: [
                        {
                            name: 'Send',
                            value: 'send',
                            description: 'Send SMS/Voice message',
                        },
                    ],
                    default: 'send',
                    description: 'The operation to perform.',
                },
                {
                    displayName: 'From',
                    name: 'from',
                    type: 'string',
                    default: '',
                    placeholder: 'Sender Number',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                                'voice',
                            ],
                        },
                    },
                    description: 'The number to which to send the message',
                },
                {
                    displayName: 'To',
                    name: 'to',
                    type: 'string',
                    default: '',
                    placeholder: 'Receipient number',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                                'voice',
                            ],
                        },
                    },
                    description: 'The number from which to send the message',
                },
                {
                    displayName: 'Language',
                    name: 'language',
                    type: 'options',
                    options: [
                        {
                            name: 'Chinese Mandarin (China)',
                            value: 'cmn-CN',
                        },
                        {
                            name: 'English (United Kingdom)',
                            value: 'en-GB',
                        },
                        {
                            name: 'English (United States)',
                            value: 'en-US',
                        },
                        {
                            name: 'Japanese (Japan)',
                            value: 'ja-JP',
                        },
                        {
                            name: 'Korean (Korea)',
                            value: 'ko-KR',
                        },
                    ],
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'voice',
                            ],
                        },
                    },
                    default: 'en-US',
                },
                {
                    displayName: 'Message',
                    name: 'message',
                    type: 'string',
                    default: '',
                    placeholder: '',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                                'voice',
                            ],
                        },
                    },
                    description: 'The message to send',
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        let endpoint;
        let operation;
        let requesetMethod;
        let resource;
        let text;
        let dataKey;
        let body;
        let qs;
        for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
            body = {};
            qs = {};
            try {
                resource = this.getNodeParameter('resource', itemIndex, '');
                operation = this.getNodeParameter('operation', itemIndex, '');
                text = this.getNodeParameter('message', itemIndex, '');
                requesetMethod = 'POST';
                body['mocean-from'] = this.getNodeParameter('from', itemIndex, '');
                body['mocean-to'] = this.getNodeParameter('to', itemIndex, '');
                if (resource === 'voice') {
                    const language = this.getNodeParameter('language', itemIndex);
                    const command = [
                        {
                            action: 'say',
                            language,
                            text,
                        },
                    ];
                    dataKey = 'voice';
                    body['mocean-command'] = JSON.stringify(command);
                    endpoint = '/rest/2/voice/dial';
                }
                else if (resource === 'sms') {
                    dataKey = 'messages';
                    body['mocean-text'] = text;
                    endpoint = '/rest/2/sms';
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unknown resource ${resource}`);
                }
                if (operation === 'send') {
                    const responseData = await GenericFunctions_1.moceanApiRequest.call(this, requesetMethod, endpoint, body, qs);
                    for (const item of responseData[dataKey]) {
                        item.type = resource;
                        returnData.push(item);
                    }
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unknown operation ${operation}`);
                }
            }
            catch (error) {
                if (this.continueOnFail()) {
                    returnData.push({ error: error.message });
                    continue;
                }
                throw error;
            }
        }
        return [this.helpers.returnJsonArray(returnData)];
    }
}
exports.Mocean = Mocean;
