// @n8n/n8n-nodes-langchain/dist/credentials/HuggingFaceApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HuggingFaceApi = void 0;
class HuggingFaceApi {
    constructor() {
        this.name = 'huggingFaceApi';
        this.displayName = 'HuggingFaceApi';
        this.documentationUrl = 'huggingface';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api-inference.huggingface.co',
                url: '/models/gpt2',
            },
        };
    }
}
exports.HuggingFaceApi = HuggingFaceApi;
