// @n8n/n8n-nodes-langchain/dist/credentials/ZepApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZepApi = void 0;
class ZepApi {
    constructor() {
        this.name = 'zepApi';
        this.displayName = 'Zep Api';
        this.documentationUrl = 'zep';
        this.properties = [
            {
                displayName: 'API URL',
                name: 'apiUrl',
                required: true,
                type: 'string',
                default: 'http://localhost:8000',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: false,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{$credentials.apiKey ? "Bearer " + $credentials.apiKey : undefined }}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.apiUrl}}',
                url: '/api/v1/collection',
            },
        };
    }
}
exports.ZepApi = ZepApi;
