// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolSerpApi/ToolSerpApi.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolSerpApi = void 0;
const serpapi_1 = require("@langchain/community/tools/serpapi");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
class ToolSerpApi {
    constructor() {
        this.description = {
            displayName: 'SerpApi (Google Search)',
            name: 'toolSerpApi',
            icon: 'file:serpApi.svg',
            group: ['transform'],
            version: 1,
            description: 'Search in Google using SerpAPI',
            defaults: {
                name: 'SerpAPI',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolserpapi/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            credentials: [
                {
                    name: 'serpApi',
                    required: true,
                },
            ],
            properties: [
                (0, sharedFields_1.getConnectionHintNoticeField)(["ai_agent"]),
                {
                    displayName: 'Options',
                    name: 'options',
                    type: 'collection',
                    placeholder: 'Add Option',
                    default: {},
                    options: [
                        {
                            displayName: 'Country',
                            name: 'gl',
                            type: 'string',
                            default: 'us',
                            description: 'Defines the country to use for search. Head to <a href="https://serpapi.com/google-countries">Google countries page</a> for a full list of supported countries.',
                        },
                        {
                            displayName: 'Device',
                            name: 'device',
                            type: 'options',
                            options: [
                                {
                                    name: 'Desktop',
                                    value: 'desktop',
                                },
                                {
                                    name: 'Mobile',
                                    value: 'mobile',
                                },
                                {
                                    name: 'Tablet',
                                    value: 'tablet',
                                },
                            ],
                            default: 'desktop',
                            description: 'Device to use to get the results',
                        },
                        {
                            displayName: 'Explicit Array',
                            name: 'no_cache',
                            type: 'boolean',
                            default: false,
                            description: 'Whether to force SerpApi to fetch the Google results even if a cached version is already present. Cache expires after 1h. Cached searches are free, and are not counted towards your searches per month.',
                        },
                        {
                            displayName: 'Google Domain',
                            name: 'google_domain',
                            type: 'string',
                            default: 'google.com',
                            description: 'Defines the domain to use for search. Head to <a href="https://serpapi.com/google-domains">Google domains page</a> for a full list of supported domains.',
                        },
                        {
                            displayName: 'Language',
                            name: 'hl',
                            type: 'string',
                            default: 'en',
                            description: 'Defines the language to use. It\'s a two-letter language code. (e.g., `en` for English, `es` for Spanish, or `fr` for French). Head to <a href="https://serpapi.com/google-languages">Google languages page</a> for a full list of supported languages.',
                        },
                    ],
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const credentials = await this.getCredentials('serpApi');
        const options = this.getNodeParameter('options', itemIndex);
        return {
            response: (0, logWrapper_1.logWrapper)(new serpapi_1.SerpAPI(credentials.apiKey, options), this),
        };
    }
}
exports.ToolSerpApi = ToolSerpApi;
