// n8n-nodes-base/dist/nodes/WhatsApp/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.appWebhookSubscriptionDelete = exports.appWebhookSubscriptionCreate = exports.appWebhookSubscriptionList = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function appAccessTokenRead() {
    const credentials = await this.getCredentials('whatsAppTriggerApi');
    const options = {
        headers: {
            'content-type': 'application/x-www-form-urlencoded',
        },
        method: 'POST',
        body: {
            client_id: credentials.clientId,
            client_secret: credentials.clientSecret,
            grant_type: 'client_credentials',
        },
        url: 'https://graph.facebook.com/v19.0/oauth/access_token',
        json: true,
    };
    try {
        return await this.helpers.httpRequest.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
async function whatsappApiRequest(method, resource, body, qs = {}) {
    const tokenResponse = await appAccessTokenRead.call(this);
    const appAccessToken = tokenResponse.access_token;
    const options = {
        headers: {
            accept: 'application/json',
            authorization: `Bearer ${appAccessToken}`,
        },
        method,
        qs,
        body: body === null || body === void 0 ? void 0 : body.payload,
        url: `https://graph.facebook.com/v19.0${resource}`,
        json: true,
    };
    try {
        return await this.helpers.httpRequest.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
async function appWebhookSubscriptionList(appId) {
    const response = (await whatsappApiRequest.call(this, 'GET', `/${appId}/subscriptions`));
    return response.data;
}
exports.appWebhookSubscriptionList = appWebhookSubscriptionList;
async function appWebhookSubscriptionCreate(appId, subscription) {
    return await whatsappApiRequest.call(this, 'POST', `/${appId}/subscriptions`, {
        type: 'form',
        payload: { ...subscription },
    });
}
exports.appWebhookSubscriptionCreate = appWebhookSubscriptionCreate;
async function appWebhookSubscriptionDelete(appId, object) {
    return await whatsappApiRequest.call(this, 'DELETE', `/${appId}/subscriptions`, {
        type: 'form',
        payload: { object },
    });
}
exports.appWebhookSubscriptionDelete = appWebhookSubscriptionDelete;
