// n8n-nodes-base/dist/nodes/TheHiveProject/actions/alert/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.update = exports.promote = exports.merge = exports.status = exports.search = exports.get = exports.deleteAlert = exports.executeResponder = exports.create = void 0;
const create = __importStar(require("./create.operation"));
exports.create = create;
const executeResponder = __importStar(require("./executeResponder.operation"));
exports.executeResponder = executeResponder;
const deleteAlert = __importStar(require("./deleteAlert.operation"));
exports.deleteAlert = deleteAlert;
const get = __importStar(require("./get.operation"));
exports.get = get;
const search = __importStar(require("./search.operation"));
exports.search = search;
const status = __importStar(require("./status.operation"));
exports.status = status;
const merge = __importStar(require("./merge.operation"));
exports.merge = merge;
const promote = __importStar(require("./promote.operation"));
exports.promote = promote;
const update = __importStar(require("./update.operation"));
exports.update = update;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        required: true,
        options: [
            {
                name: 'Create',
                value: 'create',
                action: 'Create an alert',
            },
            {
                name: 'Delete',
                value: 'deleteAlert',
                action: 'Delete an alert',
            },
            {
                name: 'Execute Responder',
                value: 'executeResponder',
                action: 'Execute responder on an alert',
            },
            {
                name: 'Get',
                value: 'get',
                action: 'Get an alert',
            },
            {
                name: 'Merge Into Case',
                value: 'merge',
                action: 'Merge an alert into a case',
            },
            {
                name: 'Promote to Case',
                value: 'promote',
                action: 'Promote an alert to a case',
            },
            {
                name: 'Search',
                value: 'search',
                action: 'Search alerts',
            },
            {
                name: 'Update',
                value: 'update',
                action: 'Update an alert',
            },
            {
                name: 'Update Status',
                value: 'status',
                action: 'Update an alert status',
            },
        ],
        displayOptions: {
            show: {
                resource: ['alert'],
            },
        },
        default: 'create',
    },
    ...create.description,
    ...deleteAlert.description,
    ...executeResponder.description,
    ...get.description,
    ...search.description,
    ...status.description,
    ...merge.description,
    ...promote.description,
    ...update.description,
];
