// n8n-nodes-base/dist/nodes/TheHiveProject/actions/alert/search.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const descriptions_1 = require("../../descriptions");
const transport_1 = require("../../transport");
const utilities_1 = require("../../../../utils/utilities");
const properties = [
    ...descriptions_1.returnAllAndLimit,
    descriptions_1.genericFiltersCollection,
    descriptions_1.sortCollection,
    descriptions_1.searchOptions,
];
const displayOptions = {
    show: {
        resource: ['alert'],
        operation: ['search'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const filtersValues = this.getNodeParameter('filters.values', i, []);
    const sortFields = this.getNodeParameter('sort.fields', i, []);
    const returnAll = this.getNodeParameter('returnAll', i);
    const { returnCount, extraData } = this.getNodeParameter('options', i);
    let limit;
    if (!returnAll) {
        limit = this.getNodeParameter('limit', i);
    }
    responseData = await transport_1.theHiveApiQuery.call(this, { query: 'listAlert' }, filtersValues, sortFields, limit, returnCount, extraData);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
