// n8n-nodes-base/dist/nodes/Misp/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwOnInvalidUrl = exports.throwOnMissingSharingGroup = exports.throwOnEmptyUpdate = exports.mispApiRestSearch = exports.mispApiRequestAllItems = exports.mispApiRequest = void 0;
const url_1 = require("url");
const n8n_workflow_1 = require("n8n-workflow");
async function mispApiRequest(method, endpoint, body = {}, qs = {}) {
    var _a;
    const { baseUrl, allowUnauthorizedCerts } = (await this.getCredentials('mispApi'));
    const options = {
        method,
        body,
        qs,
        uri: `${baseUrl}${endpoint}`,
        json: true,
        rejectUnauthorized: !allowUnauthorizedCerts,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'mispApi', options);
    }
    catch (error) {
        if (error.statusCode === 403) {
            error.statusCode = 400;
        }
        const errors = (_a = error === null || error === void 0 ? void 0 : error.error) === null || _a === void 0 ? void 0 : _a.errors;
        if (errors) {
            const key = Object.keys(errors)[0];
            if (key !== undefined) {
                let message = errors[key].join();
                if (message.includes(' nter')) {
                    message = message.replace(' nter', ' enter');
                }
                error.error.message = `${error.error.message}: ${message}`;
            }
        }
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.mispApiRequest = mispApiRequest;
async function mispApiRequestAllItems(endpoint) {
    const responseData = await mispApiRequest.call(this, 'GET', endpoint);
    const returnAll = this.getNodeParameter('returnAll', 0);
    if (!returnAll) {
        const limit = this.getNodeParameter('limit', 0);
        return responseData.slice(0, limit);
    }
    return responseData;
}
exports.mispApiRequestAllItems = mispApiRequestAllItems;
async function mispApiRestSearch(resource, itemIndex) {
    let body = {};
    const useJson = this.getNodeParameter('useJson', itemIndex);
    if (useJson) {
        const json = this.getNodeParameter('jsonOutput', itemIndex);
        if (typeof json === 'string') {
            body = (0, n8n_workflow_1.jsonParse)(json);
        }
        else {
            body = json;
        }
    }
    else {
        const value = this.getNodeParameter('value', itemIndex);
        const additionalFields = this.getNodeParameter('additionalFields', itemIndex);
        body.value = value;
        if (Object.keys(additionalFields).length) {
            if (additionalFields.tags) {
                additionalFields.tags = additionalFields.tags
                    .split(',')
                    .map((tag) => tag.trim());
            }
            Object.assign(body, additionalFields);
        }
    }
    const endpoint = `/${resource}/restSearch`;
    const { response } = await mispApiRequest.call(this, 'POST', endpoint, body);
    if (response) {
        if (resource === 'attributes') {
            return response.Attribute;
        }
        if (resource === 'events') {
            return response.map((event) => event.Event);
        }
        if (resource === 'objects') {
            return response.map((obj) => obj.Object);
        }
    }
    else {
        return [];
    }
}
exports.mispApiRestSearch = mispApiRestSearch;
function throwOnEmptyUpdate(resource, updateFields) {
    if (!Object.keys(updateFields).length) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`);
    }
}
exports.throwOnEmptyUpdate = throwOnEmptyUpdate;
const SHARING_GROUP_OPTION_ID = 4;
function throwOnMissingSharingGroup(fields) {
    if (fields.distribution === SHARING_GROUP_OPTION_ID && !fields.sharing_group_id) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please specify a sharing group');
    }
}
exports.throwOnMissingSharingGroup = throwOnMissingSharingGroup;
const isValidUrl = (str) => {
    try {
        new url_1.URL(str);
        return true;
    }
    catch (error) {
        return false;
    }
};
function throwOnInvalidUrl(str) {
    if (!isValidUrl(str)) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please specify a valid URL, protocol included. Example: https://site.com');
    }
}
exports.throwOnInvalidUrl = throwOnInvalidUrl;
