// n8n-nodes-base/dist/nodes/Microsoft/Outlook/v2/actions/message/delete.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = exports.properties = void 0;
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utilities_1 = require("../../../../../../utils/utilities");
exports.properties = [descriptions_1.messageRLC];
const displayOptions = {
    show: {
        resource: ['message'],
        operation: ['delete'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
async function execute(index) {
    const messageId = this.getNodeParameter('messageId', index, undefined, {
        extractValue: true,
    });
    await transport_1.microsoftApiRequest.call(this, 'DELETE', `/messages/${messageId}`);
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: index } });
    return executionData;
}
exports.execute = execute;
