// n8n-nodes-base/dist/nodes/Microsoft/Outlook/v2/actions/message/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.update = exports.send = exports.reply = exports.move = exports.getAll = exports.get = exports.delete = void 0;
const del = __importStar(require("./delete.operation"));
exports.delete = del;
const get = __importStar(require("./get.operation"));
exports.get = get;
const getAll = __importStar(require("./getAll.operation"));
exports.getAll = getAll;
const move = __importStar(require("./move.operation"));
exports.move = move;
const reply = __importStar(require("./reply.operation"));
exports.reply = reply;
const send = __importStar(require("./send.operation"));
exports.send = send;
const update = __importStar(require("./update.operation"));
exports.update = update;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['message'],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'delete',
                description: 'Delete a message',
                action: 'Delete a message',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Retrieve a single message',
                action: 'Get a message',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'List and search messages',
                action: 'Get many messages',
            },
            {
                name: 'Move',
                value: 'move',
                description: 'Move a message to a folder',
                action: 'Move a message',
            },
            {
                name: 'Reply',
                value: 'reply',
                description: 'Create a reply to a message',
                action: 'Reply to a message',
            },
            {
                name: 'Send',
                value: 'send',
                description: 'Send a message',
                action: 'Send a message',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update a message',
                action: 'Update a message',
            },
        ],
        default: 'send',
    },
    ...del.description,
    ...get.description,
    ...getAll.description,
    ...move.description,
    ...reply.description,
    ...send.description,
    ...update.description,
];
