// n8n-nodes-base/dist/nodes/Code/descriptions/PythonCodeDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pythonCodeDescription = void 0;
const commonDescription = {
    displayName: 'Python',
    name: 'pythonCode',
    type: 'string',
    typeOptions: {
        editor: 'codeNodeEditor',
        editorLanguage: 'python',
    },
    default: '',
    description: 'Python code to execute.<br><br>Tip: You can use built-in methods and variables like <code>_today</code> for dates and <code>_jmespath</code> for querying JSON structures. <a href="https://docs.n8n.io/code/builtin/">Learn more</a>.',
    noDataExpression: true,
};
exports.pythonCodeDescription = [
    {
        ...commonDescription,
        displayOptions: {
            show: {
                language: ['python'],
                mode: ['runOnceForAllItems'],
            },
        },
    },
    {
        ...commonDescription,
        displayOptions: {
            show: {
                language: ['python'],
                mode: ['runOnceForEachItem'],
            },
        },
    },
    {
        displayName: 'Debug by using <code>print()</code> statements and viewing their output in the browser console.',
        name: 'notice',
        type: 'notice',
        displayOptions: {
            show: {
                language: ['python'],
            },
        },
        default: '',
    },
];
