// n8n-nodes-base/dist/nodes/Copper/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.copperApiRequestAllItems = exports.handleListing = exports.adjustTaskFields = exports.adjustPersonFields = exports.adjustLeadFields = exports.adjustCompanyFields = exports.adjustEmail = exports.adjustProjectIds = exports.adjustEmails = exports.adjustPhoneNumbers = exports.adjustAddress = exports.getAutomaticSecret = exports.copperApiRequest = void 0;
const crypto_1 = require("crypto");
const n8n_workflow_1 = require("n8n-workflow");
const lodash_1 = require("lodash");
async function copperApiRequest(method, resource, body = {}, qs = {}, uri = '', option = {}) {
    const credentials = await this.getCredentials('copperApi');
    let options = {
        headers: {
            'X-PW-AccessToken': credentials.apiKey,
            'X-PW-Application': 'developer_api',
            'X-PW-UserEmail': credentials.email,
            'Content-Type': 'application/json',
        },
        method,
        qs,
        body,
        uri: uri || `https://api.prosperworks.com/developer_api/v1${resource}`,
        json: true,
    };
    options = Object.assign({}, options, option);
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    if (Object.keys(options.body).length === 0) {
        delete options.body;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.copperApiRequest = copperApiRequest;
function getAutomaticSecret(credentials) {
    const data = `${credentials.email},${credentials.apiKey}`;
    return crypto_1.createHash('md5').update(data).digest('hex');
}
exports.getAutomaticSecret = getAutomaticSecret;
function adjustAddress(fixedCollection) {
    if (!fixedCollection.address)
        return fixedCollection;
    const adjusted = lodash_1.omit(fixedCollection, ['address']);
    if (fixedCollection.address) {
        adjusted.address = fixedCollection.address.addressFields;
    }
    return adjusted;
}
exports.adjustAddress = adjustAddress;
function adjustPhoneNumbers(fixedCollection) {
    if (!fixedCollection.phone_numbers)
        return fixedCollection;
    const adjusted = lodash_1.omit(fixedCollection, ['phone_numbers']);
    if (fixedCollection.phone_numbers) {
        adjusted.phone_numbers = fixedCollection.phone_numbers.phoneFields;
    }
    return adjusted;
}
exports.adjustPhoneNumbers = adjustPhoneNumbers;
function adjustEmails(fixedCollection) {
    if (!fixedCollection.emails)
        return fixedCollection;
    const adjusted = lodash_1.omit(fixedCollection, ['emails']);
    if (fixedCollection.emails) {
        adjusted.emails = fixedCollection.emails.emailFields;
    }
    return adjusted;
}
exports.adjustEmails = adjustEmails;
function adjustProjectIds(fields) {
    if (!fields.project_ids)
        return fields;
    const adjusted = lodash_1.omit(fields, ['project_ids']);
    adjusted.project_ids = fields.project_ids.includes(',')
        ? fields.project_ids.split(',')
        : [fields.project_ids];
    return adjusted;
}
exports.adjustProjectIds = adjustProjectIds;
function adjustEmail(fixedCollection) {
    if (!fixedCollection.email)
        return fixedCollection;
    const adjusted = lodash_1.omit(fixedCollection, ['email']);
    if (fixedCollection.email) {
        adjusted.email = fixedCollection.email.emailFields;
    }
    return adjusted;
}
exports.adjustEmail = adjustEmail;
exports.adjustCompanyFields = lodash_1.flow(adjustAddress, adjustPhoneNumbers);
exports.adjustLeadFields = lodash_1.flow(exports.adjustCompanyFields, adjustEmail);
exports.adjustPersonFields = lodash_1.flow(exports.adjustCompanyFields, adjustEmails);
exports.adjustTaskFields = lodash_1.flow(exports.adjustLeadFields, adjustProjectIds);
async function handleListing(method, endpoint, qs = {}, body = {}, uri = '') {
    let responseData;
    const returnAll = this.getNodeParameter('returnAll', 0);
    const option = { resolveWithFullResponse: true };
    if (returnAll) {
        return await copperApiRequestAllItems.call(this, method, endpoint, body, qs, uri, option);
    }
    const limit = this.getNodeParameter('limit', 0);
    responseData = await copperApiRequestAllItems.call(this, method, endpoint, body, qs, uri, option);
    return responseData.slice(0, limit);
}
exports.handleListing = handleListing;
async function copperApiRequestAllItems(method, resource, body = {}, qs = {}, uri = '', option = {}) {
    let responseData;
    qs.page_size = 200;
    let totalItems = 0;
    const returnData = [];
    do {
        responseData = await copperApiRequest.call(this, method, resource, body, qs, uri, option);
        totalItems = responseData.headers['x-pw-total'];
        returnData.push(...responseData.body);
    } while (totalItems > returnData.length);
    return returnData;
}
exports.copperApiRequestAllItems = copperApiRequestAllItems;
