// n8n-nodes-base/dist/nodes/Google/Drive/v2/actions/drive/Drive.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.update = exports.list = exports.get = exports.deleteDrive = exports.create = void 0;
const create = __importStar(require("./create.operation"));
exports.create = create;
const deleteDrive = __importStar(require("./deleteDrive.operation"));
exports.deleteDrive = deleteDrive;
const get = __importStar(require("./get.operation"));
exports.get = get;
const list = __importStar(require("./list.operation"));
exports.list = list;
const update = __importStar(require("./update.operation"));
exports.update = update;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create a shared drive',
                action: 'Create shared drive',
            },
            {
                name: 'Delete',
                value: 'deleteDrive',
                description: 'Permanently delete a shared drive',
                action: 'Delete shared drive',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Get a shared drive',
                action: 'Get shared drive',
            },
            {
                name: 'Get Many',
                value: 'list',
                description: 'Get the list of shared drives',
                action: 'Get many shared drives',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update a shared drive',
                action: 'Update shared drive',
            },
        ],
        default: 'create',
        displayOptions: {
            show: {
                resource: ['drive'],
            },
        },
    },
    ...create.description,
    ...deleteDrive.description,
    ...get.description,
    ...list.description,
    ...update.description,
];
