// n8n-nodes-base/dist/nodes/Airtable/v2/actions/versionDescription.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const record = __importStar(require("./record/Record.resource"));
const base = __importStar(require("./base/Base.resource"));
exports.versionDescription = {
    displayName: 'Airtable',
    name: 'airtable',
    icon: 'file:airtable.svg',
    group: ['input'],
    version: [2, 2.1],
    subtitle: '={{ $parameter["operation"] + ": " + $parameter["resource"] }}',
    description: 'Read, update, write and delete data from Airtable',
    defaults: {
        name: 'Airtable',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'airtableTokenApi',
            required: true,
            displayOptions: {
                show: {
                    authentication: ['airtableTokenApi'],
                },
            },
        },
        {
            name: 'airtableOAuth2Api',
            required: true,
            displayOptions: {
                show: {
                    authentication: ['airtableOAuth2Api'],
                },
            },
        },
    ],
    properties: [
        {
            displayName: 'Authentication',
            name: 'authentication',
            type: 'options',
            options: [
                {
                    name: 'Access Token',
                    value: 'airtableTokenApi',
                },
                {
                    name: 'OAuth2',
                    value: 'airtableOAuth2Api',
                },
            ],
            default: 'airtableTokenApi',
        },
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            noDataExpression: true,
            options: [
                {
                    name: 'Base',
                    value: 'base',
                },
                {
                    name: 'Record',
                    value: 'record',
                },
            ],
            default: 'record',
        },
        ...record.description,
        ...base.description,
    ],
};
