// @n8n/n8n-nodes-langchain/dist/nodes/text_splitters/TextSplitterTokenSplitter/TextSplitterTokenSplitter.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextSplitterTokenSplitter = void 0;
const textsplitters_1 = require("@langchain/textsplitters");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
class TextSplitterTokenSplitter {
    constructor() {
        this.description = {
            displayName: 'Token Splitter',
            name: 'textSplitterTokenSplitter',
            icon: 'fa:grip-lines-vertical',
            group: ['transform'],
            version: 1,
            description: 'Split text into chunks by tokens',
            defaults: {
                name: 'Token Splitter',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Text Splitters'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.textsplittertokensplitter/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_textSplitter"],
            outputNames: ['Text Splitter'],
            properties: [
                (0, sharedFields_1.getConnectionHintNoticeField)(["ai_document"]),
                {
                    displayName: 'Chunk Size',
                    name: 'chunkSize',
                    type: 'number',
                    default: 1000,
                },
                {
                    displayName: 'Chunk Overlap',
                    name: 'chunkOverlap',
                    type: 'number',
                    default: 0,
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        this.logger.verbose('Supply Data for Text Splitter');
        const chunkSize = this.getNodeParameter('chunkSize', itemIndex);
        const chunkOverlap = this.getNodeParameter('chunkOverlap', itemIndex);
        const splitter = new textsplitters_1.TokenTextSplitter({
            chunkSize,
            chunkOverlap,
            allowedSpecial: 'all',
            disallowedSpecial: 'all',
            encodingName: 'cl100k_base',
            keepSeparator: false,
        });
        return {
            response: (0, logWrapper_1.logWrapper)(splitter, this),
        };
    }
}
exports.TextSplitterTokenSplitter = TextSplitterTokenSplitter;
