// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolWikipedia/ToolWikipedia.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolWikipedia = void 0;
const wikipedia_query_run_1 = require("@langchain/community/tools/wikipedia_query_run");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
class ToolWikipedia {
    constructor() {
        this.description = {
            displayName: 'Wikipedia',
            name: 'toolWikipedia',
            icon: 'file:wikipedia.svg',
            group: ['transform'],
            version: 1,
            description: 'Search in Wikipedia',
            defaults: {
                name: 'Wikipedia',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolwikipedia/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            properties: [(0, sharedFields_1.getConnectionHintNoticeField)(["ai_agent"])],
        };
    }
    async supplyData() {
        const WikiTool = new wikipedia_query_run_1.WikipediaQueryRun();
        WikiTool.description =
            'A tool for interacting with and fetching data from the Wikipedia API. The input should always be a string query.';
        return {
            response: (0, logWrapper_1.logWrapper)(WikiTool, this),
        };
    }
}
exports.ToolWikipedia = ToolWikipedia;
