// n8n-nodes-base/dist/nodes/Microsoft/Outlook/v2/actions/contact/create.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = exports.properties = void 0;
const utils_1 = require("../../helpers/utils");
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utilities_1 = require("../../../../../../utils/utilities");
exports.properties = [
    {
        displayName: 'First Name',
        name: 'givenName',
        type: 'string',
        default: '',
        required: true,
    },
    {
        displayName: 'Last Name',
        name: 'surname',
        type: 'string',
        default: '',
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        options: descriptions_1.contactFields,
    },
];
const displayOptions = {
    show: {
        resource: ['contact'],
        operation: ['create'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
async function execute(index) {
    const additionalFields = this.getNodeParameter('additionalFields', index);
    const givenName = this.getNodeParameter('givenName', index);
    const surname = this.getNodeParameter('surname', index);
    const body = {
        givenName,
        ...(0, utils_1.prepareContactFields)(additionalFields),
    };
    if (surname) {
        body.surname = surname;
    }
    const responseData = await transport_1.microsoftApiRequest.call(this, 'POST', '/contacts', body);
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: index } });
    return executionData;
}
exports.execute = execute;
