// n8n/dist/decorators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectScope = exports.GlobalScope = exports.Licensed = exports.ControllerRegistry = exports.Middleware = exports.Delete = exports.Patch = exports.Put = exports.Post = exports.Get = exports.RestController = void 0;
var RestController_1 = require("./RestController");
Object.defineProperty(exports, "RestController", { enumerable: true, get: function () { return RestController_1.RestController; } });
var Route_1 = require("./Route");
Object.defineProperty(exports, "Get", { enumerable: true, get: function () { return Route_1.Get; } });
Object.defineProperty(exports, "Post", { enumerable: true, get: function () { return Route_1.Post; } });
Object.defineProperty(exports, "Put", { enumerable: true, get: function () { return Route_1.Put; } });
Object.defineProperty(exports, "Patch", { enumerable: true, get: function () { return Route_1.Patch; } });
Object.defineProperty(exports, "Delete", { enumerable: true, get: function () { return Route_1.Delete; } });
var Middleware_1 = require("./Middleware");
Object.defineProperty(exports, "Middleware", { enumerable: true, get: function () { return Middleware_1.Middleware; } });
var controller_registry_1 = require("./controller.registry");
Object.defineProperty(exports, "ControllerRegistry", { enumerable: true, get: function () { return controller_registry_1.ControllerRegistry; } });
var Licensed_1 = require("./Licensed");
Object.defineProperty(exports, "Licensed", { enumerable: true, get: function () { return Licensed_1.Licensed; } });
var Scoped_1 = require("./Scoped");
Object.defineProperty(exports, "GlobalScope", { enumerable: true, get: function () { return Scoped_1.GlobalScope; } });
Object.defineProperty(exports, "ProjectScope", { enumerable: true, get: function () { return Scoped_1.ProjectScope; } });
