// n8n-nodes-base/dist/nodes/Raindrop/descriptions/BookmarkDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bookmarkFields = exports.bookmarkOperations = void 0;
exports.bookmarkOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        default: 'get',
        description: 'Operation to perform',
        options: [
            {
                name: 'Create',
                value: 'create',
            },
            {
                name: 'Delete',
                value: 'delete',
            },
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
            {
                name: 'Update',
                value: 'update',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
            },
        },
    },
];
exports.bookmarkFields = [
    {
        displayName: 'Collection ID',
        name: 'collectionId',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'create',
                ],
            },
        },
        typeOptions: {
            loadOptionsMethod: 'getCollections',
        },
        default: '',
    },
    {
        displayName: 'Link',
        name: 'link',
        type: 'string',
        required: true,
        default: '',
        description: 'Link of the bookmark to be created.',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'create',
                ],
            },
        },
        options: [
            {
                displayName: 'Important',
                name: 'important',
                type: 'boolean',
                default: false,
                description: 'Whether this bookmark is marked as favorite.',
            },
            {
                displayName: 'Order',
                name: 'order',
                type: 'number',
                default: 0,
                description: 'Sort order for the bookmark. For example, to move it to first place, enter 0.',
            },
            {
                displayName: 'Tags',
                name: 'tags',
                type: 'string',
                default: '',
                description: 'Bookmark tags. Multiple can be set separated by comma.',
            },
            {
                displayName: 'Title',
                name: 'title',
                type: 'string',
                default: '',
                description: 'Title of the bookmark to create.',
            },
        ],
    },
    {
        displayName: 'Bookmark ID',
        name: 'bookmarkId',
        type: 'string',
        default: '',
        required: true,
        description: 'The ID of the bookmark to delete.',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'delete',
                ],
            },
        },
    },
    {
        displayName: 'Bookmark ID',
        name: 'bookmarkId',
        type: 'string',
        default: '',
        required: true,
        description: 'The ID of the bookmark to retrieve.',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    {
        displayName: 'Collection ID',
        name: 'collectionId',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getCollections',
        },
        default: [],
        required: true,
        description: 'The ID of the collection from which to retrieve all bookmarks.',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        default: false,
        description: 'If all results should be returned or only up to a given limit.',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 10,
        },
        default: 5,
        description: 'How many results to return.',
    },
    {
        displayName: 'Bookmark ID',
        name: 'bookmarkId',
        type: 'string',
        default: '',
        required: true,
        description: 'The ID of the bookmark to update.',
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'update',
                ],
            },
        },
    },
    {
        displayName: 'Update Fields',
        name: 'updateFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'bookmark',
                ],
                operation: [
                    'update',
                ],
            },
        },
        options: [
            {
                displayName: 'Collection ID',
                name: 'collectionId',
                type: 'options',
                typeOptions: {
                    loadOptionsMethod: 'getCollections',
                },
                default: '',
            },
            {
                displayName: 'Important',
                name: 'important',
                type: 'boolean',
                default: false,
                description: 'Whether this bookmark is marked as favorite.',
            },
            {
                displayName: 'Order',
                name: 'order',
                type: 'number',
                default: 0,
                description: 'For example if you want to move bookmark to the first place set this field to 0',
            },
            {
                displayName: 'Tags',
                name: 'tags',
                type: 'string',
                default: '',
                description: 'Bookmark tags. Multiple can be set separated by comma.',
            },
            {
                displayName: 'Title',
                name: 'title',
                type: 'string',
                default: '',
                description: 'Title of the bookmark to be created.',
            },
        ],
    },
];
