// n8n-nodes-base/dist/nodes/Twitter/V2/UserDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userFields = exports.userOperations = void 0;
exports.userOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['user'],
            },
        },
        options: [
            {
                name: 'Get',
                value: 'searchUser',
                description: 'Retrieve a user by username',
                action: 'Get User',
            },
        ],
        default: 'searchUser',
    },
];
exports.userFields = [
    {
        displayName: 'User',
        name: 'user',
        type: 'resourceLocator',
        default: { mode: 'username', value: '' },
        required: true,
        description: 'The user you want to search',
        displayOptions: {
            show: {
                operation: ['searchUser'],
                resource: ['user'],
            },
            hide: {
                me: [true],
            },
        },
        modes: [
            {
                displayName: 'By Username',
                name: 'username',
                type: 'string',
                validation: [],
                placeholder: 'e.g. n8n',
                url: '',
            },
            {
                displayName: 'By ID',
                name: 'id',
                type: 'string',
                validation: [],
                placeholder: 'e.g. 1068479892537384960',
                url: '',
            },
        ],
    },
    {
        displayName: 'Me',
        name: 'me',
        type: 'boolean',
        displayOptions: {
            show: {
                operation: ['searchUser'],
                resource: ['user'],
            },
        },
        default: false,
        description: 'Whether you want to search the authenticated user',
    },
];
