// n8n/dist/executionLifecycleHooks/toSaveSettings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSaveSettings = void 0;
const config_1 = __importDefault(require("../config"));
function toSaveSettings(workflowSettings = {}) {
    var _a, _b;
    const DEFAULTS = {
        ERROR: config_1.default.getEnv('executions.saveDataOnError'),
        SUCCESS: config_1.default.getEnv('executions.saveDataOnSuccess'),
        MANUAL: config_1.default.getEnv('executions.saveDataManualExecutions'),
        PROGRESS: config_1.default.getEnv('executions.saveExecutionProgress'),
    };
    return {
        error: workflowSettings.saveDataErrorExecution
            ? workflowSettings.saveDataErrorExecution !== 'none'
            : DEFAULTS.ERROR !== 'none',
        success: workflowSettings.saveDataSuccessExecution
            ? workflowSettings.saveDataSuccessExecution !== 'none'
            : DEFAULTS.SUCCESS !== 'none',
        manual: workflowSettings === undefined || workflowSettings.saveManualExecutions === 'DEFAULT'
            ? DEFAULTS.MANUAL
            : (_a = workflowSettings.saveManualExecutions) !== null && _a !== void 0 ? _a : DEFAULTS.MANUAL,
        progress: workflowSettings === undefined || workflowSettings.saveExecutionProgress === 'DEFAULT'
            ? DEFAULTS.PROGRESS
            : (_b = workflowSettings.saveExecutionProgress) !== null && _b !== void 0 ? _b : DEFAULTS.PROGRESS,
    };
}
exports.toSaveSettings = toSaveSettings;
