// @n8n/n8n-nodes-langchain/dist/nodes/embeddings/EmbeddingsOllama/EmbeddingsOllama.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmbeddingsOllama = void 0;
const ollama_1 = require("@langchain/community/embeddings/ollama");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
const description_1 = require("../../llms/LMOllama/description");
class EmbeddingsOllama {
    constructor() {
        this.description = {
            displayName: 'Embeddings Ollama',
            name: 'embeddingsOllama',
            icon: 'file:ollama.svg',
            group: ['transform'],
            version: 1,
            description: 'Use Ollama Embeddings',
            defaults: {
                name: 'Embeddings Ollama',
            },
            ...description_1.ollamaDescription,
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Embeddings'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.embeddingsollama/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_embedding"],
            outputNames: ['Embeddings'],
            properties: [(0, sharedFields_1.getConnectionHintNoticeField)(["ai_vectorStore"]), description_1.ollamaModel],
        };
    }
    async supplyData(itemIndex) {
        this.logger.verbose('Supply data for embeddings Ollama');
        const modelName = this.getNodeParameter('model', itemIndex);
        const credentials = await this.getCredentials('ollamaApi');
        const embeddings = new ollama_1.OllamaEmbeddings({
            baseUrl: credentials.baseUrl,
            model: modelName,
        });
        return {
            response: (0, logWrapper_1.logWrapper)(embeddings, this),
        };
    }
}
exports.EmbeddingsOllama = EmbeddingsOllama;
