// n8n-nodes-base/dist/nodes/TheHiveProject/actions/case/deleteAttachment.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utilities_1 = require("../../../../utils/utilities");
const properties = [
    descriptions_1.caseRLC,
    {
        displayName: 'Attachment Name or ID',
        name: 'attachmentId',
        type: 'options',
        default: '',
        required: true,
        description: 'ID of the attachment. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
        typeOptions: {
            loadOptionsMethod: 'loadCaseAttachments',
        },
    },
];
const displayOptions = {
    show: {
        resource: ['case'],
        operation: ['deleteAttachment'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    const caseId = this.getNodeParameter('caseId', i, '', { extractValue: true });
    const attachmentId = this.getNodeParameter('attachmentId', i);
    await transport_1.theHiveApiRequest.call(this, 'DELETE', `/v1/case/${caseId}/attachment/${attachmentId}`);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)({ success: true }), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
