// n8n-nodes-base/dist/nodes/TheHiveProject/actions/case/get.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utilities_1 = require("../../../../utils/utilities");
const properties = [descriptions_1.caseRLC];
const displayOptions = {
    show: {
        resource: ['case'],
        operation: ['get'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const caseId = this.getNodeParameter('caseId', i, '', { extractValue: true });
    const qs = {};
    const body = {
        query: [
            {
                _name: 'getCase',
                idOrName: caseId,
            },
            {
                _name: 'page',
                from: 0,
                to: 10,
                extraData: ['attachmentCount'],
            },
        ],
    };
    qs.name = `get-case-${caseId}`;
    responseData = await transport_1.theHiveApiRequest.call(this, 'POST', '/v1/query', body, qs);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
