// n8n-nodes-base/dist/nodes/TheHiveProject/actions/case/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.update = exports.getTimeline = exports.getAttachment = exports.search = exports.get = exports.executeResponder = exports.deleteCase = exports.deleteAttachment = exports.create = exports.addAttachment = void 0;
const addAttachment = __importStar(require("./addAttachment.operation"));
exports.addAttachment = addAttachment;
const create = __importStar(require("./create.operation"));
exports.create = create;
const deleteAttachment = __importStar(require("./deleteAttachment.operation"));
exports.deleteAttachment = deleteAttachment;
const deleteCase = __importStar(require("./deleteCase.operation"));
exports.deleteCase = deleteCase;
const executeResponder = __importStar(require("./executeResponder.operation"));
exports.executeResponder = executeResponder;
const get = __importStar(require("./get.operation"));
exports.get = get;
const getAttachment = __importStar(require("./getAttachment.operation"));
exports.getAttachment = getAttachment;
const search = __importStar(require("./search.operation"));
exports.search = search;
const getTimeline = __importStar(require("./getTimeline.operation"));
exports.getTimeline = getTimeline;
const update = __importStar(require("./update.operation"));
exports.update = update;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        default: 'create',
        type: 'options',
        noDataExpression: true,
        required: true,
        options: [
            {
                name: 'Add Attachment',
                value: 'addAttachment',
                action: 'Add attachment to a case',
            },
            {
                name: 'Create',
                value: 'create',
                action: 'Create a case',
            },
            {
                name: 'Delete Attachment',
                value: 'deleteAttachment',
                action: 'Delete attachment from a case',
            },
            {
                name: 'Delete Case',
                value: 'deleteCase',
                action: 'Delete an case',
            },
            {
                name: 'Execute Responder',
                value: 'executeResponder',
                action: 'Execute responder on a case',
            },
            {
                name: 'Get',
                value: 'get',
                action: 'Get a case',
            },
            {
                name: 'Get Attachment',
                value: 'getAttachment',
                action: 'Get attachment from a case',
            },
            {
                name: 'Get Timeline',
                value: 'getTimeline',
                action: 'Get timeline of a case',
            },
            {
                name: 'Search',
                value: 'search',
                action: 'Search cases',
            },
            {
                name: 'Update',
                value: 'update',
                action: 'Update a case',
            },
        ],
        displayOptions: {
            show: {
                resource: ['case'],
            },
        },
    },
    ...addAttachment.description,
    ...create.description,
    ...deleteAttachment.description,
    ...deleteCase.description,
    ...executeResponder.description,
    ...get.description,
    ...getAttachment.description,
    ...search.description,
    ...getTimeline.description,
    ...update.description,
];
