// n8n-nodes-base/dist/nodes/CrowdDev/descriptions/noteFields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noteFields = exports.noteOperations = void 0;
const GenericFunctions_1 = require("../GenericFunctions");
const utils_1 = require("./utils");
const displayOpts = (0, utils_1.showFor)(['note']);
const displayFor = {
    resource: displayOpts(),
    createOrUpdate: displayOpts(['create', 'update']),
    id: displayOpts(['delete', 'find', 'update']),
};
const noteOperations = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    displayOptions: displayFor.resource.displayOptions,
    noDataExpression: true,
    default: 'find',
    options: [
        {
            name: 'Create',
            value: 'create',
            description: 'Create a note',
            action: 'Create a note',
            routing: {
                send: { preSend: [GenericFunctions_1.notePresend] },
                request: {
                    method: 'POST',
                    url: '/note',
                },
            },
        },
        {
            name: 'Delete',
            value: 'delete',
            description: 'Delete a note',
            action: 'Delete a note',
            routing: {
                request: {
                    method: 'DELETE',
                    url: '=/note',
                },
            },
        },
        {
            name: 'Find',
            value: 'find',
            description: 'Find a note',
            action: 'Find a note',
            routing: {
                request: {
                    method: 'GET',
                    url: '=/note/{{$parameter["id"]}}',
                },
            },
        },
        {
            name: 'Update',
            value: 'update',
            description: 'Update a note',
            action: 'Update a note',
            routing: {
                send: { preSend: [GenericFunctions_1.notePresend] },
                request: {
                    method: 'PUT',
                    url: '=/note/{{$parameter["id"]}}',
                },
            },
        },
    ],
};
exports.noteOperations = noteOperations;
const commonFields = [
    {
        displayName: 'Body',
        name: 'body',
        description: 'The body of the note',
        type: 'string',
        typeOptions: {
            rows: 4,
        },
        default: '',
    },
];
const noteFields = [
    Object.assign((0, utils_1.getId)(), { description: 'The ID of the note' }, displayFor.id),
    ...commonFields.map((0, utils_1.mapWith)(displayFor.createOrUpdate)),
];
exports.noteFields = noteFields;
