// n8n-nodes-base/dist/nodes/CrowdDev/descriptions/organizationFields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.organizationFields = exports.organizationOperations = void 0;
const GenericFunctions_1 = require("../GenericFunctions");
const utils_1 = require("./utils");
const displayOpts = (0, utils_1.showFor)(['organization']);
const displayFor = {
    resource: displayOpts(),
    createOrUpdate: displayOpts(['create', 'update']),
    id: displayOpts(['delete', 'find', 'update']),
};
const organizationOperations = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    displayOptions: displayFor.resource.displayOptions,
    noDataExpression: true,
    default: 'find',
    options: [
        {
            name: 'Create',
            value: 'create',
            description: 'Create an organization',
            action: 'Create an organization',
            routing: {
                send: { preSend: [GenericFunctions_1.organizationPresend] },
                request: {
                    method: 'POST',
                    url: '/organization',
                },
            },
        },
        {
            name: 'Delete',
            value: 'delete',
            description: 'Delete an organization',
            action: 'Delete an organization',
            routing: {
                request: {
                    method: 'DELETE',
                    url: '=/organization',
                },
            },
        },
        {
            name: 'Find',
            value: 'find',
            description: 'Find an organization',
            action: 'Find an organization',
            routing: {
                request: {
                    method: 'GET',
                    url: '=/organization/{{$parameter["id"]}}',
                },
            },
        },
        {
            name: 'Update',
            value: 'update',
            description: 'Update an organization',
            action: 'Update an organization',
            routing: {
                send: { preSend: [GenericFunctions_1.organizationPresend] },
                request: {
                    method: 'PUT',
                    url: '=/organization/{{$parameter["id"]}}',
                },
            },
        },
    ],
};
exports.organizationOperations = organizationOperations;
const commonFields = [
    {
        displayName: 'Name',
        name: 'name',
        description: 'The name of the organization',
        type: 'string',
        required: true,
        default: '',
    },
];
const additionalOptions = [
    {
        displayName: 'Url',
        name: 'url',
        description: 'The URL of the organization',
        type: 'string',
        default: '',
    },
    {
        displayName: 'Description',
        name: 'description',
        description: 'A short description of the organization',
        type: 'string',
        typeOptions: {
            rows: 3,
        },
        default: '',
    },
    {
        displayName: 'Logo',
        name: 'logo',
        description: 'A URL for logo of the organization',
        type: 'string',
        default: '',
    },
    {
        displayName: 'Employees',
        name: 'employees',
        description: 'The number of employees of the organization',
        type: 'number',
        default: '',
    },
    {
        displayName: 'Members',
        name: 'members',
        description: 'Members associated with the organization. Each element in the array is the ID of the member.',
        type: 'fixedCollection',
        typeOptions: {
            multipleValues: true,
        },
        default: {},
        options: [
            {
                displayName: 'Item Choice',
                name: 'itemChoice',
                values: [
                    {
                        displayName: 'Member',
                        name: 'member',
                        type: 'string',
                        default: '',
                    },
                ],
            },
        ],
    },
];
const organizationFields = [
    Object.assign((0, utils_1.getId)(), { description: 'The ID of the organization' }, displayFor.id),
    ...commonFields.map((0, utils_1.mapWith)(displayFor.createOrUpdate)),
    Object.assign({}, (0, utils_1.getAdditionalOptions)(additionalOptions), displayFor.createOrUpdate),
];
exports.organizationFields = organizationFields;
