// n8n-nodes-base/dist/nodes/CrowdDev/descriptions/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAdditionalOptions = exports.getId = exports.mapWith = exports.showFor = void 0;
const showFor = (resources) => (operations) => {
    return operations !== undefined
        ? {
            displayOptions: {
                show: {
                    resource: resources,
                    operation: operations,
                },
            },
        }
        : {
            displayOptions: {
                show: {
                    resource: resources,
                },
            },
        };
};
exports.showFor = showFor;
const mapWith = (...objects) => (item) => Object.assign({}, item, ...objects);
exports.mapWith = mapWith;
const getId = () => ({
    displayName: 'ID',
    name: 'id',
    type: 'string',
    required: true,
    default: '',
    routing: {
        send: {
            type: 'query',
            property: 'ids[]',
        },
    },
});
exports.getId = getId;
const getAdditionalOptions = (fields) => {
    return {
        displayName: 'Additional Options',
        name: 'additionalOptions',
        type: 'collection',
        displayOptions: {
            show: {
                operation: ['getAll'],
            },
        },
        default: {},
        placeholder: 'Add Option',
        options: fields,
    };
};
exports.getAdditionalOptions = getAdditionalOptions;
