// n8n/dist/commands/user-management/reset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reset = void 0;
const typedi_1 = require("typedi");
const User_1 = require("../../databases/entities/User");
const credentials_repository_1 = require("../../databases/repositories/credentials.repository");
const settings_repository_1 = require("../../databases/repositories/settings.repository");
const sharedCredentials_repository_1 = require("../../databases/repositories/sharedCredentials.repository");
const sharedWorkflow_repository_1 = require("../../databases/repositories/sharedWorkflow.repository");
const user_repository_1 = require("../../databases/repositories/user.repository");
const BaseCommand_1 = require("../BaseCommand");
const project_repository_1 = require("../../databases/repositories/project.repository");
const defaultUserProps = {
    firstName: null,
    lastName: null,
    email: null,
    password: null,
    role: 'global:owner',
};
class Reset extends BaseCommand_1.BaseCommand {
    async run() {
        const owner = await this.getInstanceOwner();
        const personalProject = await typedi_1.Container.get(project_repository_1.ProjectRepository).getPersonalProjectForUserOrFail(owner.id);
        await typedi_1.Container.get(sharedWorkflow_repository_1.SharedWorkflowRepository).makeOwnerOfAllWorkflows(personalProject);
        await typedi_1.Container.get(sharedCredentials_repository_1.SharedCredentialsRepository).makeOwnerOfAllCredentials(personalProject);
        await typedi_1.Container.get(user_repository_1.UserRepository).deleteAllExcept(owner);
        await typedi_1.Container.get(user_repository_1.UserRepository).save(Object.assign(owner, defaultUserProps));
        const danglingCredentials = await typedi_1.Container.get(credentials_repository_1.CredentialsRepository)
            .createQueryBuilder('credentials')
            .leftJoinAndSelect('credentials.shared', 'shared')
            .where('shared.credentialsId is null')
            .getMany();
        const newSharedCredentials = danglingCredentials.map((credentials) => typedi_1.Container.get(sharedCredentials_repository_1.SharedCredentialsRepository).create({
            credentials,
            projectId: personalProject.id,
            role: 'credential:owner',
        }));
        await typedi_1.Container.get(sharedCredentials_repository_1.SharedCredentialsRepository).save(newSharedCredentials);
        await typedi_1.Container.get(settings_repository_1.SettingsRepository).update({ key: 'userManagement.isInstanceOwnerSetUp' }, { value: 'false' });
        this.logger.info('Successfully reset the database to default user state.');
    }
    async getInstanceOwner() {
        const owner = await typedi_1.Container.get(user_repository_1.UserRepository).findOneBy({ role: 'global:owner' });
        if (owner)
            return owner;
        const user = new User_1.User();
        Object.assign(user, defaultUserProps);
        await typedi_1.Container.get(user_repository_1.UserRepository).save(user);
        return await typedi_1.Container.get(user_repository_1.UserRepository).findOneByOrFail({ role: 'global:owner' });
    }
    async catch(error) {
        this.logger.error('Error resetting database. See log messages for details.');
        this.logger.error(error.message);
        this.exit(1);
    }
}
exports.Reset = Reset;
Reset.description = 'Resets the database to the default user state';
Reset.examples = ['$ n8n user-management:reset'];
