// @n8n/n8n-nodes-langchain/dist/nodes/vendors/OpenAi/actions/file/list.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const transport_1 = require("../../transport");
const properties = [
    {
        displayName: 'Options',
        name: 'options',
        placeholder: 'Add Option',
        type: 'collection',
        default: {},
        options: [
            {
                displayName: 'Purpose',
                name: 'purpose',
                type: 'options',
                default: 'any',
                description: 'Only return files with the given purpose',
                options: [
                    {
                        name: 'Any [Default]',
                        value: 'any',
                    },
                    {
                        name: 'Assistants',
                        value: 'assistants',
                    },
                    {
                        name: 'Fine-Tune',
                        value: 'fine-tune',
                    },
                ],
            },
        ],
    },
];
const displayOptions = {
    show: {
        operation: ['list'],
        resource: ['file'],
    },
};
exports.description = (0, n8n_workflow_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    const options = this.getNodeParameter('options', i, {});
    const qs = {};
    if (options.purpose && options.purpose !== 'any') {
        qs.purpose = options.purpose;
    }
    const { data } = await transport_1.apiRequest.call(this, 'GET', '/files', { qs });
    return (data || []).map((file) => ({
        json: file,
        pairedItem: { item: i },
    }));
}
exports.execute = execute;
