// n8n-nodes-base/dist/nodes/SplitInBatches/SplitInBatches.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplitInBatches = void 0;
class SplitInBatches {
    constructor() {
        this.description = {
            displayName: 'Split In Batches',
            name: 'splitInBatches',
            icon: 'fa:th-large',
            group: ['organization'],
            version: 1,
            description: 'Split data into batches and iterate over each batch',
            defaults: {
                name: 'SplitInBatches',
                color: '#007755',
            },
            inputs: ['main'],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'You may not need this node — n8n nodes automatically run once for each input item. <a href="https://docs.n8n.io/getting-started/key-concepts/looping.html#using-loops-in-n8n" target="_blank">More info</a>',
                    name: 'splitInBatchesNotice',
                    type: 'notice',
                    default: '',
                },
                {
                    displayName: 'Batch Size',
                    name: 'batchSize',
                    type: 'number',
                    typeOptions: {
                        minValue: 1,
                    },
                    default: 10,
                    description: 'The number of items to return with each call.',
                },
                {
                    displayName: 'Options',
                    name: 'options',
                    type: 'collection',
                    placeholder: 'Add Option',
                    default: {},
                    options: [
                        {
                            displayName: 'Reset',
                            name: 'reset',
                            type: 'boolean',
                            default: false,
                            description: 'If set to true, the node will be reset and so with the current input-data newly initialized.',
                        },
                    ],
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData().slice();
        const nodeContext = this.getContext('node');
        const batchSize = this.getNodeParameter('batchSize', 0);
        const returnItems = [];
        const options = this.getNodeParameter('options', 0, {});
        if (nodeContext.items === undefined || options.reset === true) {
            nodeContext.currentRunIndex = 0;
            nodeContext.maxRunIndex = Math.ceil(items.length / batchSize);
            returnItems.push.apply(returnItems, items.splice(0, batchSize));
            nodeContext.items = items;
        }
        else {
            nodeContext.currentRunIndex += 1;
            returnItems.push.apply(returnItems, nodeContext.items.splice(0, batchSize));
        }
        nodeContext.noItemsLeft = nodeContext.items.length === 0;
        if (returnItems.length === 0) {
            return null;
        }
        return this.prepareOutputData(returnItems);
    }
}
exports.SplitInBatches = SplitInBatches;
