// n8n-nodes-base/dist/nodes/CrowdDev/CrowdDev.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrowdDev = void 0;
const descriptions_1 = require("./descriptions");
class CrowdDev {
    constructor() {
        this.description = {
            displayName: 'crowd.dev',
            name: 'crowdDev',
            icon: { light: 'file:crowdDev.svg', dark: 'file:crowdDev.dark.svg' },
            group: ['transform'],
            version: 1,
            subtitle: '={{ $parameter["operation"] + ": " + $parameter["resource"] }}',
            description: 'crowd.dev is an open-source suite of community and data tools built to unlock community-led growth for your organization.',
            defaults: {
                name: 'crowd.dev',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'crowdDevApi',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: '={{$credentials.url}}/api/tenant/{{$credentials.tenantId}}',
                json: true,
                skipSslCertificateValidation: '={{ $credentials.allowUnauthorizedCerts }}',
            },
            properties: descriptions_1.allProperties,
        };
    }
}
exports.CrowdDev = CrowdDev;
