// n8n-nodes-base/dist/nodes/LinkedIn/LinkedIn.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkedIn = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
const PostDescription_1 = require("./PostDescription");
class LinkedIn {
    constructor() {
        this.description = {
            displayName: 'LinkedIn',
            name: 'linkedIn',
            icon: 'file:linkedin.svg',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume LinkedIn API',
            defaults: {
                name: 'LinkedIn',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'linkedInOAuth2Api',
                    required: true,
                    displayOptions: {
                        show: {
                            authentication: ['standard'],
                        },
                    },
                },
                {
                    name: 'linkedInCommunityManagementOAuth2Api',
                    required: true,
                    displayOptions: {
                        show: {
                            authentication: ['communityManagement'],
                        },
                    },
                },
            ],
            properties: [
                {
                    displayName: 'Authentication',
                    name: 'authentication',
                    type: 'options',
                    options: [
                        {
                            name: 'Standard',
                            value: 'standard',
                        },
                        {
                            name: 'Community Management',
                            value: 'communityManagement',
                        },
                    ],
                    default: 'standard',
                },
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Post',
                            value: 'post',
                        },
                    ],
                    default: 'post',
                },
                ...PostDescription_1.postOperations,
                ...PostDescription_1.postFields,
            ],
        };
        this.methods = {
            loadOptions: {
                async getPersonUrn() {
                    var _a, _b, _c;
                    const authentication = this.getNodeParameter('authentication', 0);
                    let endpoint = '/me';
                    if (authentication === 'standard') {
                        const { legacy } = await this.getCredentials('linkedInOAuth2Api');
                        if (!legacy) {
                            endpoint = '/v2/userinfo';
                        }
                    }
                    const person = await GenericFunctions_1.linkedInApiRequest.call(this, 'GET', endpoint, {});
                    const firstName = (_a = person.localizedFirstName) !== null && _a !== void 0 ? _a : person.given_name;
                    const lastName = (_b = person.localizedLastName) !== null && _b !== void 0 ? _b : person.family_name;
                    const name = `${firstName} ${lastName}`;
                    const returnData = [
                        {
                            name,
                            value: (_c = person.id) !== null && _c !== void 0 ? _c : person.sub,
                        },
                    ];
                    return returnData;
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        let body = {};
        for (let i = 0; i < items.length; i++) {
            try {
                if (resource === 'post') {
                    if (operation === 'create') {
                        const text = this.getNodeParameter('text', i);
                        const shareMediaCategory = this.getNodeParameter('shareMediaCategory', i);
                        const postAs = this.getNodeParameter('postAs', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        let authorUrn = '';
                        let visibility = 'PUBLIC';
                        if (postAs === 'person') {
                            const personUrn = this.getNodeParameter('person', i);
                            visibility = additionalFields.visibility || 'PUBLIC';
                            authorUrn = `urn:li:person:${personUrn}`;
                        }
                        else {
                            const organizationUrn = this.getNodeParameter('organization', i);
                            authorUrn = `urn:li:organization:${organizationUrn}`;
                        }
                        let description = '';
                        let title = '';
                        let originalUrl = '';
                        body = {
                            author: authorUrn,
                            lifecycleState: 'PUBLISHED',
                            distribution: {
                                feedDistribution: 'MAIN_FEED',
                                thirdPartyDistributionChannels: [],
                            },
                            visibility,
                        };
                        if (shareMediaCategory === 'IMAGE') {
                            if (additionalFields.title) {
                                title = additionalFields.title;
                            }
                            const registerRequest = {
                                initializeUploadRequest: {
                                    owner: authorUrn,
                                },
                            };
                            const registerObject = await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', '/images?action=initializeUpload', registerRequest);
                            const binaryPropertyName = this.getNodeParameter('binaryPropertyName', i);
                            const imageMetadata = this.helpers.assertBinaryData(i, binaryPropertyName);
                            const buffer = await this.helpers.getBinaryDataBuffer(i, binaryPropertyName);
                            const { uploadUrl, image } = registerObject.value;
                            const headers = {};
                            Object.assign(headers, { 'Content-Type': imageMetadata.mimeType });
                            await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', uploadUrl, buffer, true, headers);
                            const imageBody = {
                                content: {
                                    media: {
                                        title,
                                        id: image,
                                    },
                                },
                                commentary: text,
                            };
                            Object.assign(body, imageBody);
                        }
                        else if (shareMediaCategory === 'ARTICLE') {
                            if (additionalFields.description) {
                                description = additionalFields.description;
                            }
                            if (additionalFields.title) {
                                title = additionalFields.title;
                            }
                            if (additionalFields.originalUrl) {
                                originalUrl = additionalFields.originalUrl;
                            }
                            const articleBody = {
                                content: {
                                    article: {
                                        title,
                                        description,
                                        source: originalUrl,
                                    },
                                },
                                commentary: text,
                            };
                            if (additionalFields.thumbnailBinaryPropertyName) {
                                const registerRequest = {
                                    initializeUploadRequest: {
                                        owner: authorUrn,
                                    },
                                };
                                const registerObject = await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', '/images?action=initializeUpload', registerRequest);
                                const binaryPropertyName = additionalFields.thumbnailBinaryPropertyName;
                                const imageMetadata = this.helpers.assertBinaryData(i, binaryPropertyName);
                                const buffer = await this.helpers.getBinaryDataBuffer(i, binaryPropertyName);
                                const { uploadUrl, image } = registerObject.value;
                                const headers = {};
                                Object.assign(headers, { 'Content-Type': imageMetadata.mimeType });
                                await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', uploadUrl, buffer, true, headers);
                                Object.assign(articleBody.content.article, { thumbnail: image });
                            }
                            Object.assign(body, articleBody);
                            if (description === '') {
                                delete body.description;
                            }
                            if (title === '') {
                                delete body.title;
                            }
                        }
                        else {
                            Object.assign(body, { commentary: text });
                        }
                        const endpoint = '/posts';
                        responseData = await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', endpoint, body);
                    }
                }
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            catch (error) {
                if (this.continueOnFail(error)) {
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                    returnData.push(...executionData);
                    continue;
                }
                throw error;
            }
        }
        return [returnData];
    }
}
exports.LinkedIn = LinkedIn;
