// n8n-nodes-base/dist/nodes/Google/Firebase/CloudFirestore/CollectionDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectionFields = exports.collectionOperations = void 0;
exports.collectionOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
            },
        },
        options: [
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Get all root collections',
            },
        ],
        default: 'getAll',
        description: 'The operation to perform.',
    },
];
exports.collectionFields = [
    {
        displayName: 'Project ID',
        name: 'projectId',
        type: 'options',
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getProjects',
        },
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        description: 'As displayed in firebase console URL',
        required: true,
    },
    {
        displayName: 'Database',
        name: 'database',
        type: 'string',
        default: '(default)',
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        description: 'Usually the provided default value will work',
        required: true,
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        description: 'If all results should be returned or only up to a given limit.',
        required: true,
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                resource: [
                    'collection',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 500,
        },
        default: 100,
        description: 'How many results to return.',
    },
];
