// n8n-nodes-base/dist/nodes/Airtable/v2/actions/record/Record.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.upsert = exports.update = exports.search = exports.get = exports.deleteRecord = exports.create = void 0;
const common_descriptions_1 = require("../common.descriptions");
const create = __importStar(require("./create.operation"));
exports.create = create;
const deleteRecord = __importStar(require("./deleteRecord.operation"));
exports.deleteRecord = deleteRecord;
const get = __importStar(require("./get.operation"));
exports.get = get;
const search = __importStar(require("./search.operation"));
exports.search = search;
const update = __importStar(require("./update.operation"));
exports.update = update;
const upsert = __importStar(require("./upsert.operation"));
exports.upsert = upsert;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create a new record in a table',
                action: 'Create a record',
            },
            {
                name: 'Create or Update',
                value: 'upsert',
                description: 'Create a new record, or update the current one if it already exists (upsert)',
                action: 'Create or update a record',
            },
            {
                name: 'Delete',
                value: 'deleteRecord',
                description: 'Delete a record from a table',
                action: 'Delete a record',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Retrieve a record from a table',
                action: 'Get a record',
            },
            {
                name: 'Search',
                value: 'search',
                description: 'Search for specific records or list all',
                action: 'Search records',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update a record in a table',
                action: 'Update record',
            },
        ],
        default: 'read',
        displayOptions: {
            show: {
                resource: ['record'],
            },
        },
    },
    {
        ...common_descriptions_1.baseRLC,
        displayOptions: {
            show: {
                resource: ['record'],
            },
        },
    },
    {
        ...common_descriptions_1.tableRLC,
        displayOptions: {
            show: {
                resource: ['record'],
            },
        },
    },
    ...create.description,
    ...deleteRecord.description,
    ...get.description,
    ...search.description,
    ...update.description,
    ...upsert.description,
];
