// n8n-workflow/dist/Extensions/BooleanExtensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.booleanExtensions = exports.toDateTime = exports.toInt = exports.toBoolean = void 0;
function toBoolean(value) {
    return value;
}
exports.toBoolean = toBoolean;
function toInt(value) {
    return value ? 1 : 0;
}
exports.toInt = toInt;
function toDateTime() {
    return undefined;
}
exports.toDateTime = toDateTime;
const toFloat = toInt;
const toNumber = toInt.bind({});
toNumber.doc = {
    name: 'toNumber',
    description: 'Converts <code>true</code> to <code>1</code> and <code>false</code> to <code>0</code>.',
    examples: [
        { example: 'true.toNumber()', evaluated: '1' },
        { example: 'false.toNumber()', evaluated: '0' },
    ],
    section: 'cast',
    returnType: 'number',
    docURL: 'https://docs.n8n.io/code/builtin/data-transformation-functions/booleans/#boolean-toNumber',
};
exports.booleanExtensions = {
    typeName: 'Boolean',
    functions: {
        toBoolean,
        toInt,
        toFloat,
        toNumber,
        toDateTime,
    },
};
