// n8n-nodes-base/dist/nodes/Strava/Strava.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Strava = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
const ActivityDescription_1 = require("./ActivityDescription");
const moment = require("moment");
class Strava {
    constructor() {
        this.description = {
            displayName: 'Strava',
            name: 'strava',
            icon: 'file:strava.svg',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Strava API',
            defaults: {
                name: 'Strava',
                color: '#ea5929',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'stravaOAuth2Api',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    options: [
                        {
                            name: 'Activity',
                            value: 'activity',
                        },
                    ],
                    default: 'activity',
                    description: 'The resource to operate on.',
                },
                ...ActivityDescription_1.activityOperations,
                ...ActivityDescription_1.activityFields,
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        const qs = {};
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        for (let i = 0; i < length; i++) {
            try {
                if (resource === 'activity') {
                    if (operation === 'create') {
                        const name = this.getNodeParameter('name', i);
                        const type = this.getNodeParameter('type', i);
                        const startDate = this.getNodeParameter('startDate', i);
                        const elapsedTime = this.getNodeParameter('elapsedTime', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        if (additionalFields.trainer === true) {
                            additionalFields.trainer = 1;
                        }
                        if (additionalFields.commute === true) {
                            additionalFields.commute = 1;
                        }
                        const body = {
                            name,
                            type,
                            start_date_local: moment(startDate).toISOString(),
                            elapsed_time: elapsedTime,
                        };
                        Object.assign(body, additionalFields);
                        responseData = await GenericFunctions_1.stravaApiRequest.call(this, 'POST', '/activities', body);
                    }
                    if (operation === 'get') {
                        const activityId = this.getNodeParameter('activityId', i);
                        responseData = await GenericFunctions_1.stravaApiRequest.call(this, 'GET', `/activities/${activityId}`);
                    }
                    if (['getLaps', 'getZones', 'getKudos', 'getComments'].includes(operation)) {
                        const path = {
                            'getComments': 'comments',
                            'getZones': 'zones',
                            'getKudos': 'kudos',
                            'getLaps': 'laps',
                        };
                        const activityId = this.getNodeParameter('activityId', i);
                        const returnAll = this.getNodeParameter('returnAll', i);
                        responseData = await GenericFunctions_1.stravaApiRequest.call(this, 'GET', `/activities/${activityId}/${path[operation]}`);
                        if (returnAll === false) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.splice(0, limit);
                        }
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        if (returnAll) {
                            responseData = await GenericFunctions_1.stravaApiRequestAllItems.call(this, 'GET', `/activities`, {}, qs);
                        }
                        else {
                            qs.per_page = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.stravaApiRequest.call(this, 'GET', `/activities`, {}, qs);
                        }
                    }
                    if (operation === 'update') {
                        const activityId = this.getNodeParameter('activityId', i);
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if (updateFields.trainer === true) {
                            updateFields.trainer = 1;
                        }
                        if (updateFields.commute === true) {
                            updateFields.commute = 1;
                        }
                        const body = {};
                        Object.assign(body, updateFields);
                        responseData = await GenericFunctions_1.stravaApiRequest.call(this, 'PUT', `/activities/${activityId}`, body);
                    }
                }
                if (Array.isArray(responseData)) {
                    returnData.push.apply(returnData, responseData);
                }
                else if (responseData !== undefined) {
                    returnData.push(responseData);
                }
            }
            catch (error) {
                if (this.continueOnFail()) {
                    returnData.push({ error: error.message });
                    continue;
                }
                throw error;
            }
        }
        return [this.helpers.returnJsonArray(returnData)];
    }
}
exports.Strava = Strava;
