// n8n-workflow/dist/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OBFUSCATED_ERROR_MESSAGE = exports.SINGLE_EXECUTION_NODES = exports.LANGCHAIN_CUSTOM_TOOLS = exports.HTTP_REQUEST_TOOL_LANGCHAIN_NODE_TYPE = exports.WORKFLOW_TOOL_LANGCHAIN_NODE_TYPE = exports.CODE_TOOL_LANGCHAIN_NODE_TYPE = exports.CHAIN_SUMMARIZATION_LANGCHAIN_NODE_TYPE = exports.OPENAI_LANGCHAIN_NODE_TYPE = exports.CHAIN_LLM_LANGCHAIN_NODE_TYPE = exports.AGENT_LANGCHAIN_NODE_TYPE = exports.MANUAL_CHAT_TRIGGER_LANGCHAIN_NODE_TYPE = exports.NODES_WITH_RENAMABLE_CONTENT = exports.SCRIPTING_NODE_TYPES = exports.STARTING_NODE_TYPES = exports.FUNCTION_ITEM_NODE_TYPE = exports.FUNCTION_NODE_TYPE = exports.CODE_NODE_TYPE = exports.EXECUTE_WORKFLOW_TRIGGER_NODE_TYPE = exports.START_NODE_TYPE = exports.ERROR_TRIGGER_NODE_TYPE = exports.MANUAL_TRIGGER_NODE_TYPE = exports.WEBHOOK_NODE_TYPE = exports.HTTP_REQUEST_NODE_TYPE = exports.NO_OP_NODE_TYPE = exports.STICKY_NODE_TYPE = exports.UNKNOWN_ERROR_MESSAGE_CRED = exports.UNKNOWN_ERROR_DESCRIPTION = exports.UNKNOWN_ERROR_MESSAGE = exports.FORM_TRIGGER_PATH_IDENTIFIER = exports.CREDENTIAL_EMPTY_VALUE = exports.CODE_EXECUTION_MODES = exports.CODE_LANGUAGES = exports.LOG_LEVELS = exports.WAIT_TIME_UNLIMITED = exports.BINARY_ENCODING = exports.ALPHABET = exports.LOWERCASE_LETTERS = exports.UPPERCASE_LETTERS = exports.DIGITS = void 0;
exports.DIGITS = '0123456789';
exports.UPPERCASE_LETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
exports.LOWERCASE_LETTERS = exports.UPPERCASE_LETTERS.toLowerCase();
exports.ALPHABET = [exports.DIGITS, exports.UPPERCASE_LETTERS, exports.LOWERCASE_LETTERS].join('');
exports.BINARY_ENCODING = 'base64';
exports.WAIT_TIME_UNLIMITED = '3000-01-01T00:00:00.000Z';
exports.LOG_LEVELS = ['silent', 'error', 'warn', 'info', 'debug', 'verbose'];
exports.CODE_LANGUAGES = ['javaScript', 'python'];
exports.CODE_EXECUTION_MODES = ['runOnceForAllItems', 'runOnceForEachItem'];
exports.CREDENTIAL_EMPTY_VALUE = '__n8n_EMPTY_VALUE_7b1af746-3729-4c60-9b9b-e08eb29e58da';
exports.FORM_TRIGGER_PATH_IDENTIFIER = 'n8n-form';
exports.UNKNOWN_ERROR_MESSAGE = 'There was an unknown issue while executing the node';
exports.UNKNOWN_ERROR_DESCRIPTION = 'Double-check the node configuration and the service it connects to. Check the error details below and refer to the <a href="https://docs.n8n.io" target="_blank">n8n documentation</a> to troubleshoot the issue.';
exports.UNKNOWN_ERROR_MESSAGE_CRED = 'UNKNOWN ERROR';
exports.STICKY_NODE_TYPE = 'n8n-nodes-base.stickyNote';
exports.NO_OP_NODE_TYPE = 'n8n-nodes-base.noOp';
exports.HTTP_REQUEST_NODE_TYPE = 'n8n-nodes-base.httpRequest';
exports.WEBHOOK_NODE_TYPE = 'n8n-nodes-base.webhook';
exports.MANUAL_TRIGGER_NODE_TYPE = 'n8n-nodes-base.manualTrigger';
exports.ERROR_TRIGGER_NODE_TYPE = 'n8n-nodes-base.errorTrigger';
exports.START_NODE_TYPE = 'n8n-nodes-base.start';
exports.EXECUTE_WORKFLOW_TRIGGER_NODE_TYPE = 'n8n-nodes-base.executeWorkflowTrigger';
exports.CODE_NODE_TYPE = 'n8n-nodes-base.code';
exports.FUNCTION_NODE_TYPE = 'n8n-nodes-base.function';
exports.FUNCTION_ITEM_NODE_TYPE = 'n8n-nodes-base.functionItem';
exports.STARTING_NODE_TYPES = [
    exports.MANUAL_TRIGGER_NODE_TYPE,
    exports.EXECUTE_WORKFLOW_TRIGGER_NODE_TYPE,
    exports.ERROR_TRIGGER_NODE_TYPE,
    exports.START_NODE_TYPE,
];
exports.SCRIPTING_NODE_TYPES = [exports.FUNCTION_NODE_TYPE, exports.FUNCTION_ITEM_NODE_TYPE, exports.CODE_NODE_TYPE];
exports.NODES_WITH_RENAMABLE_CONTENT = new Set([
    exports.CODE_NODE_TYPE,
    exports.FUNCTION_NODE_TYPE,
    exports.FUNCTION_ITEM_NODE_TYPE,
]);
exports.MANUAL_CHAT_TRIGGER_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.manualChatTrigger';
exports.AGENT_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.agent';
exports.CHAIN_LLM_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.chainLlm';
exports.OPENAI_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.openAi';
exports.CHAIN_SUMMARIZATION_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.chainSummarization';
exports.CODE_TOOL_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.toolCode';
exports.WORKFLOW_TOOL_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.toolWorkflow';
exports.HTTP_REQUEST_TOOL_LANGCHAIN_NODE_TYPE = '@n8n/n8n-nodes-langchain.toolHttpRequest';
exports.LANGCHAIN_CUSTOM_TOOLS = [
    exports.CODE_TOOL_LANGCHAIN_NODE_TYPE,
    exports.WORKFLOW_TOOL_LANGCHAIN_NODE_TYPE,
    exports.HTTP_REQUEST_TOOL_LANGCHAIN_NODE_TYPE,
];
exports.SINGLE_EXECUTION_NODES = {
    'n8n-nodes-base.code': {
        mode: [undefined, 'runOnceForAllItems'],
    },
    'n8n-nodes-base.executeWorkflow': {
        mode: [undefined, 'once'],
    },
    'n8n-nodes-base.crateDb': {
        operation: [undefined, 'update'],
    },
    'n8n-nodes-base.timescaleDb': {
        operation: [undefined, 'update'],
    },
    'n8n-nodes-base.microsoftSql': {
        operation: [undefined, 'update', 'delete'],
    },
    'n8n-nodes-base.questDb': {
        operation: [undefined],
    },
    'n8n-nodes-base.mongoDb': {
        operation: ['insert', 'update'],
    },
    'n8n-nodes-base.redis': {
        operation: [undefined],
    },
};
exports.OBFUSCATED_ERROR_MESSAGE = 'Internal error';
