// n8n-workflow/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.randomString = exports.randomInt = exports.updateDisplayOptions = exports.removeCircularRefs = exports.isTraversableObject = exports.assert = exports.fileTypeFromMimeType = exports.sleep = exports.jsonStringify = exports.replaceCircularReferences = exports.jsonParse = exports.deepCopy = exports.isObjectEmpty = void 0;
const form_data_1 = __importDefault(require("form-data"));
const lodash_1 = require("lodash");
const Constants_1 = require("./Constants");
const application_error_1 = require("./errors/application.error");
const readStreamClasses = new Set(['ReadStream', 'Readable', 'ReadableStream']);
BigInt.prototype.toJSON = function () {
    return this.toString();
};
const isObjectEmpty = (obj) => {
    if (obj === undefined || obj === null)
        return true;
    if (typeof obj === 'object') {
        if (obj instanceof form_data_1.default)
            return obj.getLengthSync() === 0;
        if (Array.isArray(obj))
            return obj.length === 0;
        if (obj instanceof Set || obj instanceof Map)
            return obj.size === 0;
        if (ArrayBuffer.isView(obj) || obj instanceof ArrayBuffer)
            return obj.byteLength === 0;
        if (Symbol.iterator in obj || readStreamClasses.has(obj.constructor.name))
            return false;
        return Object.keys(obj).length === 0;
    }
    return true;
};
exports.isObjectEmpty = isObjectEmpty;
const deepCopy = (source, hash = new WeakMap(), path = '') => {
    const hasOwnProp = Object.prototype.hasOwnProperty.bind(source);
    if (typeof source !== 'object' || source === null || typeof source === 'function') {
        return source;
    }
    if (typeof source.toJSON === 'function') {
        return source.toJSON();
    }
    if (hash.has(source)) {
        return hash.get(source);
    }
    if (Array.isArray(source)) {
        const clone = [];
        const len = source.length;
        for (let i = 0; i < len; i++) {
            clone[i] = (0, exports.deepCopy)(source[i], hash, path + `[${i}]`);
        }
        return clone;
    }
    const clone = Object.create(Object.getPrototypeOf({}));
    hash.set(source, clone);
    for (const i in source) {
        if (hasOwnProp(i)) {
            clone[i] = (0, exports.deepCopy)(source[i], hash, path + `.${i}`);
        }
    }
    return clone;
};
exports.deepCopy = deepCopy;
const jsonParse = (jsonString, options) => {
    try {
        return JSON.parse(jsonString);
    }
    catch (error) {
        if ((options === null || options === void 0 ? void 0 : options.fallbackValue) !== undefined) {
            return options.fallbackValue;
        }
        else if (options === null || options === void 0 ? void 0 : options.errorMessage) {
            throw new application_error_1.ApplicationError(options.errorMessage);
        }
        throw error;
    }
};
exports.jsonParse = jsonParse;
const replaceCircularReferences = (value, knownObjects = new WeakSet()) => {
    if (typeof value !== 'object' || value === null || value instanceof RegExp)
        return value;
    if ('toJSON' in value && typeof value.toJSON === 'function')
        return value.toJSON();
    if (knownObjects.has(value))
        return '[Circular Reference]';
    knownObjects.add(value);
    const copy = (Array.isArray(value) ? [] : {});
    for (const key in value) {
        copy[key] = (0, exports.replaceCircularReferences)(value[key], knownObjects);
    }
    knownObjects.delete(value);
    return copy;
};
exports.replaceCircularReferences = replaceCircularReferences;
const jsonStringify = (obj, options = {}) => {
    return JSON.stringify((options === null || options === void 0 ? void 0 : options.replaceCircularRefs) ? (0, exports.replaceCircularReferences)(obj) : obj);
};
exports.jsonStringify = jsonStringify;
const sleep = async (ms) => await new Promise((resolve) => {
    setTimeout(resolve, ms);
});
exports.sleep = sleep;
function fileTypeFromMimeType(mimeType) {
    if (mimeType.startsWith('application/json'))
        return 'json';
    if (mimeType.startsWith('text/html'))
        return 'html';
    if (mimeType.startsWith('image/'))
        return 'image';
    if (mimeType.startsWith('audio/'))
        return 'audio';
    if (mimeType.startsWith('video/'))
        return 'video';
    if (mimeType.startsWith('text/') || mimeType.startsWith('application/javascript'))
        return 'text';
    if (mimeType.startsWith('application/pdf'))
        return 'pdf';
    return;
}
exports.fileTypeFromMimeType = fileTypeFromMimeType;
function assert(condition, msg) {
    if (!condition) {
        const error = new Error(msg !== null && msg !== void 0 ? msg : 'Invalid assertion');
        if (Error.hasOwnProperty('captureStackTrace')) {
            Error.captureStackTrace(error, assert);
        }
        else if (error.stack) {
            error.stack = error.stack
                .split('\n')
                .slice(1)
                .join('\n');
        }
        throw error;
    }
}
exports.assert = assert;
const isTraversableObject = (value) => {
    return value && typeof value === 'object' && !Array.isArray(value) && !!Object.keys(value).length;
};
exports.isTraversableObject = isTraversableObject;
const removeCircularRefs = (obj, seen = new Set()) => {
    seen.add(obj);
    Object.entries(obj).forEach(([key, value]) => {
        if ((0, exports.isTraversableObject)(value)) {
            seen.has(value) ? (obj[key] = { circularReference: true }) : (0, exports.removeCircularRefs)(value, seen);
            return;
        }
        if (Array.isArray(value)) {
            value.forEach((val, index) => {
                if (seen.has(val)) {
                    value[index] = { circularReference: true };
                    return;
                }
                if ((0, exports.isTraversableObject)(val)) {
                    (0, exports.removeCircularRefs)(val, seen);
                }
            });
        }
    });
};
exports.removeCircularRefs = removeCircularRefs;
function updateDisplayOptions(displayOptions, properties) {
    return properties.map((nodeProperty) => {
        return {
            ...nodeProperty,
            displayOptions: (0, lodash_1.merge)({}, nodeProperty.displayOptions, displayOptions),
        };
    });
}
exports.updateDisplayOptions = updateDisplayOptions;
function randomInt(min, max) {
    if (max === undefined) {
        max = min;
        min = 0;
    }
    return min + (crypto.getRandomValues(new Uint32Array(1))[0] % (max - min));
}
exports.randomInt = randomInt;
function randomString(minLength, maxLength) {
    const length = maxLength === undefined ? minLength : randomInt(minLength, maxLength + 1);
    return [...crypto.getRandomValues(new Uint32Array(length))]
        .map((byte) => Constants_1.ALPHABET[byte % Constants_1.ALPHABET.length])
        .join('');
}
exports.randomString = randomString;
