// n8n-nodes-base/dist/nodes/Cortex/AnalyzerDescriptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzerFields = exports.analyzersOperations = void 0;
const AnalyzerInterface_1 = require("./AnalyzerInterface");
exports.analyzersOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        required: true,
        description: 'Choose an operation',
        displayOptions: {
            show: {
                resource: [
                    'analyzer',
                ],
            },
        },
        default: 'execute',
        options: [
            {
                name: 'Execute',
                value: 'execute',
                description: 'Execute Analyzer',
            },
        ],
    },
];
exports.analyzerFields = [
    {
        displayName: 'Analyzer Type',
        name: 'analyzer',
        type: 'options',
        required: true,
        typeOptions: {
            loadOptionsMethod: 'loadActiveAnalyzers',
        },
        displayOptions: {
            show: {
                resource: [
                    'analyzer',
                ],
                operation: [
                    'execute',
                ],
            },
        },
        description: 'Choose the analyzer',
        default: '',
    },
    {
        displayName: 'Observable Type',
        name: 'observableType',
        type: 'options',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'analyzer',
                ],
                operation: [
                    'execute',
                ],
            },
            hide: {
                analyzer: [
                    '',
                ],
            },
        },
        typeOptions: {
            loadOptionsMethod: 'loadObservableOptions',
            loadOptionsDependsOn: [
                'analyzer',
            ],
        },
        default: '',
        description: 'Choose the observable type',
    },
    {
        displayName: 'Observable Value',
        name: 'observableValue',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'analyzer',
                ],
                operation: [
                    'execute',
                ],
            },
            hide: {
                observableType: [
                    'file',
                ],
                analyzer: [
                    '',
                ],
            },
        },
        default: '',
        description: 'Enter the observable value',
    },
    {
        displayName: 'Binary Property',
        name: 'binaryPropertyName',
        type: 'string',
        default: 'data',
        required: true,
        displayOptions: {
            show: {
                observableType: [
                    'file',
                ],
                resource: [
                    'analyzer',
                ],
                operation: [
                    'execute',
                ],
            },
        },
        description: 'Name of the binary property to which to write the data of the read file.',
    },
    {
        displayName: 'TLP',
        name: 'tlp',
        type: 'options',
        required: false,
        displayOptions: {
            show: {
                resource: [
                    'analyzer',
                ],
                operation: [
                    'execute',
                ],
            },
            hide: {
                observableType: [
                    '',
                ],
                analyzer: [
                    '',
                ],
            },
        },
        options: [
            {
                name: 'White',
                value: AnalyzerInterface_1.TLP.white,
            },
            {
                name: 'Green',
                value: AnalyzerInterface_1.TLP.green,
            },
            {
                name: 'Amber',
                value: AnalyzerInterface_1.TLP.amber,
            }, {
                name: 'Red',
                value: AnalyzerInterface_1.TLP.red,
            },
        ],
        default: 2,
        description: 'The TLP of the analyzed observable',
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'analyzer',
                ],
                operation: [
                    'execute',
                ],
            },
        },
        options: [
            {
                displayName: 'Force',
                name: 'force',
                type: 'boolean',
                default: false,
                description: 'To force bypassing the cache, set this parameter to true',
            },
            {
                displayName: 'Timeout (seconds)',
                name: 'timeout',
                type: 'number',
                default: 3,
                description: 'Timeout to wait for the report in case it is not available at the time the query was made',
            },
        ],
    },
];
