// n8n-nodes-base/dist/nodes/Cortex/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareParameters = exports.splitTags = exports.getEntityLabel = exports.cortexApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const moment = require("moment");
async function cortexApiRequest(method, resource, body = {}, query = {}, uri, option = {}) {
    const credentials = await this.getCredentials('cortexApi');
    if (credentials === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials got returned!');
    }
    const headerWithAuthentication = Object.assign({}, { Authorization: ` Bearer ${credentials.cortexApiKey}` });
    let options = {
        headers: headerWithAuthentication,
        method,
        qs: query,
        uri: uri || `${credentials.host}/api${resource}`,
        body,
        json: true,
    };
    if (Object.keys(option).length !== 0) {
        options = Object.assign({}, options, option);
    }
    if (Object.keys(body).length === 0) {
        delete options.body;
    }
    if (Object.keys(query).length === 0) {
        delete options.qs;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.cortexApiRequest = cortexApiRequest;
function getEntityLabel(entity) {
    let label = '';
    switch (entity._type) {
        case 'case':
            label = `#${entity.caseId} ${entity.title}`;
            break;
        case 'case_artifact':
            label = `[${entity.dataType}] ${entity.data ? entity.data : (entity.attachment.name)}`;
            break;
        case 'alert':
            label = `[${entity.source}:${entity.sourceRef}] ${entity.title}`;
            break;
        case 'case_task_log':
            label = `${entity.message} from ${entity.createdBy}`;
            break;
        case 'case_task':
            label = `${entity.title} (${entity.status})`;
            break;
        case 'job':
            label = `${entity.analyzerName} (${entity.status})`;
            break;
        default:
            break;
    }
    return label;
}
exports.getEntityLabel = getEntityLabel;
function splitTags(tags) {
    return tags.split(',').filter(tag => tag !== ' ' && tag);
}
exports.splitTags = splitTags;
function prepareParameters(values) {
    const response = {};
    for (const key in values) {
        if (values[key] !== undefined && values[key] !== null && values[key] !== '') {
            if (moment(values[key], moment.ISO_8601).isValid()) {
                response[key] = Date.parse(values[key]);
            }
            else if (key === 'tags') {
                response[key] = splitTags(values[key]);
            }
            else {
                response[key] = values[key];
            }
        }
    }
    return response;
}
exports.prepareParameters = prepareParameters;
