// n8n-nodes-base/dist/nodes/Microsoft/Teams/v2/actions/versionDescription.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const channel = __importStar(require("./channel"));
const channelMessage = __importStar(require("./channelMessage"));
const chatMessage = __importStar(require("./chatMessage"));
const task = __importStar(require("./task"));
exports.versionDescription = {
    displayName: 'Microsoft Teams',
    name: 'microsoftTeams',
    icon: 'file:teams.svg',
    group: ['input'],
    version: 2,
    subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
    description: 'Consume Microsoft Teams API',
    defaults: {
        name: 'Microsoft Teams',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'microsoftTeamsOAuth2Api',
            required: true,
        },
    ],
    properties: [
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            noDataExpression: true,
            options: [
                {
                    name: 'Channel',
                    value: 'channel',
                },
                {
                    name: 'Channel Message',
                    value: 'channelMessage',
                },
                {
                    name: 'Chat Message',
                    value: 'chatMessage',
                },
                {
                    name: 'Task',
                    value: 'task',
                },
            ],
            default: 'channel',
        },
        ...channel.description,
        ...channelMessage.description,
        ...chatMessage.description,
        ...task.description,
    ],
};
