// @n8n/n8n-nodes-langchain/dist/nodes/llms/LmChatGoogleGemini/LmChatGoogleGemini.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LmChatGoogleGemini = void 0;
const google_genai_1 = require("@langchain/google-genai");
const sharedFields_1 = require("../../../utils/sharedFields");
const N8nLlmTracing_1 = require("../N8nLlmTracing");
const options_1 = require("./options");
class LmChatGoogleGemini {
    constructor() {
        this.description = {
            displayName: 'Google Gemini Chat Model',
            name: 'lmChatGoogleGemini',
            icon: 'file:google.svg',
            group: ['transform'],
            version: 1,
            description: 'Chat Model Google Gemini',
            defaults: {
                name: 'Google Gemini Chat Model',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Language Models'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.lmchatgooglegemini/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_languageModel"],
            outputNames: ['Model'],
            credentials: [
                {
                    name: 'googlePalmApi',
                    required: true,
                },
            ],
            requestDefaults: {
                ignoreHttpStatusErrors: true,
                baseURL: '={{ $credentials.host }}',
            },
            properties: [
                (0, sharedFields_1.getConnectionHintNoticeField)(["ai_chain", "ai_agent"]),
                {
                    displayName: 'Model',
                    name: 'modelName',
                    type: 'options',
                    description: 'The model which will generate the completion. <a href="https://developers.generativeai.google/api/rest/generativelanguage/models/list">Learn more</a>.',
                    typeOptions: {
                        loadOptions: {
                            routing: {
                                request: {
                                    method: 'GET',
                                    url: '/v1beta/models',
                                },
                                output: {
                                    postReceive: [
                                        {
                                            type: 'rootProperty',
                                            properties: {
                                                property: 'models',
                                            },
                                        },
                                        {
                                            type: 'filter',
                                            properties: {
                                                pass: "={{ !$responseItem.name.includes('embedding') }}",
                                            },
                                        },
                                        {
                                            type: 'setKeyValue',
                                            properties: {
                                                name: '={{$responseItem.name}}',
                                                value: '={{$responseItem.name}}',
                                                description: '={{$responseItem.description}}',
                                            },
                                        },
                                        {
                                            type: 'sort',
                                            properties: {
                                                key: 'name',
                                            },
                                        },
                                    ],
                                },
                            },
                        },
                    },
                    routing: {
                        send: {
                            type: 'body',
                            property: 'model',
                        },
                    },
                    default: 'models/gemini-1.0-pro',
                },
                {
                    displayName: 'Options',
                    name: 'options',
                    placeholder: 'Add Option',
                    description: 'Additional options to add',
                    type: 'collection',
                    default: {},
                    options: [
                        {
                            displayName: 'Maximum Number of Tokens',
                            name: 'maxOutputTokens',
                            default: 2048,
                            description: 'The maximum number of tokens to generate in the completion',
                            type: 'number',
                        },
                        {
                            displayName: 'Sampling Temperature',
                            name: 'temperature',
                            default: 0.4,
                            typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
                            description: 'Controls randomness: Lowering results in less random completions. As the temperature approaches zero, the model will become deterministic and repetitive.',
                            type: 'number',
                        },
                        {
                            displayName: 'Top K',
                            name: 'topK',
                            default: 32,
                            typeOptions: { maxValue: 40, minValue: -1, numberPrecision: 1 },
                            description: 'Used to remove "long tail" low probability responses. Defaults to -1, which disables it.',
                            type: 'number',
                        },
                        {
                            displayName: 'Top P',
                            name: 'topP',
                            default: 1,
                            typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
                            description: 'Controls diversity via nucleus sampling: 0.5 means half of all likelihood-weighted options are considered. We generally recommend altering this or temperature but not both.',
                            type: 'number',
                        },
                        {
                            displayName: 'Safety Settings',
                            name: 'safetySettings',
                            type: 'fixedCollection',
                            typeOptions: { multipleValues: true },
                            default: {
                                values: {
                                    category: options_1.harmCategories[0].name,
                                    threshold: options_1.harmThresholds[0].name,
                                },
                            },
                            placeholder: 'Add Option',
                            options: [
                                {
                                    name: 'values',
                                    displayName: 'Values',
                                    values: [
                                        {
                                            displayName: 'Safety Category',
                                            name: 'category',
                                            type: 'options',
                                            description: 'The category of harmful content to block',
                                            default: 'HARM_CATEGORY_UNSPECIFIED',
                                            options: options_1.harmCategories,
                                        },
                                        {
                                            displayName: 'Safety Threshold',
                                            name: 'threshold',
                                            type: 'options',
                                            description: 'The threshold of harmful content to block',
                                            default: 'HARM_BLOCK_THRESHOLD_UNSPECIFIED',
                                            options: options_1.harmThresholds,
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const credentials = await this.getCredentials('googlePalmApi');
        const modelName = this.getNodeParameter('modelName', itemIndex);
        const options = this.getNodeParameter('options', itemIndex, {
            maxOutputTokens: 1024,
            temperature: 0.7,
            topK: 40,
            topP: 0.9,
        });
        const safetySettings = this.getNodeParameter('options.safetySettings.values', itemIndex, null);
        const model = new google_genai_1.ChatGoogleGenerativeAI({
            apiKey: credentials.apiKey,
            modelName,
            topK: options.topK,
            topP: options.topP,
            temperature: options.temperature,
            maxOutputTokens: options.maxOutputTokens,
            safetySettings,
            callbacks: [new N8nLlmTracing_1.N8nLlmTracing(this)],
        });
        return {
            response: model,
        };
    }
}
exports.LmChatGoogleGemini = LmChatGoogleGemini;
