// n8n-nodes-base/dist/nodes/HighLevel/v2/HighLevelV2.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HighLevelV2 = void 0;
const ContactDescription_1 = require("./description/ContactDescription");
const OpportunityDescription_1 = require("./description/OpportunityDescription");
const TaskDescription_1 = require("./description/TaskDescription");
const GenericFunctions_1 = require("./GenericFunctions");
const resources = [
    {
        displayName: 'Resource',
        name: 'resource',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Contact',
                value: 'contact',
            },
            {
                name: 'Opportunity',
                value: 'opportunity',
            },
            {
                name: 'Task',
                value: 'task',
            },
        ],
        default: 'contact',
        required: true,
    },
];
const versionDescription = {
    displayName: 'HighLevel',
    name: 'highLevel',
    icon: 'file:highLevel.svg',
    group: ['transform'],
    version: 2,
    description: 'Consume HighLevel API v2',
    subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
    defaults: {
        name: 'HighLevel',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'highLevelOAuth2Api',
            required: true,
        },
    ],
    requestDefaults: {
        baseURL: 'https://services.leadconnectorhq.com',
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            Version: '2021-07-28',
        },
    },
    requestOperations: {
        pagination: GenericFunctions_1.highLevelApiPagination,
    },
    properties: [
        ...resources,
        ...ContactDescription_1.contactOperations,
        ...ContactDescription_1.contactNotes,
        ...ContactDescription_1.contactFields,
        ...OpportunityDescription_1.opportunityOperations,
        ...OpportunityDescription_1.opportunityFields,
        ...TaskDescription_1.taskOperations,
        ...TaskDescription_1.taskFields,
    ],
};
class HighLevelV2 {
    constructor(baseDescription) {
        this.methods = {
            loadOptions: {
                getPipelines: GenericFunctions_1.getPipelines,
                getContacts: GenericFunctions_1.getContacts,
                getPipelineStages: GenericFunctions_1.getPipelineStages,
                getUsers: GenericFunctions_1.getUsers,
                getTimezones: GenericFunctions_1.getTimezones,
            },
        };
        this.description = {
            ...baseDescription,
            ...versionDescription,
        };
    }
}
exports.HighLevelV2 = HighLevelV2;
