// n8n-nodes-base/dist/nodes/Microsoft/Teams/v2/actions/channelMessage/create.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const utils_1 = require("../../helpers/utils");
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utilities_1 = require("../../../../../../utils/utilities");
const properties = [
    descriptions_1.teamRLC,
    descriptions_1.channelRLC,
    {
        displayName: 'Content Type',
        name: 'contentType',
        required: true,
        type: 'options',
        options: [
            {
                name: 'Text',
                value: 'text',
            },
            {
                name: 'HTML',
                value: 'html',
            },
        ],
        default: 'text',
        description: 'Whether the message is plain text or HTML',
    },
    {
        displayName: 'Message',
        name: 'message',
        required: true,
        type: 'string',
        default: '',
        description: 'The content of the message to be sent',
        typeOptions: {
            rows: 2,
        },
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        options: [
            {
                displayName: 'Include Link to Workflow',
                name: 'includeLinkToWorkflow',
                type: 'boolean',
                default: true,
                description: 'Whether to append a link to this workflow at the end of the message. This is helpful if you have many workflows sending messages.',
            },
            {
                displayName: 'Reply to ID',
                name: 'makeReply',
                type: 'string',
                default: '',
                placeholder: 'e.g. 1673348720590',
                description: 'An optional ID of the message you want to reply to. The message ID is the number before "?tenantId" in the message URL.',
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['channelMessage'],
        operation: ['create'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i, nodeVersion, instanceId) {
    const teamId = this.getNodeParameter('teamId', i, '', { extractValue: true });
    const channelId = this.getNodeParameter('channelId', i, '', { extractValue: true });
    const contentType = this.getNodeParameter('contentType', i);
    const message = this.getNodeParameter('message', i);
    const options = this.getNodeParameter('options', i);
    let includeLinkToWorkflow = options.includeLinkToWorkflow;
    if (includeLinkToWorkflow === undefined) {
        includeLinkToWorkflow = nodeVersion >= 1.1;
    }
    const body = utils_1.prepareMessage.call(this, message, contentType, includeLinkToWorkflow, instanceId);
    if (options.makeReply) {
        const replyToId = options.makeReply;
        return await transport_1.microsoftApiRequest.call(this, 'POST', `/beta/teams/${teamId}/channels/${channelId}/messages/${replyToId}/replies`, body);
    }
    else {
        return await transport_1.microsoftApiRequest.call(this, 'POST', `/beta/teams/${teamId}/channels/${channelId}/messages`, body);
    }
}
exports.execute = execute;
