// n8n-nodes-base/dist/nodes/CustomerIo/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.eventExists = exports.customerIoApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const lodash_1 = require("lodash");
async function customerIoApiRequest(method, endpoint, body, baseApi, query) {
    const credentials = await this.getCredentials('customerIoApi');
    if (credentials === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials got returned!');
    }
    query = query || {};
    const options = {
        headers: {
            'Content-Type': 'application/json',
        },
        method,
        body,
        uri: '',
        json: true,
    };
    if (baseApi === 'tracking') {
        options.uri = `https://track.customer.io/api/v1${endpoint}`;
        const basicAuthKey = Buffer.from(`${credentials.trackingSiteId}:${credentials.trackingApiKey}`).toString('base64');
        Object.assign(options.headers, { 'Authorization': `Basic ${basicAuthKey}` });
    }
    else if (baseApi === 'api') {
        options.uri = `https://api.customer.io/v1/api${endpoint}`;
        const basicAuthKey = Buffer.from(`${credentials.trackingSiteId}:${credentials.trackingApiKey}`).toString('base64');
        Object.assign(options.headers, { 'Authorization': `Basic ${basicAuthKey}` });
    }
    else if (baseApi === 'beta') {
        options.uri = `https://beta-api.customer.io/v1/api${endpoint}`;
        Object.assign(options.headers, { 'Authorization': `Bearer ${credentials.appApiKey}` });
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.customerIoApiRequest = customerIoApiRequest;
function eventExists(currentEvents, webhookEvents) {
    for (const currentEvent of currentEvents) {
        if (lodash_1.get(webhookEvents, `${currentEvent.split('.')[0]}.${currentEvent.split('.')[1]}`) !== true) {
            return false;
        }
    }
    return true;
}
exports.eventExists = eventExists;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
