// n8n-nodes-base/dist/nodes/ClickUp/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.clickupApiRequestAllItems = exports.clickupApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function clickupApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    const options = {
        headers: {
            'Content-Type': 'application/json',
        },
        method,
        qs,
        body,
        uri: uri || `https://api.clickup.com/api/v2${resource}`,
        json: true,
    };
    try {
        const authenticationMethod = this.getNodeParameter('authentication', 0);
        if (authenticationMethod === 'accessToken') {
            const credentials = await this.getCredentials('clickUpApi');
            options.headers['Authorization'] = credentials === null || credentials === void 0 ? void 0 : credentials.accessToken;
            return await this.helpers.request(options);
        }
        else {
            const oAuth2Options = {
                keepBearer: false,
                tokenType: 'Bearer',
            };
            return await this.helpers.requestOAuth2.call(this, 'clickUpOAuth2Api', options, oAuth2Options);
        }
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.clickupApiRequest = clickupApiRequest;
async function clickupApiRequestAllItems(propertyName, method, resource, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query.page = 0;
    do {
        responseData = await clickupApiRequest.call(this, method, resource, body, query);
        returnData.push.apply(returnData, responseData[propertyName]);
        query.page++;
        if (query.limit && query.limit <= returnData.length) {
            return returnData;
        }
    } while (responseData[propertyName] &&
        responseData[propertyName].length !== 0);
    return returnData;
}
exports.clickupApiRequestAllItems = clickupApiRequestAllItems;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
