// n8n-nodes-base/dist/nodes/ClickUp/TaskListDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.taskListFields = exports.taskListOperations = void 0;
exports.taskListOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'taskList',
                ],
            },
        },
        options: [
            {
                name: 'Add',
                value: 'add',
                description: 'Add a task to a list',
            },
            {
                name: 'Remove',
                value: 'remove',
                description: 'Remove a task from a list',
            },
        ],
        default: 'add',
        description: 'The operation to perform.',
    },
];
exports.taskListFields = [
    {
        displayName: 'Task ID',
        name: 'taskId',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'taskList',
                ],
                operation: [
                    'remove',
                    'add',
                ],
            },
        },
        required: true,
    },
    {
        displayName: 'List ID',
        name: 'listId',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'taskList',
                ],
                operation: [
                    'remove',
                    'add',
                ],
            },
        },
        required: true,
    },
];
