// @n8n/tournament/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tournament = void 0;
const ExpressionBuilder_1 = require("./ExpressionBuilder");
const Analysis_1 = require("./Analysis");
const DATA_NODE_NAME = '___n8n_data';
class Tournament {
    constructor(errorHandler = () => { }, _dataNodeName = DATA_NODE_NAME) {
        this.errorHandler = errorHandler;
        this._dataNodeName = _dataNodeName;
        this._codeCache = {};
    }
    getExpressionCode(expr) {
        return (0, ExpressionBuilder_1.getExpressionCode)(expr, this._dataNodeName);
    }
    tmplDiff(expr) {
        return (0, Analysis_1.getTmplDifference)(expr, this._dataNodeName);
    }
    getFunction(expr) {
        if (expr in this._codeCache) {
            return this._codeCache[expr];
        }
        const [code, analysis] = this.getExpressionCode(expr);
        const func = new Function('E', code + ';');
        this._codeCache[expr] = [func, analysis];
        return [func, analysis];
    }
    execute(expr, data) {
        if (!expr) {
            return expr;
        }
        const fn = this.getFunction(expr)[0];
        return fn.call(data, this.errorHandler);
    }
}
exports.Tournament = Tournament;
