// n8n-nodes-base/dist/nodes/Mautic/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.mauticApiRequestAllItems = exports.mauticApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function mauticApiRequest(method, endpoint, body = {}, query, uri) {
    const authenticationMethod = this.getNodeParameter('authentication', 0, 'credentials');
    const options = {
        headers: {},
        method,
        qs: query,
        uri: uri || `/api${endpoint}`,
        body,
        json: true,
    };
    try {
        let returnData;
        if (authenticationMethod === 'credentials') {
            const credentials = await this.getCredentials('mauticApi');
            const base64Key = Buffer.from(`${credentials.username}:${credentials.password}`).toString('base64');
            options.headers.Authorization = `Basic ${base64Key}`;
            options.uri = `${credentials.url}${options.uri}`;
            returnData = await this.helpers.request(options);
        }
        else {
            const credentials = await this.getCredentials('mauticOAuth2Api');
            options.uri = `${credentials.url}${options.uri}`;
            returnData = await this.helpers.requestOAuth2.call(this, 'mauticOAuth2Api', options, { includeCredentialsOnRefreshOnBody: true });
        }
        if (returnData.errors) {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), returnData);
        }
        return returnData;
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.mauticApiRequest = mauticApiRequest;
async function mauticApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    let data = [];
    query.limit = 30;
    query.start = 0;
    do {
        responseData = await mauticApiRequest.call(this, method, endpoint, body, query);
        const values = Object.values(responseData[propertyName]);
        for (const value of values) {
            data.push(value);
        }
        returnData.push.apply(returnData, data);
        query.start++;
        data = [];
    } while (responseData.total !== undefined &&
        ((query.limit * query.start) - parseInt(responseData.total, 10)) < 0);
    return returnData;
}
exports.mauticApiRequestAllItems = mauticApiRequestAllItems;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
