// n8n-nodes-base/dist/nodes/TheHiveProject/actions/observable/get.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
const descriptions_1 = require("../../descriptions");
const utilities_1 = require("../../../../utils/utilities");
const properties = [descriptions_1.observableRLC];
const displayOptions = {
    show: {
        resource: ['observable'],
        operation: ['get'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const observableId = this.getNodeParameter('observableId', i, '', {
        extractValue: true,
    });
    const qs = {};
    const body = {
        query: [
            {
                _name: 'getObservable',
                idOrName: observableId,
            },
        ],
    };
    qs.name = `get-observable-${observableId}`;
    responseData = await transport_1.theHiveApiRequest.call(this, 'POST', '/v1/query', body, qs);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
