// n8n-nodes-base/dist/nodes/Kitemaker/descriptions/UserDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userFields = exports.userOperations = void 0;
exports.userOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        default: 'getAll',
        description: 'Operation to perform.',
        options: [
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Retrieve data on all the users in the<br>logged-in user\'s organization.',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'user',
                ],
            },
        },
    },
];
exports.userFields = [
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'user',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 5,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
            maxValue: 1000,
        },
        displayOptions: {
            show: {
                resource: [
                    'user',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
];
