// n8n-nodes-base/dist/nodes/ItemLists/V3/actions/itemList/limit.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const utilities_1 = require("../../../../../utils/utilities");
const properties = [
    {
        displayName: 'Max Items',
        name: 'maxItems',
        type: 'number',
        typeOptions: {
            minValue: 1,
        },
        default: 1,
        description: 'If there are more items than this number, some are removed',
    },
    {
        displayName: 'Keep',
        name: 'keep',
        type: 'options',
        options: [
            {
                name: 'First Items',
                value: 'firstItems',
            },
            {
                name: 'Last Items',
                value: 'lastItems',
            },
        ],
        default: 'firstItems',
        description: 'When removing items, whether to keep the ones at the start or the ending',
    },
];
const displayOptions = {
    show: {
        resource: ['itemList'],
        operation: ['limit'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(items) {
    let returnData = items;
    const maxItems = this.getNodeParameter('maxItems', 0);
    const keep = this.getNodeParameter('keep', 0);
    if (maxItems > items.length) {
        return returnData;
    }
    if (keep === 'firstItems') {
        returnData = items.slice(0, maxItems);
    }
    else {
        returnData = items.slice(items.length - maxItems, items.length);
    }
    return returnData;
}
exports.execute = execute;
