// @n8n/n8n-nodes-langchain/dist/nodes/vendors/OpenAi/actions/image/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.analyze = exports.generate = void 0;
const generate = __importStar(require("./generate.operation"));
exports.generate = generate;
const analyze = __importStar(require("./analyze.operation"));
exports.analyze = analyze;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Analyze Image',
                value: 'analyze',
                action: 'Analyze image',
                description: 'Take in images and answer questions about them',
            },
            {
                name: 'Generate an Image',
                value: 'generate',
                action: 'Generate an image',
                description: 'Creates an image from a text prompt',
            },
        ],
        default: 'generate',
        displayOptions: {
            show: {
                resource: ['image'],
            },
        },
    },
    ...generate.description,
    ...analyze.description,
];
