// n8n-nodes-base/dist/nodes/SseTrigger/SseTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SseTrigger = void 0;
const EventSource = require("eventsource");
class SseTrigger {
    constructor() {
        this.description = {
            displayName: 'SSE Trigger',
            name: 'sseTrigger',
            icon: 'fa:cloud-download-alt',
            group: ['trigger'],
            version: 1,
            description: 'Triggers the workflow when Server-Sent Events occur',
            defaults: {
                name: 'SSE Trigger',
                color: '#225577',
            },
            inputs: [],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'URL',
                    name: 'url',
                    type: 'string',
                    default: '',
                    placeholder: 'http://example.com',
                    description: 'The URL to receive the SSE from.',
                    required: true,
                },
            ],
        };
    }
    async trigger() {
        const url = this.getNodeParameter('url');
        const eventSource = new EventSource(url);
        eventSource.onmessage = (event) => {
            const eventData = JSON.parse(event.data);
            this.emit([this.helpers.returnJsonArray([eventData])]);
        };
        async function closeFunction() {
            eventSource.close();
        }
        return {
            closeFunction,
        };
    }
}
exports.SseTrigger = SseTrigger;
