// n8n-nodes-base/dist/nodes/TheHiveProject/actions/task/executeResponder.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const descriptions_1 = require("../../descriptions");
const transport_1 = require("../../transport");
const utilities_1 = require("../../../../utils/utilities");
const properties = [{ ...descriptions_1.taskRLC, name: 'id' }, descriptions_1.responderOptions];
const displayOptions = {
    show: {
        resource: ['task'],
        operation: ['executeResponder'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const taskId = this.getNodeParameter('id', i);
    const responderId = this.getNodeParameter('responder', i);
    let body;
    let response;
    responseData = [];
    const qs = {};
    body = {
        responderId,
        objectId: taskId,
        objectType: 'case_task',
    };
    response = await transport_1.theHiveApiRequest.call(this, 'POST', '/connector/cortex/action', body);
    body = {
        query: [
            {
                _name: 'listAction',
            },
            {
                _name: 'filter',
                _and: [
                    {
                        _field: 'cortexId',
                        _value: response.cortexId,
                    },
                    {
                        _field: 'objectId',
                        _value: response.objectId,
                    },
                    {
                        _field: 'startDate',
                        _value: response.startDate,
                    },
                ],
            },
        ],
    };
    qs.name = 'task-actions';
    do {
        response = await transport_1.theHiveApiRequest.call(this, 'POST', '/v1/query', body, qs);
    } while (response.status === 'Waiting' || response.status === 'InProgress');
    responseData = response;
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
exports.execute = execute;
