// n8n-nodes-base/dist/nodes/MondayCom/BoardGroupDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.boardGroupFields = exports.boardGroupOperations = void 0;
exports.boardGroupOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'boardGroup',
                ],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'delete',
                description: 'Delete a group in a board',
            },
            {
                name: 'Create',
                value: 'create',
                description: 'Create a group in a board',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: `Get list of groups in a board`,
            },
        ],
        default: 'create',
        description: 'The operation to perform.',
    },
];
exports.boardGroupFields = [
    {
        displayName: 'Board ID',
        name: 'boardId',
        type: 'options',
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getBoards',
        },
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'boardGroup',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    {
        displayName: 'Name',
        name: 'name',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                operation: [
                    'create',
                ],
                resource: [
                    'boardGroup',
                ],
            },
        },
        default: '',
        description: `The group name`,
    },
    {
        displayName: 'Board ID',
        name: 'boardId',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getBoards',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'boardGroup',
                ],
                operation: [
                    'delete',
                ],
            },
        },
    },
    {
        displayName: 'Group ID',
        name: 'groupId',
        type: 'options',
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getGroups',
            loadOptionsDependsOn: [
                'boardId',
            ],
        },
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'boardGroup',
                ],
                operation: [
                    'delete',
                ],
            },
        },
    },
    {
        displayName: 'Board ID',
        name: 'boardId',
        type: 'options',
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getBoards',
        },
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'boardGroup',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
];
