// n8n-nodes-base/dist/nodes/Mattermost/v1/actions/versionDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const channel = require("./channel");
const message = require("./message");
const reaction = require("./reaction");
const user = require("./user");
exports.versionDescription = {
    displayName: 'Mattermost',
    name: 'mattermost',
    icon: 'file:mattermost.svg',
    group: ['output'],
    version: 1,
    subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
    description: 'Sends data to Mattermost',
    defaults: {
        name: 'Mattermost',
        color: '#000000',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'mattermostApi',
            required: true,
        },
    ],
    properties: [
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            options: [
                {
                    name: 'Channel',
                    value: 'channel',
                },
                {
                    name: 'Message',
                    value: 'message',
                },
                {
                    name: 'Reaction',
                    value: 'reaction',
                },
                {
                    name: 'User',
                    value: 'user',
                },
            ],
            default: 'message',
            description: 'The resource to operate on',
        },
        ...channel.descriptions,
        ...message.descriptions,
        ...reaction.descriptions,
        ...user.descriptions,
    ],
};
