// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolCalculator/ToolCalculator.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolCalculator = void 0;
const calculator_1 = require("@langchain/community/tools/calculator");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
class ToolCalculator {
    constructor() {
        this.description = {
            displayName: 'Calculator',
            name: 'toolCalculator',
            icon: 'fa:calculator',
            group: ['transform'],
            version: 1,
            description: 'Make it easier for AI agents to perform arithmetic',
            defaults: {
                name: 'Calculator',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolcalculator/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            properties: [(0, sharedFields_1.getConnectionHintNoticeField)(["ai_agent"])],
        };
    }
    async supplyData() {
        return {
            response: (0, logWrapper_1.logWrapper)(new calculator_1.Calculator(), this),
        };
    }
}
exports.ToolCalculator = ToolCalculator;
