// n8n-nodes-base/dist/nodes/Notion/DatabaseDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.databaseFields = exports.databaseOperations = void 0;
exports.databaseOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'database',
                ],
            },
        },
        options: [
            {
                name: 'Get',
                value: 'get',
                description: 'Get a database',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Get all databases',
            },
        ],
        default: 'get',
        description: 'The operation to perform.',
    },
];
exports.databaseFields = [
    {
        displayName: 'Database ID',
        name: 'databaseId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'database',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                resource: [
                    'database',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        default: false,
        description: 'If all results should be returned or only up to a given limit.',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                resource: [
                    'database',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 100,
        },
        default: 50,
        description: 'How many results to return.',
    },
];
