// @n8n/n8n-nodes-langchain/dist/nodes/vendors/OpenAi/actions/text/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.message = exports.classify = void 0;
const classify = __importStar(require("./classify.operation"));
exports.classify = classify;
const message = __importStar(require("./message.operation"));
exports.message = message;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Message a Model',
                value: 'message',
                action: 'Message a model',
                description: 'Create a completion with GPT 3, 4, etc.',
            },
            {
                name: 'Classify Text for Violations',
                value: 'classify',
                action: 'Classify text for violations',
                description: 'Check whether content complies with usage policies',
            },
        ],
        default: 'message',
        displayOptions: {
            show: {
                resource: ['text'],
            },
        },
    },
    ...classify.description,
    ...message.description,
];
