// n8n-nodes-base/dist/nodes/ActionNetwork/descriptions/AttendanceDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attendanceFields = exports.attendanceOperations = void 0;
const SharedFields_1 = require("./SharedFields");
exports.attendanceOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
            },
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
        ],
        default: 'create',
        description: 'Operation to perform',
    },
];
exports.attendanceFields = [
    {
        displayName: 'Person ID',
        name: 'personId',
        description: 'ID of the person to create an attendance for.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    {
        displayName: 'Event ID',
        name: 'eventId',
        description: 'ID of the event to create an attendance for.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('attendance', 'create'),
    {
        displayName: 'Event ID',
        name: 'eventId',
        description: 'ID of the event whose attendance to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    {
        displayName: 'Attendance ID',
        name: 'attendanceId',
        description: 'ID of the attendance to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('attendance', 'get'),
    {
        displayName: 'Event ID',
        name: 'eventId',
        description: 'ID of the event to create an attendance for.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 50,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                resource: [
                    'attendance',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('attendance', 'getAll'),
];
