// n8n-nodes-base/dist/nodes/ActionNetwork/descriptions/EventDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventFields = exports.eventOperations = void 0;
const SharedFields_1 = require("./SharedFields");
exports.eventOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
            },
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
        ],
        default: 'create',
        description: 'Operation to perform',
    },
];
exports.eventFields = [
    {
        displayName: 'Origin System',
        name: 'originSystem',
        description: 'Source where the event originated.',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    {
        displayName: 'Title',
        name: 'title',
        description: 'Title of the event to create.',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('event', 'create'),
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
                operation: [
                    'create',
                ],
            },
        },
        options: SharedFields_1.eventAdditionalFieldsOptions,
    },
    {
        displayName: 'Event ID',
        name: 'eventId',
        description: 'ID of the event to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('event', 'get'),
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 50,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                resource: [
                    'event',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('event', 'getAll'),
];
