// n8n-nodes-base/dist/nodes/ActionNetwork/descriptions/SignatureDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signatureFields = exports.signatureOperations = void 0;
const SharedFields_1 = require("./SharedFields");
exports.signatureOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
            },
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
            {
                name: 'Update',
                value: 'update',
            },
        ],
        default: 'create',
        description: 'Operation to perform',
    },
];
exports.signatureFields = [
    {
        displayName: 'Petition ID',
        name: 'petitionId',
        description: 'ID of the petition to sign.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    {
        displayName: 'Person ID',
        name: 'personId',
        description: 'ID of the person whose signature to create.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('signature', 'create'),
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'create',
                ],
            },
        },
        options: [
            {
                displayName: 'Comments',
                name: 'comments',
                type: 'string',
                default: '',
                description: 'Comments to leave when signing this petition.',
            },
        ],
    },
    {
        displayName: 'Petition ID',
        name: 'petitionId',
        description: 'ID of the petition whose signature to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    {
        displayName: 'Signature ID',
        name: 'signatureId',
        description: 'ID of the signature to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('signature', 'get'),
    {
        displayName: 'Petition ID',
        name: 'petitionId',
        description: 'ID of the petition whose signatures to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 50,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('signature', 'getAll'),
    {
        displayName: 'Petition ID',
        name: 'petitionId',
        description: 'ID of the petition whose signature to update.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'update',
                ],
            },
        },
    },
    {
        displayName: 'Signature ID',
        name: 'signatureId',
        description: 'ID of the signature to update.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'update',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('signature', 'update'),
    {
        displayName: 'Update Fields',
        name: 'updateFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'signature',
                ],
                operation: [
                    'update',
                ],
            },
        },
        options: [
            {
                displayName: 'Comments',
                name: 'comments',
                type: 'string',
                default: '',
                description: 'Comments to leave when signing this petition.',
            },
        ],
    },
];
