// n8n-nodes-base/dist/nodes/ActionNetwork/descriptions/TagDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tagFields = exports.tagOperations = void 0;
const SharedFields_1 = require("./SharedFields");
exports.tagOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'tag',
                ],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
            },
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
        ],
        default: 'create',
        description: 'Operation to perform',
    },
];
exports.tagFields = [
    {
        displayName: 'Name',
        name: 'name',
        description: 'Name of the tag to create.',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'tag',
                ],
                operation: [
                    'create',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('tag', 'create'),
    {
        displayName: 'Tag ID',
        name: 'tagId',
        description: 'ID of the tag to retrieve.',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: [
                    'tag',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('tag', 'get'),
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Return all results.',
        displayOptions: {
            show: {
                resource: [
                    'tag',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 50,
        description: 'The number of results to return.',
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                resource: [
                    'tag',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
    },
    SharedFields_1.makeSimpleField('tag', 'getAll'),
];
