// n8n-nodes-base/dist/nodes/Trello/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiRequestAllItems = exports.apiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function apiRequest(method, endpoint, body, query) {
    const credentials = await this.getCredentials('trelloApi');
    if (credentials === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials got returned!');
    }
    query = query || {};
    query.key = credentials.apiKey;
    query.token = credentials.apiToken;
    const options = {
        headers: {},
        method,
        body,
        qs: query,
        uri: `https://api.trello.com/1/${endpoint}`,
        json: true,
    };
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.apiRequest = apiRequest;
async function apiRequestAllItems(method, endpoint, body, query = {}) {
    query.limit = 30;
    query.sort = '-id';
    const returnData = [];
    let responseData;
    do {
        responseData = await apiRequest.call(this, method, endpoint, body, query);
        returnData.push.apply(returnData, responseData);
        if (responseData.length !== 0) {
            query.before = responseData[responseData.length - 1].id;
        }
    } while (query.limit <= responseData.length);
    return returnData;
}
exports.apiRequestAllItems = apiRequestAllItems;
