// n8n-nodes-base/dist/nodes/Strapi/Strapi.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Strapi = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
const EntryDescription_1 = require("./EntryDescription");
class Strapi {
    constructor() {
        this.description = {
            displayName: 'Strapi',
            name: 'strapi',
            icon: 'file:strapi.svg',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Strapi API',
            defaults: {
                name: 'Strapi',
                color: '#725ed8',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'strapiApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    options: [
                        {
                            name: 'Entry',
                            value: 'entry',
                        },
                    ],
                    default: 'entry',
                    description: 'The resource to operate on.',
                },
                ...EntryDescription_1.entryOperations,
                ...EntryDescription_1.entryFields,
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        const qs = {};
        const headers = {};
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        const { jwt } = await GenericFunctions_1.getToken.call(this);
        headers.Authorization = `Bearer ${jwt}`;
        for (let i = 0; i < length; i++) {
            try {
                if (resource === 'entry') {
                    if (operation === 'create') {
                        const body = {};
                        const contentType = this.getNodeParameter('contentType', i);
                        const columns = this.getNodeParameter('columns', i);
                        const columnList = columns.split(',').map(column => column.trim());
                        for (const key of Object.keys(items[i].json)) {
                            if (columnList.includes(key)) {
                                body[key] = items[i].json[key];
                            }
                        }
                        responseData = await GenericFunctions_1.strapiApiRequest.call(this, 'POST', `/${contentType}`, body, qs, undefined, headers);
                        returnData.push(responseData);
                    }
                    if (operation === 'delete') {
                        const contentType = this.getNodeParameter('contentType', i);
                        const entryId = this.getNodeParameter('entryId', i);
                        responseData = await GenericFunctions_1.strapiApiRequest.call(this, 'DELETE', `/${contentType}/${entryId}`, {}, qs, undefined, headers);
                        returnData.push(responseData);
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const contentType = this.getNodeParameter('contentType', i);
                        const options = this.getNodeParameter('options', i);
                        if (options.sort && options.sort.length !== 0) {
                            const sortFields = options.sort;
                            qs._sort = sortFields.join(',');
                        }
                        if (options.where) {
                            const query = GenericFunctions_1.validateJSON(options.where);
                            if (query !== undefined) {
                                qs._where = query;
                            }
                            else {
                                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Query must be a valid JSON');
                            }
                        }
                        if (options.publicationState) {
                            qs._publicationState = options.publicationState;
                        }
                        if (returnAll) {
                            responseData = await GenericFunctions_1.strapiApiRequestAllItems.call(this, 'GET', `/${contentType}`, {}, qs, headers);
                        }
                        else {
                            qs._limit = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.strapiApiRequest.call(this, 'GET', `/${contentType}`, {}, qs, undefined, headers);
                        }
                        returnData.push.apply(returnData, responseData);
                    }
                    if (operation === 'get') {
                        const contentType = this.getNodeParameter('contentType', i);
                        const entryId = this.getNodeParameter('entryId', i);
                        responseData = await GenericFunctions_1.strapiApiRequest.call(this, 'GET', `/${contentType}/${entryId}`, {}, qs, undefined, headers);
                        returnData.push(responseData);
                    }
                    if (operation === 'update') {
                        const body = {};
                        const contentType = this.getNodeParameter('contentType', i);
                        const columns = this.getNodeParameter('columns', i);
                        const updateKey = this.getNodeParameter('updateKey', i);
                        const columnList = columns.split(',').map(column => column.trim());
                        const entryId = items[i].json[updateKey];
                        for (const key of Object.keys(items[i].json)) {
                            if (columnList.includes(key)) {
                                body[key] = items[i].json[key];
                            }
                        }
                        responseData = await GenericFunctions_1.strapiApiRequest.call(this, 'PUT', `/${contentType}/${entryId}`, body, qs, undefined, headers);
                        returnData.push(responseData);
                    }
                }
            }
            catch (error) {
                if (this.continueOnFail()) {
                    returnData.push({ error: error.message });
                    continue;
                }
                throw error;
            }
        }
        return [this.helpers.returnJsonArray(returnData)];
    }
}
exports.Strapi = Strapi;
