// n8n/dist/services/hooks.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HooksService = void 0;
const typedi_1 = require("typedi");
const auth_service_1 = require("../auth/auth.service");
const user_repository_1 = require("../databases/repositories/user.repository");
const settings_repository_1 = require("../databases/repositories/settings.repository");
const workflow_repository_1 = require("../databases/repositories/workflow.repository");
const credentials_repository_1 = require("../databases/repositories/credentials.repository");
const authUser_repository_1 = require("../databases/repositories/authUser.repository");
const user_service_1 = require("../services/user.service");
let HooksService = class HooksService {
    constructor(userService, authService, userRepository, settingsRepository, workflowRepository, credentialsRepository, authUserRepository) {
        this.userService = userService;
        this.authService = authService;
        this.userRepository = userRepository;
        this.settingsRepository = settingsRepository;
        this.workflowRepository = workflowRepository;
        this.credentialsRepository = credentialsRepository;
        this.authUserRepository = authUserRepository;
    }
    async inviteUsers(owner, attributes) {
        return await this.userService.inviteUsers(owner, attributes);
    }
    issueCookie(res, user) {
        return this.authService.issueCookie(res, user);
    }
    async findOneUser(filter) {
        return await this.authUserRepository.findOne(filter);
    }
    async saveUser(user) {
        return await this.userRepository.save(user);
    }
    async updateSettings(filter, set) {
        return await this.settingsRepository.update(filter, set);
    }
    async workflowsCount(filter) {
        return await this.workflowRepository.count(filter);
    }
    async credentialsCount(filter) {
        return await this.credentialsRepository.count(filter);
    }
    async settingsCount(filter) {
        return await this.settingsRepository.count(filter);
    }
    async authMiddleware(req, res, next) {
        return await this.authService.authMiddleware(req, res, next);
    }
    dbCollections() {
        return {
            User: this.userRepository,
            Settings: this.settingsRepository,
            Credentials: this.credentialsRepository,
            Workflow: this.workflowRepository,
        };
    }
};
exports.HooksService = HooksService;
exports.HooksService = HooksService = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [user_service_1.UserService,
        auth_service_1.AuthService,
        user_repository_1.UserRepository,
        settings_repository_1.SettingsRepository,
        workflow_repository_1.WorkflowRepository,
        credentials_repository_1.CredentialsRepository,
        authUser_repository_1.AuthUserRepository])
], HooksService);
