// n8n/dist/services/orchestration.handler.base.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrchestrationHandlerService = void 0;
const typedi_1 = __importDefault(require("typedi"));
const redis_service_1 = require("./redis.service");
class OrchestrationHandlerService {
    constructor() {
        this.initialized = false;
        this.redisService = typedi_1.default.get(redis_service_1.RedisService);
    }
    async init() {
        await this.initSubscriber();
        this.initialized = true;
    }
    async initWithOptions(options) {
        await this.initSubscriber(options);
        this.initialized = true;
    }
    async shutdown() {
        var _a;
        await ((_a = this.redisSubscriber) === null || _a === void 0 ? void 0 : _a.destroy());
        this.initialized = false;
    }
}
exports.OrchestrationHandlerService = OrchestrationHandlerService;
