// n8n/dist/services/tag.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagService = void 0;
const tag_repository_1 = require("../databases/repositories/tag.repository");
const typedi_1 = require("typedi");
const GenericHelpers_1 = require("../GenericHelpers");
const ExternalHooks_1 = require("../ExternalHooks");
let TagService = class TagService {
    constructor(externalHooks, tagRepository) {
        this.externalHooks = externalHooks;
        this.tagRepository = tagRepository;
    }
    toEntity(attrs) {
        attrs.name = attrs.name.trim();
        return this.tagRepository.create(attrs);
    }
    async save(tag, actionKind) {
        await (0, GenericHelpers_1.validateEntity)(tag);
        const action = actionKind[0].toUpperCase() + actionKind.slice(1);
        await this.externalHooks.run(`tag.before${action}`, [tag]);
        const savedTag = this.tagRepository.save(tag, { transaction: false });
        await this.externalHooks.run(`tag.after${action}`, [tag]);
        return await savedTag;
    }
    async delete(id) {
        await this.externalHooks.run('tag.beforeDelete', [id]);
        const deleteResult = this.tagRepository.delete(id);
        await this.externalHooks.run('tag.afterDelete', [id]);
        return await deleteResult;
    }
    async getAll(options) {
        if (options === null || options === void 0 ? void 0 : options.withUsageCount) {
            const allTags = await this.tagRepository.find({
                select: ['id', 'name', 'createdAt', 'updatedAt'],
                relations: ['workflowMappings'],
            });
            return allTags.map(({ workflowMappings, ...rest }) => {
                return {
                    ...rest,
                    usageCount: workflowMappings.length,
                };
            });
        }
        return await this.tagRepository.find({
            select: ['id', 'name', 'createdAt', 'updatedAt'],
        });
    }
    async getById(id) {
        return await this.tagRepository.findOneOrFail({
            where: { id },
        });
    }
    sortByRequestOrder(tags, { requestOrder }) {
        const tagMap = tags.reduce((acc, tag) => {
            acc[tag.id] = tag;
            return acc;
        }, {});
        return requestOrder.map((tagId) => tagMap[tagId]);
    }
};
exports.TagService = TagService;
exports.TagService = TagService = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [ExternalHooks_1.ExternalHooks,
        tag_repository_1.TagRepository])
], TagService);
