// n8n/dist/CredentialsOverwrites.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialsOverwrites = void 0;
const typedi_1 = require("typedi");
const n8n_workflow_1 = require("n8n-workflow");
const config_1 = __importDefault(require("./config"));
const CredentialTypes_1 = require("./CredentialTypes");
const Logger_1 = require("./Logger");
let CredentialsOverwrites = class CredentialsOverwrites {
    constructor(credentialTypes, logger) {
        this.credentialTypes = credentialTypes;
        this.logger = logger;
        this.overwriteData = {};
        this.resolvedTypes = [];
        const data = config_1.default.getEnv('credentials.overwrite.data');
        const overwriteData = (0, n8n_workflow_1.jsonParse)(data, {
            errorMessage: 'The credentials-overwrite is not valid JSON.',
        });
        this.setData(overwriteData);
    }
    setData(overwriteData) {
        this.resolvedTypes.length = 0;
        this.overwriteData = overwriteData;
        for (const type in overwriteData) {
            const overwrites = this.getOverwrites(type);
            if (overwrites && Object.keys(overwrites).length) {
                this.overwriteData[type] = overwrites;
            }
        }
    }
    applyOverwrite(type, data) {
        const overwrites = this.get(type);
        if (overwrites === undefined) {
            return data;
        }
        const returnData = (0, n8n_workflow_1.deepCopy)(data);
        for (const key of Object.keys(overwrites)) {
            if ([null, undefined, ''].includes(returnData[key])) {
                returnData[key] = overwrites[key];
            }
        }
        return returnData;
    }
    getOverwrites(type) {
        if (this.resolvedTypes.includes(type)) {
            return this.overwriteData[type];
        }
        if (!this.credentialTypes.recognizes(type)) {
            this.logger.warn(`Unknown credential type ${type} in Credential overwrites`);
            return;
        }
        const credentialTypeData = this.credentialTypes.getByName(type);
        if (credentialTypeData.extends === undefined) {
            this.resolvedTypes.push(type);
            return this.overwriteData[type];
        }
        const overwrites = {};
        for (const credentialsTypeName of credentialTypeData.extends) {
            Object.assign(overwrites, this.getOverwrites(credentialsTypeName));
        }
        if (this.overwriteData[type] !== undefined) {
            Object.assign(overwrites, this.overwriteData[type]);
        }
        this.resolvedTypes.push(type);
        return overwrites;
    }
    get(name) {
        const parentTypes = this.credentialTypes.getParentTypes(name);
        return [name, ...parentTypes]
            .reverse()
            .map((type) => this.overwriteData[type])
            .filter((type) => !!type)
            .reduce((acc, current) => Object.assign(acc, current), {});
    }
    getAll() {
        return this.overwriteData;
    }
};
exports.CredentialsOverwrites = CredentialsOverwrites;
exports.CredentialsOverwrites = CredentialsOverwrites = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [CredentialTypes_1.CredentialTypes,
        Logger_1.Logger])
], CredentialsOverwrites);
