// n8n/dist/Queue.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
const typedi_1 = __importStar(require("typedi"));
const n8n_workflow_1 = require("n8n-workflow");
const ActiveExecutions_1 = require("./ActiveExecutions");
const config_1 = __importDefault(require("./config"));
let Queue = class Queue {
    constructor(activeExecutions) {
        this.activeExecutions = activeExecutions;
    }
    async init() {
        const { default: Bull } = await Promise.resolve().then(() => __importStar(require('bull')));
        const { RedisClientService } = await Promise.resolve().then(() => __importStar(require('./services/redis/redis-client.service')));
        const redisClientService = typedi_1.default.get(RedisClientService);
        const bullPrefix = config_1.default.getEnv('queue.bull.prefix');
        const prefix = redisClientService.toValidPrefix(bullPrefix);
        this.jobQueue = new Bull('jobs', {
            prefix,
            settings: config_1.default.get('queue.bull.settings'),
            createClient: (type) => redisClientService.createClient({ type: `${type}(bull)` }),
        });
        this.jobQueue.on('global:progress', (_jobId, progress) => {
            this.activeExecutions.resolveResponsePromise(progress.executionId, this.decodeWebhookResponse(progress.response));
        });
    }
    decodeWebhookResponse(response) {
        if (typeof response === 'object' &&
            typeof response.body === 'object' &&
            response.body['__@N8nEncodedBuffer@__']) {
            response.body = Buffer.from(response.body['__@N8nEncodedBuffer@__'], n8n_workflow_1.BINARY_ENCODING);
        }
        return response;
    }
    async add(jobData, jobOptions) {
        return await this.jobQueue.add(jobData, jobOptions);
    }
    async getJob(jobId) {
        return await this.jobQueue.getJob(jobId);
    }
    async getJobs(jobTypes) {
        return await this.jobQueue.getJobs(jobTypes);
    }
    async getInProgressExecutionIds() {
        const inProgressJobs = await this.getJobs(['active', 'waiting']);
        return new Set(inProgressJobs.map((job) => job.data.executionId));
    }
    async process(concurrency, fn) {
        return await this.jobQueue.process(concurrency, fn);
    }
    async ping() {
        return await this.jobQueue.client.ping();
    }
    async pause({ isLocal, doNotWaitActive, } = {}) {
        return await this.jobQueue.pause(isLocal, doNotWaitActive);
    }
    getBullObjectInstance() {
        if (this.jobQueue === undefined) {
            throw new n8n_workflow_1.ApplicationError('Queue is not initialized yet!');
        }
        return this.jobQueue;
    }
    async stopJob(job) {
        if (await job.isActive()) {
            await job.progress(-1);
            return true;
        }
        try {
            await job.remove();
            return true;
        }
        catch (e) {
            await job.progress(-1);
        }
        return false;
    }
};
exports.Queue = Queue;
exports.Queue = Queue = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [ActiveExecutions_1.ActiveExecutions])
], Queue);
