// n8n/dist/databases/repositories/usageMetrics.repository.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsageMetricsRepository = exports.UsageMetrics = void 0;
const config_1 = __importDefault(require("../../config"));
const typedi_1 = require("typedi");
const typeorm_1 = require("@n8n/typeorm");
let UsageMetrics = class UsageMetrics {
};
exports.UsageMetrics = UsageMetrics;
exports.UsageMetrics = UsageMetrics = __decorate([
    (0, typeorm_1.Entity)()
], UsageMetrics);
let UsageMetricsRepository = class UsageMetricsRepository extends typeorm_1.Repository {
    constructor(dataSource) {
        super(UsageMetrics, dataSource.manager);
    }
    toTableName(name) {
        const tablePrefix = config_1.default.getEnv('database.tablePrefix');
        let tableName = config_1.default.getEnv('database.type') === 'mysqldb'
            ? `\`${tablePrefix}${name}\``
            : `"${tablePrefix}${name}"`;
        const pgSchema = config_1.default.getEnv('database.postgresdb.schema');
        if (pgSchema !== 'public')
            tableName = [pgSchema, tablePrefix + name].join('.');
        return tableName;
    }
    async getLicenseRenewalMetrics() {
        const userTable = this.toTableName('user');
        const workflowTable = this.toTableName('workflow_entity');
        const credentialTable = this.toTableName('credentials_entity');
        const workflowStatsTable = this.toTableName('workflow_statistics');
        const [{ enabled_user_count: enabledUsers, total_user_count: totalUsers, active_workflow_count: activeWorkflows, total_workflow_count: totalWorkflows, total_credentials_count: totalCredentials, production_executions_count: productionExecutions, manual_executions_count: manualExecutions, },] = (await this.query(`
			SELECT
				(SELECT COUNT(*) FROM ${userTable} WHERE disabled = false) AS enabled_user_count,
				(SELECT COUNT(*) FROM ${userTable}) AS total_user_count,
				(SELECT COUNT(*) FROM ${workflowTable} WHERE active = true) AS active_workflow_count,
				(SELECT COUNT(*) FROM ${workflowTable}) AS total_workflow_count,
				(SELECT COUNT(*) FROM ${credentialTable}) AS total_credentials_count,
				(SELECT SUM(count) FROM ${workflowStatsTable} WHERE name IN ('production_success', 'production_error')) AS production_executions_count,
				(SELECT SUM(count) FROM ${workflowStatsTable} WHERE name IN ('manual_success', 'manual_error')) AS manual_executions_count;
		`));
        const toNumber = (value) => (typeof value === 'number' ? value : parseInt(value, 10)) || 0;
        return {
            enabledUsers: toNumber(enabledUsers),
            totalUsers: toNumber(totalUsers),
            activeWorkflows: toNumber(activeWorkflows),
            totalWorkflows: toNumber(totalWorkflows),
            totalCredentials: toNumber(totalCredentials),
            productionExecutions: toNumber(productionExecutions),
            manualExecutions: toNumber(manualExecutions),
        };
    }
};
exports.UsageMetricsRepository = UsageMetricsRepository;
exports.UsageMetricsRepository = UsageMetricsRepository = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [typeorm_1.DataSource])
], UsageMetricsRepository);
