// n8n-nodes-base/dist/nodes/Wise/descriptions/ProfileDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.profileFields = exports.profileOperations = void 0;
exports.profileOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        default: 'get',
        description: 'Operation to perform',
        options: [
            {
                name: 'Get',
                value: 'get',
            },
            {
                name: 'Get All',
                value: 'getAll',
            },
        ],
        displayOptions: {
            show: {
                resource: [
                    'profile',
                ],
            },
        },
    },
];
exports.profileFields = [
    {
        displayName: 'Profile ID',
        name: 'profileId',
        type: 'options',
        required: true,
        default: [],
        typeOptions: {
            loadOptionsMethod: 'getProfiles',
        },
        description: 'ID of the user profile to retrieve.',
        displayOptions: {
            show: {
                resource: [
                    'profile',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
];
