// n8n-nodes-base/dist/nodes/Twilio/Twilio.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Twilio = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
class Twilio {
    constructor() {
        this.description = {
            displayName: 'Twilio',
            name: 'twilio',
            icon: 'file:twilio.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Send SMS and WhatsApp messages or make phone calls',
            defaults: {
                name: 'Twilio',
                color: '#cf272d',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'twilioApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    options: [
                        {
                            name: 'SMS',
                            value: 'sms',
                        },
                    ],
                    default: 'sms',
                    description: 'The resource to operate on.',
                },
                {
                    displayName: 'Operation',
                    name: 'operation',
                    type: 'options',
                    displayOptions: {
                        show: {
                            resource: [
                                'sms',
                            ],
                        },
                    },
                    options: [
                        {
                            name: 'Send',
                            value: 'send',
                            description: 'Send SMS/MMS/WhatsApp message',
                        },
                    ],
                    default: 'send',
                    description: 'The operation to perform.',
                },
                {
                    displayName: 'From',
                    name: 'from',
                    type: 'string',
                    default: '',
                    placeholder: '+14155238886',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                            ],
                        },
                    },
                    description: 'The number from which to send the message',
                },
                {
                    displayName: 'To',
                    name: 'to',
                    type: 'string',
                    default: '',
                    placeholder: '+14155238886',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                            ],
                        },
                    },
                    description: 'The number to which to send the message',
                },
                {
                    displayName: 'To Whatsapp',
                    name: 'toWhatsapp',
                    type: 'boolean',
                    default: false,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                            ],
                        },
                    },
                    description: 'If the message should be send to WhatsApp',
                },
                {
                    displayName: 'Message',
                    name: 'message',
                    type: 'string',
                    default: '',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'send',
                            ],
                            resource: [
                                'sms',
                            ],
                        },
                    },
                    description: 'The message to send',
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        let operation;
        let resource;
        let body;
        let qs;
        let requestMethod;
        let endpoint;
        for (let i = 0; i < items.length; i++) {
            try {
                requestMethod = 'GET';
                endpoint = '';
                body = {};
                qs = {};
                resource = this.getNodeParameter('resource', i);
                operation = this.getNodeParameter('operation', i);
                if (resource === 'sms') {
                    if (operation === 'send') {
                        requestMethod = 'POST';
                        endpoint = '/Messages.json';
                        body.From = this.getNodeParameter('from', i);
                        body.To = this.getNodeParameter('to', i);
                        body.Body = this.getNodeParameter('message', i);
                        const toWhatsapp = this.getNodeParameter('toWhatsapp', i);
                        if (toWhatsapp === true) {
                            body.From = `whatsapp:${body.From}`;
                            body.To = `whatsapp:${body.To}`;
                        }
                    }
                    else {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `The operation "${operation}" is not known!`);
                    }
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `The resource "${resource}" is not known!`);
                }
                const responseData = await GenericFunctions_1.twilioApiRequest.call(this, requestMethod, endpoint, body, qs);
                returnData.push(responseData);
            }
            catch (error) {
                if (this.continueOnFail()) {
                    returnData.push({ error: error.message });
                    continue;
                }
                throw error;
            }
        }
        return [this.helpers.returnJsonArray(returnData)];
    }
}
exports.Twilio = Twilio;
