// n8n-nodes-base/dist/nodes/Elastic/ElasticSecurity/descriptions/CaseTagDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.caseTagFields = exports.caseTagOperations = void 0;
exports.caseTagOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: [
                    'caseTag',
                ],
            },
        },
        options: [
            {
                name: 'Add',
                value: 'add',
                description: 'Add a tag to a case',
            },
            {
                name: 'Remove',
                value: 'remove',
                description: 'Remove a tag from a case',
            },
        ],
        default: 'add',
    },
];
exports.caseTagFields = [
    {
        displayName: 'Case ID',
        name: 'caseId',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'caseTag',
                ],
                operation: [
                    'add',
                ],
            },
        },
    },
    {
        displayName: 'Tag',
        name: 'tag',
        type: 'options',
        description: 'Tag to attach to the case. Choose from the list or enter a new one with an expression.',
        required: true,
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getTags',
        },
        displayOptions: {
            show: {
                resource: [
                    'caseTag',
                ],
                operation: [
                    'add',
                ],
            },
        },
    },
    {
        displayName: 'Case ID',
        name: 'caseId',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'caseTag',
                ],
                operation: [
                    'remove',
                ],
            },
        },
    },
    {
        displayName: 'Tag',
        name: 'tag',
        type: 'options',
        required: true,
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getTags',
        },
        displayOptions: {
            show: {
                resource: [
                    'caseTag',
                ],
                operation: [
                    'remove',
                ],
            },
        },
    },
];
