// n8n-nodes-base/dist/nodes/Mattermost/v1/transport/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiRequestAllItems = exports.apiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function apiRequest(method, endpoint, body = {}, query = {}) {
    const credentials = await this.getCredentials('mattermostApi');
    if (!credentials) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials returned!');
    }
    const options = {
        method,
        body,
        qs: query,
        url: `${credentials.baseUrl}/api/v4/${endpoint}`,
        headers: {
            authorization: `Bearer ${credentials.accessToken}`,
            'content-type': 'application/json; charset=utf-8',
        },
    };
    try {
        return await this.helpers.httpRequest(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.apiRequest = apiRequest;
async function apiRequestAllItems(method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query.page = 0;
    query.per_page = 100;
    do {
        responseData = await apiRequest.call(this, method, endpoint, body, query);
        query.page++;
        returnData.push.apply(returnData, responseData);
    } while (responseData.length !== 0);
    return returnData;
}
exports.apiRequestAllItems = apiRequestAllItems;
