// n8n-nodes-base/dist/nodes/N8nTrigger/N8nTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.N8nTrigger = void 0;
class N8nTrigger {
    constructor() {
        this.description = {
            displayName: 'n8n Trigger',
            name: 'n8nTrigger',
            icon: 'file:n8nTrigger.svg',
            group: ['trigger'],
            version: 1,
            description: 'Handle events from your n8n instance',
            defaults: {
                name: 'n8n Trigger',
                color: '#ff6d5a',
            },
            inputs: [],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'Events',
                    name: 'events',
                    type: 'multiOptions',
                    required: true,
                    default: [],
                    description: 'Specifies under which conditions an execution should happen:<br />' +
                        '- <b>Instance started</b>: Triggers when this n8n instance is started or re-started',
                    options: [
                        {
                            name: 'Instance started',
                            value: 'init',
                            description: 'Triggers when this n8n instance is started or re-started',
                        },
                    ],
                },
            ],
        };
    }
    async trigger() {
        const events = this.getNodeParameter('events', []);
        const activationMode = this.getActivationMode();
        if (events.includes(activationMode)) {
            let event;
            if (activationMode === 'init') {
                event = 'Instance started';
            }
            this.emit([
                this.helpers.returnJsonArray([
                    { event, timestamp: (new Date()).toISOString(), workflow_id: this.getWorkflow().id },
                ]),
            ]);
        }
        const self = this;
        async function manualTriggerFunction() {
            self.emit([self.helpers.returnJsonArray([{ event: 'Manual execution', timestamp: (new Date()).toISOString(), workflow_id: self.getWorkflow().id }])]);
        }
        return {
            manualTriggerFunction,
        };
    }
}
exports.N8nTrigger = N8nTrigger;
