// n8n-nodes-base/dist/nodes/Google/Translate/GoogleTranslate.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleTranslate = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
class GoogleTranslate {
    constructor() {
        this.description = {
            displayName: 'Google Translate',
            name: 'googleTranslate',
            icon: 'file:googletranslate.png',
            group: ['input', 'output'],
            version: 1,
            description: 'Translate data using Google Translate',
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            defaults: {
                name: 'Google Translate',
                color: '#5390f5',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'googleApi',
                    required: true,
                    displayOptions: {
                        show: {
                            authentication: [
                                'serviceAccount',
                            ],
                        },
                    },
                },
                {
                    name: 'googleTranslateOAuth2Api',
                    required: true,
                    displayOptions: {
                        show: {
                            authentication: [
                                'oAuth2',
                            ],
                        },
                    },
                },
            ],
            properties: [
                {
                    displayName: 'Authentication',
                    name: 'authentication',
                    type: 'options',
                    options: [
                        {
                            name: 'Service Account',
                            value: 'serviceAccount',
                        },
                        {
                            name: 'OAuth2',
                            value: 'oAuth2',
                        },
                    ],
                    default: 'serviceAccount',
                },
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    options: [
                        {
                            name: 'Language',
                            value: 'language',
                        },
                    ],
                    default: 'language',
                    description: 'The operation to perform',
                },
                {
                    displayName: 'Operation',
                    name: 'operation',
                    type: 'options',
                    displayOptions: {
                        show: {
                            resource: [
                                'language',
                            ],
                        },
                    },
                    options: [
                        {
                            name: 'Translate',
                            value: 'translate',
                            description: 'Translate data',
                        },
                    ],
                    default: 'translate',
                    description: 'The operation to perform',
                },
                {
                    displayName: 'Text',
                    name: 'text',
                    type: 'string',
                    default: '',
                    description: 'The input text to translate',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'translate',
                            ],
                        },
                    },
                },
                {
                    displayName: 'Translate To',
                    name: 'translateTo',
                    type: 'options',
                    typeOptions: {
                        loadOptionsMethod: 'getLanguages',
                    },
                    default: '',
                    description: 'The language to use for translation of the input text, set to one of the language codes listed in <a href="https://cloud.google.com/translate/docs/languages">Language Support</a>',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: [
                                'translate',
                            ],
                        },
                    },
                },
            ],
        };
        this.methods = {
            loadOptions: {
                async getLanguages() {
                    const returnData = [];
                    const { data: { languages } } = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/language/translate/v2/languages');
                    for (const language of languages) {
                        returnData.push({
                            name: language.language.toUpperCase(),
                            value: language.language,
                        });
                    }
                    return returnData;
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const length = items.length;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        const responseData = [];
        for (let i = 0; i < length; i++) {
            if (resource === 'language') {
                if (operation === 'translate') {
                    const text = this.getNodeParameter('text', i);
                    const translateTo = this.getNodeParameter('translateTo', i);
                    const response = await GenericFunctions_1.googleApiRequest.call(this, 'POST', `/language/translate/v2`, { q: text, target: translateTo });
                    responseData.push(response.data.translations[0]);
                }
            }
        }
        return [this.helpers.returnJsonArray(responseData)];
    }
}
exports.GoogleTranslate = GoogleTranslate;
