// n8n-nodes-base/dist/nodes/Microsoft/OneDrive/MicrosoftOneDriveTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MicrosoftOneDriveTrigger = void 0;
const luxon_1 = require("luxon");
const TriggerDescription_1 = require("./TriggerDescription");
const GenericFunctions_1 = require("./GenericFunctions");
class MicrosoftOneDriveTrigger {
    constructor() {
        this.description = {
            displayName: 'Microsoft OneDrive Trigger',
            name: 'microsoftOneDriveTrigger',
            icon: 'file:oneDrive.svg',
            group: ['trigger'],
            version: 1,
            description: 'Trigger for Microsoft OneDrive API.',
            subtitle: '={{($parameter["event"])}}',
            defaults: {
                name: 'Microsoft OneDrive Trigger',
            },
            credentials: [
                {
                    name: 'microsoftOneDriveOAuth2Api',
                    required: true,
                },
            ],
            polling: true,
            inputs: [],
            outputs: ['main'],
            properties: [...TriggerDescription_1.triggerDescription],
        };
        this.methods = {
            loadOptions: {},
        };
    }
    async poll() {
        const workflowData = this.getWorkflowStaticData('node');
        let responseData;
        const lastLink = workflowData.LastLink ||
            'https://graph.microsoft.com/v1.0/me/drive/root/delta?token=latest';
        const now = luxon_1.DateTime.now().toUTC();
        const start = luxon_1.DateTime.fromISO(workflowData.lastTimeChecked) || now;
        const end = now;
        const event = this.getNodeParameter('event', 'fileCreated');
        const watch = this.getNodeParameter('watch', 'anyFile');
        const watchFolder = this.getNodeParameter('watchFolder', false) || false;
        const folderChild = this.getNodeParameter('options.folderChild', false) || false;
        let eventType = 'created';
        let eventResource = 'file';
        if (event.includes('Updated')) {
            eventType = 'updated';
        }
        if (event.includes('folder')) {
            eventResource = 'folder';
        }
        try {
            if (this.getMode() === 'manual') {
                responseData = (await GenericFunctions_1.microsoftApiRequest.call(this, 'GET', '', {}, {}, 'https://graph.microsoft.com/v1.0/me/drive/root/delta')).value;
            }
            else {
                const response = (await GenericFunctions_1.microsoftApiRequestAllItemsDelta.call(this, lastLink, start, eventType));
                responseData = response.returnData;
                workflowData.LastLink = response.deltaLink;
            }
            workflowData.lastTimeChecked = end.toISO();
            if (watch === 'selectedFile') {
                const fileId = this.getNodeParameter('fileId', '', {
                    extractValue: true,
                }).replace('%21', '!');
                if (fileId) {
                    responseData = responseData.filter((item) => item.id === fileId);
                }
            }
            if (!folderChild &&
                (watch === 'oneSelectedFolder' || watch === 'selectedFolder' || watchFolder)) {
                const folderId = this.getNodeParameter('folderId', '', {
                    extractValue: true,
                }).replace('%21', '!');
                if (folderId) {
                    if (watch === 'oneSelectedFolder') {
                        responseData = responseData.filter((item) => item.id === folderId);
                    }
                    else {
                        responseData = responseData.filter((item) => item.parentReference.id === folderId);
                    }
                }
            }
            if (folderChild && (watch === 'selectedFolder' || watchFolder)) {
                const folderId = this.getNodeParameter('folderId', '', {
                    extractValue: true,
                }).replace('%21', '!');
                const folderPath = await GenericFunctions_1.getPath.call(this, folderId);
                responseData = responseData.filter((item) => {
                    var _a;
                    const path = (_a = item.parentReference) === null || _a === void 0 ? void 0 : _a.path;
                    return typeof path === 'string' && path.startsWith(folderPath);
                });
            }
            responseData = responseData.filter((item) => item[eventResource]);
            if (!(responseData === null || responseData === void 0 ? void 0 : responseData.length)) {
                return null;
            }
            const simplify = this.getNodeParameter('simple');
            if (simplify) {
                responseData = responseData.map((x) => {
                    var _a, _b, _c, _d;
                    return ({
                        id: x.id,
                        createdDateTime: (_a = x.fileSystemInfo) === null || _a === void 0 ? void 0 : _a.createdDateTime,
                        lastModifiedDateTime: (_b = x.fileSystemInfo) === null || _b === void 0 ? void 0 : _b.lastModifiedDateTime,
                        name: x.name,
                        webUrl: x.webUrl,
                        size: x.size,
                        path: ((_c = x.parentReference) === null || _c === void 0 ? void 0 : _c.path) || '',
                        mimeType: ((_d = x.file) === null || _d === void 0 ? void 0 : _d.mimeType) || '',
                    });
                });
            }
            return [this.helpers.returnJsonArray(responseData)];
        }
        catch (error) {
            if (this.getMode() === 'manual' || !workflowData.lastTimeChecked) {
                throw error;
            }
            const workflow = this.getWorkflow();
            const node = this.getNode();
            this.logger.error(`There was a problem in '${node.name}' node in workflow '${workflow.id}': '${error.description}'`, {
                node: node.name,
                workflowId: workflow.id,
                error,
            });
            throw error;
        }
        return null;
    }
}
exports.MicrosoftOneDriveTrigger = MicrosoftOneDriveTrigger;
