// n8n/dist/license/license.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicenseController = void 0;
const decorators_1 = require("../decorators");
const license_service_1 = require("./license.service");
let LicenseController = class LicenseController {
    constructor(licenseService) {
        this.licenseService = licenseService;
    }
    async getLicenseData() {
        return await this.licenseService.getLicenseData();
    }
    async requestEnterpriseTrial(req) {
        await this.licenseService.requestEnterpriseTrial(req.user);
    }
    async activateLicense(req) {
        const { activationKey } = req.body;
        await this.licenseService.activateLicense(activationKey);
        return await this.getTokenAndData();
    }
    async renewLicense() {
        await this.licenseService.renewLicense();
        return await this.getTokenAndData();
    }
    async getTokenAndData() {
        const managementToken = this.licenseService.getManagementJwt();
        const data = await this.licenseService.getLicenseData();
        return { ...data, managementToken };
    }
};
exports.LicenseController = LicenseController;
__decorate([
    (0, decorators_1.Get)('/'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], LicenseController.prototype, "getLicenseData", null);
__decorate([
    (0, decorators_1.Post)('/enterprise/request_trial'),
    (0, decorators_1.GlobalScope)('license:manage'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], LicenseController.prototype, "requestEnterpriseTrial", null);
__decorate([
    (0, decorators_1.Post)('/activate'),
    (0, decorators_1.GlobalScope)('license:manage'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], LicenseController.prototype, "activateLicense", null);
__decorate([
    (0, decorators_1.Post)('/renew'),
    (0, decorators_1.GlobalScope)('license:manage'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], LicenseController.prototype, "renewLicense", null);
exports.LicenseController = LicenseController = __decorate([
    (0, decorators_1.RestController)('/license'),
    __metadata("design:paramtypes", [license_service_1.LicenseService])
], LicenseController);
