// n8n-nodes-base/dist/nodes/FacebookLeadAds/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.facebookFormList = exports.installAppOnPage = exports.facebookPageApiRequest = exports.facebookEntityDetail = exports.facebookPageList = exports.appWebhookSubscriptionDelete = exports.appWebhookSubscriptionCreate = exports.appWebhookSubscriptionList = exports.facebookAppApiRequest = exports.appAccessTokenRead = exports.facebookApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function facebookApiRequest(method, resource, body = {}, qs = {}) {
    var _a, _b;
    const options = {
        headers: {
            accept: 'application/json',
        },
        method,
        qs,
        body,
        gzip: true,
        uri: `https://graph.facebook.com/v17.0${resource}`,
        json: true,
    };
    try {
        return await this.helpers.requestOAuth2.call(this, 'facebookLeadAdsOAuth2Api', options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error, {
            message: (_b = (_a = error === null || error === void 0 ? void 0 : error.error) === null || _a === void 0 ? void 0 : _a.error) === null || _b === void 0 ? void 0 : _b.message,
        });
    }
}
exports.facebookApiRequest = facebookApiRequest;
async function appAccessTokenRead() {
    const credentials = await this.getCredentials('facebookLeadAdsOAuth2Api');
    const options = {
        headers: {
            'content-type': 'application/x-www-form-urlencoded',
        },
        method: 'POST',
        form: {
            client_id: credentials.clientId,
            client_secret: credentials.clientSecret,
            grant_type: 'client_credentials',
        },
        uri: credentials.accessTokenUrl,
        json: true,
    };
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.appAccessTokenRead = appAccessTokenRead;
async function facebookAppApiRequest(method, resource, body, qs = {}) {
    const tokenResponse = await appAccessTokenRead.call(this);
    const appAccessToken = tokenResponse.access_token;
    const options = {
        headers: {
            accept: 'application/json',
            authorization: `Bearer ${appAccessToken}`,
        },
        method,
        qs,
        gzip: true,
        uri: `https://graph.facebook.com/v17.0${resource}`,
        json: true,
    };
    if ((body === null || body === void 0 ? void 0 : body.type) === 'json') {
        options.body = body.payload;
    }
    else if ((body === null || body === void 0 ? void 0 : body.type) === 'form') {
        options.form = body.payload;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.facebookAppApiRequest = facebookAppApiRequest;
async function appWebhookSubscriptionList(appId) {
    const response = (await facebookAppApiRequest.call(this, 'GET', `/${appId}/subscriptions`));
    return response.data;
}
exports.appWebhookSubscriptionList = appWebhookSubscriptionList;
async function appWebhookSubscriptionCreate(appId, subscription) {
    return await facebookAppApiRequest.call(this, 'POST', `/${appId}/subscriptions`, {
        type: 'form',
        payload: { ...subscription },
    });
}
exports.appWebhookSubscriptionCreate = appWebhookSubscriptionCreate;
async function appWebhookSubscriptionDelete(appId, object) {
    return await facebookAppApiRequest.call(this, 'DELETE', `/${appId}/subscriptions`, {
        type: 'form',
        payload: { object },
    });
}
exports.appWebhookSubscriptionDelete = appWebhookSubscriptionDelete;
async function facebookPageList(cursor) {
    const response = (await facebookApiRequest.call(this, 'GET', '/me/accounts', {}, { cursor, fields: 'id,name' }));
    return response;
}
exports.facebookPageList = facebookPageList;
async function facebookEntityDetail(entityId, fields = 'id,name,access_token') {
    return await facebookApiRequest.call(this, 'GET', `/${entityId}`, {}, { fields });
}
exports.facebookEntityDetail = facebookEntityDetail;
async function facebookPageApiRequest(method, resource, body = {}, qs = {}) {
    const pageId = this.getNodeParameter('page', '', { extractValue: true });
    const page = (await facebookEntityDetail.call(this, pageId));
    const pageAccessToken = page.access_token;
    const options = {
        headers: {
            accept: 'application/json',
            authorization: `Bearer ${pageAccessToken}`,
        },
        method,
        qs,
        body,
        gzip: true,
        uri: `https://graph.facebook.com/v17.0${resource}`,
        json: true,
    };
    try {
        return await this.helpers.request.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.facebookPageApiRequest = facebookPageApiRequest;
async function installAppOnPage(pageId, fields) {
    return await facebookPageApiRequest.call(this, 'POST', `/${pageId}/subscribed_apps`, {}, { subscribed_fields: fields });
}
exports.installAppOnPage = installAppOnPage;
async function facebookFormList(pageId, cursor) {
    const response = (await facebookPageApiRequest.call(this, 'GET', `/${pageId}/leadgen_forms`, {}, { cursor, fields: 'id,name' }));
    return response;
}
exports.facebookFormList = facebookFormList;
