// n8n-nodes-base/dist/nodes/Google/Drive/v2/helpers/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DRIVE = exports.RLC_FOLDER_DEFAULT = exports.RLC_DRIVE_DEFAULT = exports.UPLOAD_CHUNK_SIZE = void 0;
exports.UPLOAD_CHUNK_SIZE = 256 * 1024;
exports.RLC_DRIVE_DEFAULT = 'My Drive';
exports.RLC_FOLDER_DEFAULT = 'root';
var DRIVE;
(function (DRIVE) {
    DRIVE["FOLDER"] = "application/vnd.google-apps.folder";
    DRIVE["AUDIO"] = "application/vnd.google-apps.audio";
    DRIVE["DOCUMENT"] = "application/vnd.google-apps.document";
    DRIVE["SDK"] = "application/vnd.google-apps.drive-sdk";
    DRIVE["DRAWING"] = "application/vnd.google-apps.drawing";
    DRIVE["FILE"] = "application/vnd.google-apps.file";
    DRIVE["FORM"] = "application/vnd.google-apps.form";
    DRIVE["FUSIONTABLE"] = "application/vnd.google-apps.fusiontable";
    DRIVE["MAP"] = "application/vnd.google-apps.map";
    DRIVE["PHOTO"] = "application/vnd.google-apps.photo";
    DRIVE["PRESENTATION"] = "application/vnd.google-apps.presentation";
    DRIVE["APP_SCRIPTS"] = "application/vnd.google-apps.script";
    DRIVE["SITES"] = "application/vnd.google-apps.sites";
    DRIVE["SPREADSHEET"] = "application/vnd.google-apps.spreadsheet";
    DRIVE["UNKNOWN"] = "application/vnd.google-apps.unknown";
    DRIVE["VIDEO"] = "application/vnd.google-apps.video";
})(DRIVE || (exports.DRIVE = DRIVE = {}));
