// n8n-nodes-base/dist/nodes/Google/Drive/v2/helpers/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setParentFolder = exports.updateDriveScopes = exports.setUpdateCommonParams = exports.setFileProperties = exports.getItemBinaryData = exports.prepareQueryString = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const interfaces_1 = require("./interfaces");
function prepareQueryString(fields) {
    let queryFields = 'id, name';
    if (fields) {
        if (fields.includes('*')) {
            queryFields = '*';
        }
        else {
            queryFields = fields.join(', ');
        }
    }
    return queryFields;
}
exports.prepareQueryString = prepareQueryString;
async function getItemBinaryData(inputDataFieldName, i, chunkSize = interfaces_1.UPLOAD_CHUNK_SIZE) {
    let contentLength;
    let fileContent;
    let originalFilename;
    let mimeType;
    if (!inputDataFieldName) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'The name of the input field containing the binary file data must be set', {
            itemIndex: i,
        });
    }
    const binaryData = this.helpers.assertBinaryData(i, inputDataFieldName);
    if (binaryData.id) {
        fileContent = await this.helpers.getBinaryStream(binaryData.id, chunkSize);
        const metadata = await this.helpers.getBinaryMetadata(binaryData.id);
        contentLength = metadata.fileSize;
        originalFilename = metadata.fileName;
        if (metadata.mimeType)
            mimeType = binaryData.mimeType;
    }
    else {
        fileContent = Buffer.from(binaryData.data, n8n_workflow_1.BINARY_ENCODING);
        contentLength = fileContent.length;
        originalFilename = binaryData.fileName;
        mimeType = binaryData.mimeType;
    }
    return {
        contentLength,
        fileContent,
        originalFilename,
        mimeType,
    };
}
exports.getItemBinaryData = getItemBinaryData;
function setFileProperties(body, options) {
    if (options.propertiesUi) {
        const values = options.propertiesUi.propertyValues || [];
        body.properties = values.reduce((acc, value) => Object.assign(acc, { [`${value.key}`]: value.value }), {});
    }
    if (options.appPropertiesUi) {
        const values = options.appPropertiesUi.appPropertyValues || [];
        body.appProperties = values.reduce((acc, value) => Object.assign(acc, { [`${value.key}`]: value.value }), {});
    }
    return body;
}
exports.setFileProperties = setFileProperties;
function setUpdateCommonParams(qs, options) {
    if (options.keepRevisionForever) {
        qs.keepRevisionForever = options.keepRevisionForever;
    }
    if (options.ocrLanguage) {
        qs.ocrLanguage = options.ocrLanguage;
    }
    if (options.useContentAsIndexableText) {
        qs.useContentAsIndexableText = options.useContentAsIndexableText;
    }
    return qs;
}
exports.setUpdateCommonParams = setUpdateCommonParams;
function updateDriveScopes(qs, driveId, defaultDrive = interfaces_1.RLC_DRIVE_DEFAULT) {
    if (driveId) {
        if (driveId === defaultDrive) {
            qs.includeItemsFromAllDrives = false;
            qs.supportsAllDrives = false;
            qs.spaces = 'appDataFolder, drive';
            qs.corpora = 'user';
        }
        else {
            qs.driveId = driveId;
            qs.corpora = 'drive';
        }
    }
}
exports.updateDriveScopes = updateDriveScopes;
function setParentFolder(folderId, driveId, folderIdDefault = interfaces_1.RLC_FOLDER_DEFAULT, driveIdDefault = interfaces_1.RLC_DRIVE_DEFAULT) {
    if (folderId !== folderIdDefault) {
        return folderId;
    }
    else if (driveId && driveId !== driveIdDefault) {
        return driveId;
    }
    else {
        return 'root';
    }
}
exports.setParentFolder = setParentFolder;
