// n8n-nodes-base/dist/nodes/Bannerbear/TemplateDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.templateFields = exports.templateOperations = void 0;
exports.templateOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'template',
                ],
            },
        },
        options: [
            {
                name: 'Get',
                value: 'get',
                description: 'Get a template',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Get all templates',
            },
        ],
        default: 'get',
        description: 'The operation to perform.',
    },
];
exports.templateFields = [
    {
        displayName: 'Template ID',
        name: 'templateId',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'template',
                ],
                operation: [
                    'get',
                ],
            },
        },
        description: 'Unique identifier for the template.',
    },
];
