// n8n-nodes-base/dist/nodes/ItemLists/V3/actions/versionDescription.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const itemList = __importStar(require("./itemList"));
exports.versionDescription = {
    displayName: 'Item Lists',
    name: 'itemLists',
    icon: 'file:itemLists.svg',
    group: ['input'],
    subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
    description: 'Helper for working with lists of items and transforming arrays',
    version: [3, 3.1],
    defaults: {
        name: 'Item Lists',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [],
    properties: [
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'hidden',
            options: [
                {
                    name: 'Item List',
                    value: 'itemList',
                },
            ],
            default: 'itemList',
        },
        ...itemList.description,
    ],
};
