// n8n-nodes-base/dist/nodes/ActiveCampaign/ContactListDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contactListFields = exports.contactListOperations = void 0;
exports.contactListOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'contactList',
                ],
            },
        },
        options: [
            {
                name: 'Add',
                value: 'add',
                description: 'Add contact to a list',
            },
            {
                name: 'Remove',
                value: 'remove',
                description: 'Remove contact from a list',
            },
        ],
        default: 'add',
        description: 'The operation to perform.',
    },
];
exports.contactListFields = [
    {
        displayName: 'List ID',
        name: 'listId',
        type: 'number',
        default: '',
        required: true,
        displayOptions: {
            show: {
                operation: [
                    'add',
                ],
                resource: [
                    'contactList',
                ],
            },
        },
        description: 'List ID',
    },
    {
        displayName: 'Contact ID',
        name: 'contactId',
        type: 'number',
        default: '',
        required: true,
        displayOptions: {
            show: {
                operation: [
                    'add',
                ],
                resource: [
                    'contactList',
                ],
            },
        },
        description: 'Contact ID',
    },
    {
        displayName: 'List ID',
        name: 'listId',
        type: 'number',
        default: '',
        required: true,
        displayOptions: {
            show: {
                operation: [
                    'remove',
                ],
                resource: [
                    'contactList',
                ],
            },
        },
        description: 'List ID',
    },
    {
        displayName: 'Contact ID',
        name: 'contactId',
        type: 'number',
        default: '',
        required: true,
        displayOptions: {
            show: {
                operation: [
                    'remove',
                ],
                resource: [
                    'contactList',
                ],
            },
        },
        description: 'Contact ID',
    },
];
