// n8n-nodes-base/dist/nodes/Airtable/v2/methods/loadOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAttachmentColumns = exports.getColumnsWithoutColumnToMatchOn = exports.getColumnsWithRecordId = exports.getColumns = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const transport_1 = require("../transport");
async function getColumns() {
    const base = this.getNodeParameter('base', undefined, {
        extractValue: true,
    });
    const tableId = encodeURI(this.getNodeParameter('table', undefined, {
        extractValue: true,
    }));
    const response = await transport_1.apiRequest.call(this, 'GET', `meta/bases/${base}/tables`);
    const tableData = (response.tables || []).find((table) => {
        return table.id === tableId;
    });
    if (!tableData) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Table information could not be found!', {
            level: 'warning',
        });
    }
    const result = [];
    for (const field of tableData.fields) {
        result.push({
            name: field.name,
            value: field.name,
            description: `Type: ${field.type}`,
        });
    }
    return result;
}
exports.getColumns = getColumns;
async function getColumnsWithRecordId() {
    const returnData = await getColumns.call(this);
    return [
        {
            name: 'id',
            value: 'id',
            description: 'Type: primaryFieldId',
        },
        ...returnData,
    ];
}
exports.getColumnsWithRecordId = getColumnsWithRecordId;
async function getColumnsWithoutColumnToMatchOn() {
    const columnToMatchOn = this.getNodeParameter('columnToMatchOn');
    const returnData = await getColumns.call(this);
    return returnData.filter((column) => column.value !== columnToMatchOn);
}
exports.getColumnsWithoutColumnToMatchOn = getColumnsWithoutColumnToMatchOn;
async function getAttachmentColumns() {
    var _a, _b;
    const base = this.getNodeParameter('base', undefined, {
        extractValue: true,
    });
    const tableId = encodeURI(this.getNodeParameter('table', undefined, {
        extractValue: true,
    }));
    const response = await transport_1.apiRequest.call(this, 'GET', `meta/bases/${base}/tables`);
    const tableData = (response.tables || []).find((table) => {
        return table.id === tableId;
    });
    if (!tableData) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Table information could not be found!', {
            level: 'warning',
        });
    }
    const result = [];
    for (const field of tableData.fields) {
        if (!((_b = (_a = field.type) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === null || _b === void 0 ? void 0 : _b.includes('attachment'))) {
            continue;
        }
        result.push({
            name: field.name,
            value: field.name,
            description: `Type: ${field.type}`,
        });
    }
    return result;
}
exports.getAttachmentColumns = getAttachmentColumns;
