// n8n-nodes-base/dist/nodes/Uplead/CompanyDesciption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.companyFields = exports.companyOperations = void 0;
exports.companyOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'company',
                ],
            },
        },
        options: [
            {
                name: 'Enrich',
                value: 'enrich',
            },
        ],
        default: 'enrich',
        description: 'The operation to perform.',
    },
];
exports.companyFields = [
    {
        displayName: 'Company',
        name: 'company',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'company',
                ],
                operation: [
                    'enrich',
                ],
            },
        },
        description: 'the name of the company (e.g – amazon)',
    },
    {
        displayName: 'Domain',
        name: 'domain',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: [
                    'company',
                ],
                operation: [
                    'enrich',
                ],
            },
        },
        description: 'the domain name (e.g – amazon.com)',
    },
];
