// n8n-nodes-base/dist/nodes/SpreadsheetFile/v2/SpreadsheetFileV2.node.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpreadsheetFileV2 = void 0;
const description_1 = require("../description");
const fromFile = __importStar(require("./fromFile.operation"));
const toFile = __importStar(require("./toFile.operation"));
class SpreadsheetFileV2 {
    constructor(baseDescription) {
        this.description = {
            ...baseDescription,
            version: 2,
            defaults: {
                name: 'Spreadsheet File',
                color: '#2244FF',
            },
            inputs: ['main'],
            outputs: ['main'],
            properties: [description_1.operationProperty, ...fromFile.description, ...toFile.description],
        };
    }
    async execute() {
        const items = this.getInputData();
        const operation = this.getNodeParameter('operation', 0);
        let returnData = [];
        if (operation === 'fromFile') {
            returnData = await fromFile.execute.call(this, items);
        }
        if (operation === 'toFile') {
            returnData = await toFile.execute.call(this, items);
        }
        return [returnData];
    }
}
exports.SpreadsheetFileV2 = SpreadsheetFileV2;
