// n8n-nodes-base/dist/nodes/Demio/ReportDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reportFields = exports.reportOperations = void 0;
exports.reportOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'report',
                ],
            },
        },
        options: [
            {
                name: 'Get',
                value: 'get',
                description: 'Get an event report',
            },
        ],
        default: 'get',
        description: 'The operation to perform.',
    },
];
exports.reportFields = [
    {
        displayName: 'Event ID',
        name: 'eventId',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getEvents',
        },
        displayOptions: {
            show: {
                resource: [
                    'report',
                ],
                operation: [
                    'get',
                ],
            },
        },
        default: '',
        description: 'Event ID',
    },
    {
        displayName: 'Session ID',
        name: 'dateId',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getEventSessions',
            loadOptionsDependsOn: [
                'eventId',
            ],
        },
        default: '',
        required: true,
        description: 'ID of the session',
        displayOptions: {
            show: {
                resource: [
                    'report',
                ],
                operation: [
                    'get',
                ],
            },
        },
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: [
                    'report',
                ],
                operation: [
                    'get',
                ],
            },
        },
        options: [
            {
                displayName: 'Status',
                name: 'status',
                type: 'options',
                options: [
                    {
                        name: 'Attended',
                        value: 'attended',
                    },
                    {
                        name: 'Banned',
                        value: 'banned',
                    },
                    {
                        name: 'Completed',
                        value: 'completed',
                    },
                    {
                        name: 'Did Not Attend',
                        value: 'did-not-attend',
                    },
                    {
                        name: 'Left Early',
                        value: 'left-early',
                    },
                ],
                default: '',
                description: 'Filter results by participation status',
            },
        ],
    },
];
