// n8n-nodes-base/dist/nodes/Ghost/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.ghostApiRequestAllItems = exports.ghostApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const jwt = require("jsonwebtoken");
async function ghostApiRequest(method, endpoint, body = {}, query = {}, uri) {
    const source = this.getNodeParameter('source', 0);
    let credentials;
    let version;
    let token;
    if (source === 'contentApi') {
        version = 'v3';
        credentials = await this.getCredentials('ghostContentApi');
        query.key = credentials.apiKey;
    }
    else {
        version = 'v2';
        credentials = await this.getCredentials('ghostAdminApi');
        const [id, secret] = credentials.apiKey.split(':');
        token = jwt.sign({}, Buffer.from(secret, 'hex'), {
            keyid: id,
            algorithm: 'HS256',
            expiresIn: '5m',
            audience: `/${version}/admin/`,
        });
    }
    const options = {
        method,
        qs: query,
        uri: uri || `${credentials.url}/ghost/api/${version}${endpoint}`,
        body,
        json: true,
    };
    if (token) {
        options.headers = {
            Authorization: `Ghost ${token}`,
        };
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.ghostApiRequest = ghostApiRequest;
async function ghostApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query.limit = 50;
    query.page = 1;
    do {
        responseData = await ghostApiRequest.call(this, method, endpoint, body, query);
        query.page = responseData.meta.pagination.next;
        returnData.push.apply(returnData, responseData[propertyName]);
    } while (query.page !== null);
    return returnData;
}
exports.ghostApiRequestAllItems = ghostApiRequestAllItems;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
