// n8n-nodes-base/dist/nodes/LoneScale/LoneScaleTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoneScaleTrigger = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
class LoneScaleTrigger {
    constructor() {
        this.description = {
            displayName: 'LoneScale Trigger',
            name: 'loneScaleTrigger',
            icon: 'file:lonescale-logo.svg',
            group: ['trigger'],
            version: 1,
            description: 'Trigger LoneScale Workflow',
            defaults: {
                name: 'LoneScale Trigger',
            },
            inputs: [],
            outputs: ['main'],
            credentials: [
                {
                    name: 'loneScaleApi',
                    required: true,
                },
            ],
            webhooks: [
                {
                    name: 'default',
                    httpMethod: 'POST',
                    responseMode: 'onReceived',
                    path: 'webhook',
                },
            ],
            properties: [
                {
                    displayName: 'Workflow Name',
                    name: 'workflow',
                    type: 'options',
                    noDataExpression: true,
                    typeOptions: {
                        loadOptionsMethod: 'getWorkflows',
                    },
                    default: '',
                    description: 'Select one workflow. Choose from the list',
                    required: true,
                },
            ],
        };
        this.methods = {
            loadOptions: {
                async getWorkflows() {
                    const data = await GenericFunctions_1.lonescaleApiRequest.call(this, 'GET', '/workflows');
                    return data === null || data === void 0 ? void 0 : data.map((d) => ({
                        name: d.title,
                        value: d.id,
                    }));
                },
            },
        };
        this.webhookMethods = {
            default: {
                async checkExists() {
                    const webhookData = this.getWorkflowStaticData('node');
                    const webhookUrl = this.getNodeWebhookUrl('default');
                    const workflowId = this.getNodeParameter('workflow');
                    const webhook = await GenericFunctions_1.lonescaleApiRequest.call(this, 'GET', `/workflows/${workflowId}/hook?type=n8n`);
                    if (webhook.target_url === webhookUrl) {
                        webhookData.webhookId = webhook.webhook_id;
                        return true;
                    }
                    return false;
                },
                async create() {
                    const webhookUrl = this.getNodeWebhookUrl('default');
                    const webhookData = this.getWorkflowStaticData('node');
                    const workflowId = this.getNodeParameter('workflow');
                    const body = {
                        type: 'n8n',
                        target_url: webhookUrl,
                    };
                    const webhook = await GenericFunctions_1.lonescaleApiRequest.call(this, 'POST', `/workflows/${workflowId}/hook`, body);
                    webhookData.webhookId = webhook.webhook_id;
                    return true;
                },
                async delete() {
                    const webhookData = this.getWorkflowStaticData('node');
                    try {
                        await GenericFunctions_1.lonescaleApiRequest.call(this, 'DELETE', `/workflows/${webhookData.webhookId}/hook?type=n8n`);
                    }
                    catch (error) {
                        return false;
                    }
                    delete webhookData.webhookId;
                    return true;
                },
            },
        };
    }
    async webhook() {
        const req = this.getRequestObject();
        return {
            workflowData: [this.helpers.returnJsonArray(req.body)],
        };
    }
}
exports.LoneScaleTrigger = LoneScaleTrigger;
