// @n8n/n8n-nodes-langchain/dist/nodes/llms/LmChatAwsBedrock/LmChatAwsBedrock.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LmChatAwsBedrock = void 0;
const bedrock_1 = require("@langchain/community/chat_models/bedrock");
const sharedFields_1 = require("../../../utils/sharedFields");
require("@aws-sdk/credential-provider-node");
require("@aws-sdk/client-bedrock-runtime");
const N8nLlmTracing_1 = require("../N8nLlmTracing");
class LmChatAwsBedrock {
    constructor() {
        this.description = {
            displayName: 'AWS Bedrock Chat Model',
            name: 'lmChatAwsBedrock',
            icon: 'file:bedrock.svg',
            group: ['transform'],
            version: 1,
            description: 'Language Model AWS Bedrock',
            defaults: {
                name: 'AWS Bedrock Chat Model',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Language Models'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.lmchatawsbedrock/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_languageModel"],
            outputNames: ['Model'],
            credentials: [
                {
                    name: 'aws',
                    required: true,
                },
            ],
            requestDefaults: {
                ignoreHttpStatusErrors: true,
                baseURL: '=https://bedrock.{{$credentials?.region ?? "eu-central-1"}}.amazonaws.com',
            },
            properties: [
                (0, sharedFields_1.getConnectionHintNoticeField)(["ai_chain", "ai_chain"]),
                {
                    displayName: 'Model',
                    name: 'model',
                    type: 'options',
                    description: 'The model which will generate the completion. <a href="https://docs.aws.amazon.com/bedrock/latest/userguide/foundation-models.html">Learn more</a>.',
                    typeOptions: {
                        loadOptions: {
                            routing: {
                                request: {
                                    method: 'GET',
                                    url: '/foundation-models?&byOutputModality=TEXT&byInferenceType=ON_DEMAND',
                                },
                                output: {
                                    postReceive: [
                                        {
                                            type: 'rootProperty',
                                            properties: {
                                                property: 'modelSummaries',
                                            },
                                        },
                                        {
                                            type: 'setKeyValue',
                                            properties: {
                                                name: '={{$responseItem.modelName}}',
                                                description: '={{$responseItem.modelArn}}',
                                                value: '={{$responseItem.modelId}}',
                                            },
                                        },
                                        {
                                            type: 'sort',
                                            properties: {
                                                key: 'name',
                                            },
                                        },
                                    ],
                                },
                            },
                        },
                    },
                    routing: {
                        send: {
                            type: 'body',
                            property: 'model',
                        },
                    },
                    default: '',
                },
                {
                    displayName: 'Options',
                    name: 'options',
                    placeholder: 'Add Option',
                    description: 'Additional options to add',
                    type: 'collection',
                    default: {},
                    options: [
                        {
                            displayName: 'Maximum Number of Tokens',
                            name: 'maxTokensToSample',
                            default: 2000,
                            description: 'The maximum number of tokens to generate in the completion',
                            type: 'number',
                        },
                        {
                            displayName: 'Sampling Temperature',
                            name: 'temperature',
                            default: 0.7,
                            typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
                            description: 'Controls randomness: Lowering results in less random completions. As the temperature approaches zero, the model will become deterministic and repetitive.',
                            type: 'number',
                        },
                    ],
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const credentials = await this.getCredentials('aws');
        const modelName = this.getNodeParameter('model', itemIndex);
        const options = this.getNodeParameter('options', itemIndex, {});
        const model = new bedrock_1.BedrockChat({
            region: credentials.region,
            model: modelName,
            temperature: options.temperature,
            maxTokens: options.maxTokensToSample,
            credentials: {
                secretAccessKey: credentials.secretAccessKey,
                accessKeyId: credentials.accessKeyId,
                sessionToken: credentials.sessionToken,
            },
            callbacks: [new N8nLlmTracing_1.N8nLlmTracing(this)],
        });
        return {
            response: model,
        };
    }
}
exports.LmChatAwsBedrock = LmChatAwsBedrock;
