// n8n-nodes-base/dist/nodes/Storyblok/StoryManagementDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storyManagementFields = exports.storyManagementOperations = void 0;
exports.storyManagementOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'delete',
                description: 'Delete a story',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Get a story',
            },
            {
                name: 'Get All',
                value: 'getAll',
                description: 'Get all stories',
            },
            {
                name: 'Publish',
                value: 'publish',
                description: 'Publish a story',
            },
            {
                name: 'Unpublish',
                value: 'unpublish',
                description: 'Unpublish a story',
            },
        ],
        default: 'get',
        description: 'The operation to perform.',
    },
];
exports.storyManagementFields = [
    {
        displayName: 'Space ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'delete',
                ],
            },
        },
        description: 'The name of the space.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'delete',
                ],
            },
        },
        description: 'Numeric ID of the story.',
    },
    {
        displayName: 'Space ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'get',
                ],
            },
        },
        description: 'The name of the space.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'get',
                ],
            },
        },
        description: 'Numeric ID of the story.',
    },
    {
        displayName: 'Space ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        description: 'The name of the space',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        default: false,
        description: 'Returns a list of your user contacts.',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'getAll',
                ],
                returnAll: [
                    false,
                ],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 100,
        },
        default: 50,
        description: 'How many results to return.',
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        default: {},
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'getAll',
                ],
            },
        },
        options: [
            {
                displayName: 'Starts With',
                name: 'starts_with',
                type: 'string',
                default: '',
                description: 'Filter by slug.',
            },
        ],
    },
    {
        displayName: 'Space ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'publish',
                ],
            },
        },
        description: 'The name of the space.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'publish',
                ],
            },
        },
        description: 'Numeric ID of the story.',
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'publish',
                ],
            },
        },
        default: {},
        options: [
            {
                displayName: 'Release ID',
                name: 'releaseId',
                type: 'string',
                default: '',
                description: 'Numeric ID of release.',
            },
            {
                displayName: 'Language',
                name: 'language',
                type: 'string',
                default: '',
                description: 'Language code to publish the story individually (must be enabled in the space settings).',
            },
        ],
    },
    {
        displayName: 'Space ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'unpublish',
                ],
            },
        },
        description: 'The name of the space.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: [
                    'managementApi',
                ],
                resource: [
                    'story',
                ],
                operation: [
                    'unpublish',
                ],
            },
        },
        description: 'Numeric ID of the story.',
    },
];
