// n8n-nodes-base/dist/nodes/Cockpit/SingletonDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.singletonFields = exports.singletonOperations = void 0;
exports.singletonOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'singleton',
                ],
            },
        },
        options: [
            {
                name: 'Get',
                value: 'get',
                description: 'Get a singleton',
            },
        ],
        default: 'get',
        description: 'The operation to perform.',
    },
];
exports.singletonFields = [
    {
        displayName: 'Singleton',
        name: 'singleton',
        type: 'options',
        default: '',
        typeOptions: {
            loadOptionsMethod: 'getSingletons',
        },
        displayOptions: {
            show: {
                resource: [
                    'singleton',
                ],
            },
        },
        required: true,
        description: 'Name of the singleton to operate on.',
    },
];
