// n8n-nodes-base/dist/nodes/Wufoo/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wufooApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function wufooApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    const credentials = await this.getCredentials('wufooApi');
    if (credentials === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials got returned!');
    }
    let options = {
        auth: {
            username: credentials.apiKey,
            password: '',
        },
        method,
        form: body,
        body,
        qs,
        uri: `https://${credentials.subdomain}.wufoo.com/api/v3/${resource}`,
        json: true,
    };
    options = Object.assign({}, options, option);
    if (Object.keys(options.body).length === 0 || method === 'PUT') {
        delete options.body;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.wufooApiRequest = wufooApiRequest;
