// n8n-nodes-base/dist/nodes/Postgres/Postgres.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Postgres = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const pgPromise = require("pg-promise");
const Postgres_node_functions_1 = require("./Postgres.node.functions");
class Postgres {
    constructor() {
        this.description = {
            displayName: 'Postgres',
            name: 'postgres',
            icon: 'file:postgres.svg',
            group: ['input'],
            version: 1,
            description: 'Get, add and update data in Postgres',
            defaults: {
                name: 'Postgres',
                color: '#336791',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'postgres',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Operation',
                    name: 'operation',
                    type: 'options',
                    options: [
                        {
                            name: 'Execute Query',
                            value: 'executeQuery',
                            description: 'Execute an SQL query',
                        },
                        {
                            name: 'Insert',
                            value: 'insert',
                            description: 'Insert rows in database',
                        },
                        {
                            name: 'Update',
                            value: 'update',
                            description: 'Update rows in database',
                        },
                    ],
                    default: 'insert',
                    description: 'The operation to perform.',
                },
                {
                    displayName: 'Query',
                    name: 'query',
                    type: 'string',
                    typeOptions: {
                        alwaysOpenEditWindow: true,
                    },
                    displayOptions: {
                        show: {
                            operation: ['executeQuery'],
                        },
                    },
                    default: '',
                    placeholder: 'SELECT id, name FROM product WHERE quantity > $1 AND price <= $2',
                    required: true,
                    description: 'The SQL query to execute. You can use n8n expressions or $1 and $2 in conjunction with query parameters.',
                },
                {
                    displayName: 'Schema',
                    name: 'schema',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['insert'],
                        },
                    },
                    default: 'public',
                    required: true,
                    description: 'Name of the schema the table belongs to',
                },
                {
                    displayName: 'Table',
                    name: 'table',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['insert'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Name of the table in which to insert data to.',
                },
                {
                    displayName: 'Columns',
                    name: 'columns',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['insert'],
                        },
                    },
                    default: '',
                    placeholder: 'id:int,name:text,description',
                    description: 'Comma separated list of the properties which should used as columns for the new rows.<br>You can use type casting with colons (:) like id:int.',
                },
                {
                    displayName: 'Schema',
                    name: 'schema',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['update'],
                        },
                    },
                    default: 'public',
                    required: false,
                    description: 'Name of the schema the table belongs to',
                },
                {
                    displayName: 'Table',
                    name: 'table',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['update'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Name of the table in which to update data in',
                },
                {
                    displayName: 'Update Key',
                    name: 'updateKey',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['update'],
                        },
                    },
                    default: 'id',
                    required: true,
                    description: 'Comma separated list of the properties which decides which rows in the database should be updated. Normally that would be "id".',
                },
                {
                    displayName: 'Columns',
                    name: 'columns',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['update'],
                        },
                    },
                    default: '',
                    placeholder: 'name:text,description',
                    description: 'Comma separated list of the properties which should used as columns for rows to update.<br>You can use type casting with colons (:) like id:int.',
                },
                {
                    displayName: 'Return Fields',
                    name: 'returnFields',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['insert', 'update'],
                        },
                    },
                    default: '*',
                    description: 'Comma separated list of the fields that the operation will return',
                },
                {
                    displayName: 'Additional Fields',
                    name: 'additionalFields',
                    type: 'collection',
                    placeholder: 'Add Field',
                    default: {},
                    options: [
                        {
                            displayName: 'Mode',
                            name: 'mode',
                            type: 'options',
                            options: [
                                {
                                    name: 'Independently',
                                    value: 'independently',
                                    description: 'Execute each query independently',
                                },
                                {
                                    name: 'Multiple queries',
                                    value: 'multiple',
                                    description: '<b>Default</b>. Sends multiple queries at once to database.',
                                },
                                {
                                    name: 'Transaction',
                                    value: 'transaction',
                                    description: 'Executes all queries in a single transaction',
                                },
                            ],
                            default: 'multiple',
                            description: [
                                'The way queries should be sent to database.',
                                'Can be used in conjunction with <b>Continue on Fail</b>.',
                                'See the docs for more examples',
                            ].join('<br>'),
                        },
                        {
                            displayName: 'Query Parameters',
                            name: 'queryParams',
                            type: 'string',
                            displayOptions: {
                                show: {
                                    '/operation': [
                                        'executeQuery',
                                    ],
                                },
                            },
                            default: '',
                            placeholder: 'quantity,price',
                            description: 'Comma separated list of properties which should be used as query parameters.',
                        },
                    ],
                },
            ],
        };
    }
    async execute() {
        const credentials = await this.getCredentials('postgres');
        if (credentials === undefined) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials got returned!');
        }
        const pgp = pgPromise();
        const config = {
            host: credentials.host,
            port: credentials.port,
            database: credentials.database,
            user: credentials.user,
            password: credentials.password,
        };
        if (credentials.allowUnauthorizedCerts === true) {
            config.ssl = {
                rejectUnauthorized: false,
            };
        }
        else {
            config.ssl = !['disable', undefined].includes(credentials.ssl);
            config.sslmode = credentials.ssl || 'disable';
        }
        const db = pgp(config);
        let returnItems = [];
        const items = this.getInputData();
        const operation = this.getNodeParameter('operation', 0);
        if (operation === 'executeQuery') {
            const queryResult = await Postgres_node_functions_1.pgQuery(this.getNodeParameter, pgp, db, items, this.continueOnFail());
            returnItems = this.helpers.returnJsonArray(queryResult);
        }
        else if (operation === 'insert') {
            const insertData = await Postgres_node_functions_1.pgInsert(this.getNodeParameter, pgp, db, items, this.continueOnFail());
            for (let i = 0; i < insertData.length; i++) {
                returnItems.push({
                    json: insertData[i],
                });
            }
        }
        else if (operation === 'update') {
            const updateItems = await Postgres_node_functions_1.pgUpdate(this.getNodeParameter, pgp, db, items, this.continueOnFail());
            returnItems = this.helpers.returnJsonArray(updateItems);
        }
        else {
            await pgp.end();
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `The operation "${operation}" is not supported!`);
        }
        await pgp.end();
        return this.prepareOutputData(returnItems);
    }
}
exports.Postgres = Postgres;
