// n8n/dist/push/websocket.push.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketPush = void 0;
const typedi_1 = require("typedi");
const Logger_1 = require("../Logger");
const abstract_push_1 = require("./abstract.push");
const orchestration_service_1 = require("../services/orchestration.service");
function heartbeat() {
    this.isAlive = true;
}
let WebSocketPush = class WebSocketPush extends abstract_push_1.AbstractPush {
    constructor(logger, orchestrationService) {
        super(logger, orchestrationService);
        setInterval(() => this.pingAll(), 60 * 1000);
    }
    add(pushRef, userId, connection) {
        connection.isAlive = true;
        connection.on('pong', heartbeat);
        super.add(pushRef, userId, connection);
        const onMessage = (data) => {
            try {
                const buffer = Array.isArray(data) ? Buffer.concat(data) : Buffer.from(data);
                this.onMessageReceived(pushRef, JSON.parse(buffer.toString('utf8')));
            }
            catch (error) {
                this.logger.error("Couldn't parse message from editor-UI", {
                    error: error,
                    pushRef,
                    data,
                });
            }
        };
        connection.once('close', () => {
            connection.off('pong', heartbeat);
            connection.off('message', onMessage);
            this.remove(pushRef);
        });
        connection.on('message', onMessage);
    }
    close(connection) {
        connection.close();
    }
    sendToOneConnection(connection, data) {
        connection.send(data);
    }
    pingAll() {
        for (const pushRef in this.connections) {
            const connection = this.connections[pushRef];
            if (!connection.isAlive) {
                delete this.connections[pushRef];
                return connection.terminate();
            }
            connection.isAlive = false;
            connection.ping();
        }
    }
};
exports.WebSocketPush = WebSocketPush;
exports.WebSocketPush = WebSocketPush = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [Logger_1.Logger, orchestration_service_1.OrchestrationService])
], WebSocketPush);
