// n8n-nodes-base/dist/nodes/TheHiveProject/methods/listSearch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.observableSearch = exports.logSearch = exports.pageSearch = exports.taskSearch = exports.alertSearch = exports.commentSearch = exports.caseSearch = void 0;
const transport_1 = require("../transport");
async function listResource(resource, filterField, nameField, urlPlaceholder, filter, paginationToken) {
    const query = [
        {
            _name: resource,
        },
    ];
    if (filter) {
        query.push({
            _name: 'filter',
            _like: {
                _field: filterField,
                _value: filter,
            },
        });
    }
    const from = paginationToken !== undefined ? parseInt(paginationToken, 10) : 0;
    const to = from + 100;
    query.push({
        _name: 'page',
        from,
        to,
    });
    const response = await transport_1.theHiveApiRequest.call(this, 'POST', '/v1/query', { query });
    if (response.length === 0) {
        return {
            results: [],
            paginationToken: undefined,
        };
    }
    const credentials = await this.getCredentials('theHiveProjectApi');
    const url = credentials === null || credentials === void 0 ? void 0 : credentials.url;
    return {
        results: response.map((entry) => ({
            name: entry[nameField],
            value: entry._id,
            url: urlPlaceholder !== undefined ? `${url}/${urlPlaceholder}/${entry._id}/details` : undefined,
        })),
        paginationToken: to,
    };
}
async function caseSearch(filter, paginationToken) {
    return await listResource.call(this, 'listCase', 'title', 'title', 'cases', filter, paginationToken);
}
exports.caseSearch = caseSearch;
async function commentSearch(filter, paginationToken) {
    return await listResource.call(this, 'listComment', 'message', 'message', undefined, filter, paginationToken);
}
exports.commentSearch = commentSearch;
async function alertSearch(filter, paginationToken) {
    return await listResource.call(this, 'listAlert', 'title', 'title', 'alerts', filter, paginationToken);
}
exports.alertSearch = alertSearch;
async function taskSearch(filter, paginationToken) {
    return await listResource.call(this, 'listTask', 'title', 'title', undefined, filter, paginationToken);
}
exports.taskSearch = taskSearch;
async function pageSearch(filter, paginationToken) {
    let caseId;
    try {
        caseId = this.getNodeParameter('caseId', '', { extractValue: true });
    }
    catch (error) {
        caseId = undefined;
    }
    let query;
    if (caseId) {
        query = [
            {
                _name: 'getCase',
                idOrName: caseId,
            },
            {
                _name: 'pages',
            },
        ];
    }
    else {
        query = [
            {
                _name: 'listOrganisationPage',
            },
        ];
    }
    if (filter) {
        query.push({
            _name: 'filter',
            _like: {
                _field: 'title',
                _value: filter,
            },
        });
    }
    const from = paginationToken !== undefined ? parseInt(paginationToken, 10) : 0;
    const to = from + 100;
    query.push({
        _name: 'page',
        from,
        to,
    });
    const response = await transport_1.theHiveApiRequest.call(this, 'POST', '/v1/query', { query });
    if (response.length === 0) {
        return {
            results: [],
            paginationToken: undefined,
        };
    }
    return {
        results: response.map((entry) => ({
            name: entry.title,
            value: entry._id,
        })),
        paginationToken: to,
    };
}
exports.pageSearch = pageSearch;
async function logSearch(filter, paginationToken) {
    return await listResource.call(this, 'listLog', 'message', 'message', undefined, filter, paginationToken);
}
exports.logSearch = logSearch;
async function observableSearch(filter, paginationToken) {
    const query = [
        {
            _name: 'listObservable',
        },
    ];
    if (filter) {
        query.push({
            _name: 'filter',
            _or: [
                {
                    _like: {
                        _field: 'data',
                        _value: filter,
                    },
                },
                {
                    _like: {
                        _field: 'message',
                        _value: filter,
                    },
                },
                {
                    _like: {
                        _field: 'attachment.name',
                        _value: filter,
                    },
                },
            ],
        });
    }
    const from = paginationToken !== undefined ? parseInt(paginationToken, 10) : 0;
    const to = from + 100;
    query.push({
        _name: 'page',
        from,
        to,
    });
    const response = await transport_1.theHiveApiRequest.call(this, 'POST', '/v1/query', { query });
    if (response.length === 0) {
        return {
            results: [],
            paginationToken: undefined,
        };
    }
    return {
        results: response.map((entry) => {
            var _a;
            return ({
                name: entry.data || ((_a = entry.attachment) === null || _a === void 0 ? void 0 : _a.name) || entry.message || entry._id,
                value: entry._id,
            });
        }),
        paginationToken: to,
    };
}
exports.observableSearch = observableSearch;
