// n8n-nodes-base/dist/nodes/Autopilot/ContactJourneyDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contactJourneyFields = exports.contactJourneyOperations = void 0;
exports.contactJourneyOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'contactJourney',
                ],
            },
        },
        options: [
            {
                name: 'Add',
                value: 'add',
                description: 'Add contact to list',
            },
        ],
        default: 'add',
        description: 'The operation to perform.',
    },
];
exports.contactJourneyFields = [
    {
        displayName: 'Trigger ID',
        name: 'triggerId',
        required: true,
        typeOptions: {
            loadOptionsMethod: 'getTriggers',
        },
        type: 'options',
        displayOptions: {
            show: {
                operation: [
                    'add',
                ],
                resource: [
                    'contactJourney',
                ],
            },
        },
        default: '',
        description: 'List ID.',
    },
    {
        displayName: 'Contact ID',
        name: 'contactId',
        required: true,
        type: 'string',
        displayOptions: {
            show: {
                operation: [
                    'add',
                ],
                resource: [
                    'contactJourney',
                ],
            },
        },
        default: '',
        description: 'Can be ID or email.',
    },
];
