// n8n-nodes-base/dist/nodes/Form/v2/FormTriggerV2.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormTriggerV2 = void 0;
const utils_1 = require("../utils");
const common_descriptions_1 = require("../common.descriptions");
const descriptionV2 = {
    displayName: 'n8n Form Trigger',
    name: 'formTrigger',
    icon: 'file:form.svg',
    group: ['trigger'],
    version: 2,
    description: 'Runs the flow when an n8n generated webform is submitted',
    defaults: {
        name: 'n8n Form Trigger',
    },
    inputs: [],
    outputs: ['main'],
    webhooks: [
        {
            name: 'setup',
            httpMethod: 'GET',
            responseMode: 'onReceived',
            isFullPath: true,
            path: '={{$parameter["path"]}}',
            ndvHideUrl: true,
            isForm: true,
        },
        {
            name: 'default',
            httpMethod: 'POST',
            responseMode: '={{$parameter["responseMode"]}}',
            responseData: '={{$parameter["responseMode"] === "lastNode" ? "noData" : undefined}}',
            isFullPath: true,
            path: '={{$parameter["path"]}}',
            ndvHideMethod: true,
            isForm: true,
        },
    ],
    eventTriggerDescription: 'Waiting for you to submit the form',
    activationMessage: 'You can now make calls to your production Form URL.',
    triggerPanel: common_descriptions_1.formTriggerPanel,
    properties: [
        common_descriptions_1.webhookPath,
        common_descriptions_1.formTitle,
        common_descriptions_1.formDescription,
        common_descriptions_1.formFields,
        common_descriptions_1.formRespondMode,
        {
            displayName: "In the 'Respond to Webhook' node, select 'Respond With JSON' and set the <strong>formSubmittedText</strong> key to display a custom response in the form, or the <strong>redirectURL</strong> key to redirect users to a URL",
            name: 'formNotice',
            type: 'notice',
            displayOptions: {
                show: { responseMode: ['responseNode'] },
            },
            default: '',
        },
        {
            displayName: 'Options',
            name: 'options',
            type: 'collection',
            placeholder: 'Add Option',
            default: {},
            options: [
                {
                    displayName: 'Append n8n Attribution',
                    name: 'appendAttribution',
                    type: 'boolean',
                    default: true,
                    description: 'Whether to include the link “Form automated with n8n” at the bottom of the form',
                },
                {
                    ...common_descriptions_1.respondWithOptions,
                    displayOptions: {
                        hide: {
                            '/responseMode': ['responseNode'],
                        },
                    },
                },
            ],
        },
    ],
};
class FormTriggerV2 {
    constructor(baseDescription) {
        this.description = {
            ...baseDescription,
            ...descriptionV2,
        };
    }
    async webhook() {
        return await (0, utils_1.formWebhook)(this);
    }
}
exports.FormTriggerV2 = FormTriggerV2;
