// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolWolframAlpha/ToolWolframAlpha.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolWolframAlpha = void 0;
const wolframalpha_1 = require("@langchain/community/tools/wolframalpha");
const logWrapper_1 = require("../../../utils/logWrapper");
const sharedFields_1 = require("../../../utils/sharedFields");
class ToolWolframAlpha {
    constructor() {
        this.description = {
            displayName: 'Wolfram|Alpha',
            name: 'toolWolframAlpha',
            icon: 'file:wolfram-alpha.svg',
            group: ['transform'],
            version: 1,
            description: "Connects to WolframAlpha's computational intelligence engine.",
            defaults: {
                name: 'Wolfram Alpha',
            },
            credentials: [
                {
                    name: 'wolframAlphaApi',
                    required: true,
                },
            ],
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolwolframalpha/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            properties: [(0, sharedFields_1.getConnectionHintNoticeField)(["ai_agent"])],
        };
    }
    async supplyData() {
        const credentials = await this.getCredentials('wolframAlphaApi');
        return {
            response: (0, logWrapper_1.logWrapper)(new wolframalpha_1.WolframAlphaTool({ appid: credentials.appId }), this),
        };
    }
}
exports.ToolWolframAlpha = ToolWolframAlpha;
